// @rds/angular-components/esm2022/lib/dialog/dialog-content.directive.mjs
import { ChangeDetectorRef, DestroyRef, Directive, ElementRef, HostBinding, inject, NgZone, Renderer2, } from '@angular/core';
import { concat, merge } from 'rxjs';
import { ContentObserver } from '@angular/cdk/observers';
import { DimensionsObserver } from '../utils';
import { debounceTime, take } from 'rxjs/operators';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import * as i0 from "@angular/core";
const HEADER_SELECTOR = '.rds-dialog__header';
const SCROLLABLE_CLASS = 'rds-dialog--scrollable';
const ACTIONS_SELECTOR = '.rds-dialog__actions-container';
export class RdsDialogContentDirective {
    constructor() {
        this.nativeElement = inject(ElementRef).nativeElement;
        this.destroyRef = inject(DestroyRef);
        this.renderer = inject(Renderer2);
        this.cdr = inject(ChangeDetectorRef);
        this.ngZone = inject(NgZone);
        this.contentObserver = inject(ContentObserver);
        this.dimensionsObserver = inject(DimensionsObserver);
        this._height = '100%';
        this.hostClass = 'rds-dialog__content';
    }
    get hostHeight() {
        return this._height;
    }
    ngAfterContentInit() {
        const refreshTriggers$ = merge(this.dimensionsObserver.observe(this.nativeElement), this.contentObserver.observe(this.nativeElement));
        // thanks to this concat we debounce all but first emission
        concat(refreshTriggers$.pipe(take(1)), refreshTriggers$.pipe(debounceTime(200)))
            .pipe(takeUntilDestroyed(this.destroyRef))
            .subscribe(() => {
            this.ngZone.runOutsideAngular(() => {
                this._refreshHeight();
            });
        });
    }
    _refreshHeight() {
        const parentElement = this.nativeElement.parentElement;
        this.toggleScrollableClass();
        const header = parentElement.querySelector(HEADER_SELECTOR);
        const actionsContainer = parentElement.querySelector(ACTIONS_SELECTOR);
        const actionsContainerHeight = actionsContainer?.scrollHeight || 0;
        Promise.resolve().then(() => {
            const headerHeight = header?.scrollHeight || 0;
            const newHeight = `calc(100% - ${headerHeight}px - ${actionsContainerHeight}px)`;
            if (this._height !== newHeight) {
                this.ngZone.run(() => {
                    this._height = newHeight;
                    this.cdr.markForCheck();
                });
            }
        });
    }
    toggleScrollableClass() {
        const rdsDialog = this.nativeElement.parentElement?.parentElement;
        if (!rdsDialog) {
            return;
        }
        if (this.nativeElement.scrollHeight > this.nativeElement.offsetHeight) {
            this.renderer.addClass(rdsDialog, SCROLLABLE_CLASS);
        }
        else {
            this.renderer.removeChild(rdsDialog, SCROLLABLE_CLASS);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDialogContentDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsDialogContentDirective, selector: "[rds-dialog-content], [rdsDialogContent]", host: { properties: { "class": "this.hostClass", "style.height": "this.hostHeight" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDialogContentDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: `[rds-dialog-content], [rdsDialogContent]`,
                }]
        }], propDecorators: { hostClass: [{
                type: HostBinding,
                args: ['class']
            }], hostHeight: [{
                type: HostBinding,
                args: ['style.height']
            }] } });
