// @rds/angular-components/esm2022/lib/dialog/dialog.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RdsDialogContainerComponent } from './dialog-container.component';
import { PortalModule } from '@angular/cdk/portal';
import { A11yModule } from '@angular/cdk/a11y';
import { RdsDialogHeaderDirective } from './dialog-header.directive';
import { RdsDialogActionsContainerDirective } from './dialog-actions-container.directive';
import { RdsDialogContentDirective } from './dialog-content.directive';
import { RdsDialogIconPlaceholderDirective } from './dialog-icon-placeholder.directive';
import { RdsDialogCloseDirective } from './dialog-close.directive';
import { OverlayModule } from '@angular/cdk/overlay';
import { ObserversModule } from '@angular/cdk/observers';
import * as i0 from "@angular/core";
const ELEMENTS = [
    RdsDialogHeaderDirective,
    RdsDialogActionsContainerDirective,
    RdsDialogContentDirective,
    RdsDialogIconPlaceholderDirective,
    RdsDialogCloseDirective,
];
export class RdsDialogModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDialogModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsDialogModule, declarations: [RdsDialogContainerComponent, RdsDialogHeaderDirective,
            RdsDialogActionsContainerDirective,
            RdsDialogContentDirective,
            RdsDialogIconPlaceholderDirective,
            RdsDialogCloseDirective], imports: [CommonModule,
            PortalModule,
            A11yModule,
            OverlayModule,
            ObserversModule], exports: [RdsDialogHeaderDirective,
            RdsDialogActionsContainerDirective,
            RdsDialogContentDirective,
            RdsDialogIconPlaceholderDirective,
            RdsDialogCloseDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDialogModule, imports: [CommonModule,
            PortalModule,
            A11yModule,
            OverlayModule,
            ObserversModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDialogModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        PortalModule,
                        A11yModule,
                        OverlayModule,
                        ObserversModule,
                    ],
                    declarations: [RdsDialogContainerComponent, ...ELEMENTS],
                    exports: [...ELEMENTS],
                }]
        }] });
