// @rds/angular-components/esm2022/lib/dialog/dialog-header.directive.mjs
import { ChangeDetectorRef, DestroyRef, Directive, ElementRef, HostBinding, inject, Input, NgZone, } from '@angular/core';
import { RdsDialogRef } from './model';
import { RdsDialogService } from './dialog.service';
import { getClosestDialog } from './get-closet-dialog';
import { debounceTime } from 'rxjs';
import { ContentObserver } from '@angular/cdk/observers';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import * as i0 from "@angular/core";
let nextId = 0;
export class RdsDialogHeaderDirective {
    constructor() {
        this.dialogService = inject(RdsDialogService);
        this.destroyRef = inject(DestroyRef);
        this.elementRef = inject(ElementRef);
        this.cdr = inject(ChangeDetectorRef);
        this.ngZone = inject(NgZone);
        this.contentObserver = inject(ContentObserver);
        this.dialogRef = inject(RdsDialogRef, {
            optional: true,
        });
        this._id = `rds-dialog-header-${nextId++}`;
        this._width = '100%';
        this.hostClass = 'rds-dialog__header';
    }
    get id() {
        return this._id;
    }
    set id(newValue) {
        this._id = newValue;
    }
    ngOnInit() {
        if (!this.dialogRef) {
            const dialogRef = getClosestDialog(this.elementRef, this.dialogService.openDialogs);
            if (dialogRef) {
                this.dialogRef = dialogRef;
            }
        }
        if (this.dialogRef) {
            Promise.resolve().then(() => {
                if (this.dialogRef && !this.dialogRef.dialogContainer.ariaLabelledBy) {
                    this.dialogRef.dialogContainer.ariaLabelledBy = this.id;
                }
            });
        }
    }
    ngAfterViewInit() {
        this.ngZone.runOutsideAngular(() => this._refreshWidth());
        this.contentObserver
            .observe(this.elementRef)
            .pipe(debounceTime(50), takeUntilDestroyed(this.destroyRef))
            .subscribe(() => this._refreshWidth());
    }
    _refreshWidth() {
        const contentElement = this.elementRef.nativeElement;
        const parentElement = contentElement.parentElement;
        const iconPlaceholder = parentElement.querySelector('.rds-dialog__icon-placeholder');
        const iconsWidth = iconPlaceholder?.clientWidth || 0;
        Promise.resolve().then(() => {
            const newWidth = `calc(100% - ${iconsWidth}px)`;
            if (newWidth !== this._width) {
                this.ngZone.run(() => {
                    this._width = newWidth;
                    this.cdr.markForCheck();
                });
            }
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDialogHeaderDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsDialogHeaderDirective, selector: "[rds-dialog-header], [rdsDialogHeader]", inputs: { id: "id" }, host: { properties: { "style.width": "this._width", "class": "this.hostClass", "id": "this.id" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDialogHeaderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rds-dialog-header], [rdsDialogHeader]',
                }]
        }], propDecorators: { _width: [{
                type: HostBinding,
                args: ['style.width']
            }], hostClass: [{
                type: HostBinding,
                args: ['class']
            }], id: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['id']
            }] } });
