// @rds/angular-components/esm2022/lib/dialog/dialog-close.directive.mjs
import { getClosestDialog } from './get-closet-dialog';
import { Directive, ElementRef, HostBinding, HostListener, inject, Input, } from '@angular/core';
import { RdsDialogRef } from './model';
import { RdsDialogService } from './dialog.service';
import * as i0 from "@angular/core";
export class RdsDialogCloseDirective {
    constructor() {
        this.elementRef = inject(ElementRef);
        this.dialogService = inject(RdsDialogService);
        this.dialogRef = inject(RdsDialogRef, {
            optional: true,
        });
        this._type = 'button';
    }
    get hostType() {
        return this._type;
    }
    set hostType(newValue) {
        this._type = newValue;
    }
    onHostClick() {
        this.dialogRef?.close(this.dialogResult);
    }
    ngOnInit() {
        if (!this.dialogRef) {
            this.dialogRef = getClosestDialog(this.elementRef, this.dialogService.openDialogs);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDialogCloseDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsDialogCloseDirective, selector: "[rds-dialog-close], [rdsDialogClose]", inputs: { dialogResult: ["rds-dialog-close", "dialogResult"], hostType: ["type", "hostType"] }, host: { listeners: { "click": "onHostClick()" }, properties: { "attr.type": "this.hostType" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDialogCloseDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rds-dialog-close], [rdsDialogClose]',
                }]
        }], propDecorators: { dialogResult: [{
                type: Input,
                args: ['rds-dialog-close']
            }], hostType: [{
                type: Input,
                args: ['type']
            }, {
                type: HostBinding,
                args: ['attr.type']
            }], onHostClick: [{
                type: HostListener,
                args: ['click']
            }] } });
