// @rds/angular-components/esm2022/lib/tab/tab-label-wrapper.directive.mjs
import { FocusMonitor, } from '@angular/cdk/a11y';
import { booleanAttribute, DestroyRef, Directive, ElementRef, inject, Input, } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import * as i0 from "@angular/core";
export class RdsTabLabelWrapperDirective {
    constructor() {
        this.elementRef = inject(ElementRef);
        this.focusMonitor = inject(FocusMonitor);
        this.destroyRef = inject(DestroyRef);
        this.disabled = false;
        this._focused = new Subject();
    }
    ngAfterViewInit() {
        this.focusMonitor
            .monitor(this.elementRef, true)
            .pipe(takeUntilDestroyed(this.destroyRef))
            .subscribe(() => this._focused.next(this));
    }
    ngOnDestroy() {
        this.focusMonitor.stopMonitoring(this.elementRef);
        this._focused.complete();
    }
    focus(origin, options) {
        this.focusMonitor.focusVia(this.elementRef.nativeElement, origin, options);
        this._focused.next(this);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTabLabelWrapperDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "16.1.0", version: "17.3.1", type: RdsTabLabelWrapperDirective, selector: "[rdsTabLabelWrapper]", inputs: { disabled: ["disabled", "disabled", booleanAttribute] }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTabLabelWrapperDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rdsTabLabelWrapper]',
                }]
        }], propDecorators: { disabled: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }] } });
