// @rds/angular-components/esm2022/lib/tab/tab.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RdsTabsComponent } from './tabs.component';
import { RdsTabComponent } from './tab.component';
import { RdsTabLabelDirective } from './tab-label.directive';
import { PortalModule } from '@angular/cdk/portal';
import { A11yModule } from '@angular/cdk/a11y';
import { RdsTabLabelWrapperDirective } from './tab-label-wrapper.directive';
import { ObserversModule } from '@angular/cdk/observers';
import * as i0 from "@angular/core";
const componentsToExpert = [
    RdsTabsComponent,
    RdsTabComponent,
    RdsTabLabelDirective,
    RdsTabLabelWrapperDirective,
];
export class RdsTabModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTabModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsTabModule, declarations: [RdsTabsComponent,
            RdsTabComponent,
            RdsTabLabelDirective,
            RdsTabLabelWrapperDirective], imports: [CommonModule, PortalModule, A11yModule, ObserversModule], exports: [RdsTabsComponent,
            RdsTabComponent,
            RdsTabLabelDirective,
            RdsTabLabelWrapperDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTabModule, imports: [CommonModule, PortalModule, A11yModule, ObserversModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTabModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [...componentsToExpert],
                    exports: [...componentsToExpert],
                    imports: [CommonModule, PortalModule, A11yModule, ObserversModule],
                }]
        }] });
