// @rds/angular-components/esm2022/lib/tab/tab.component.mjs
import { booleanAttribute, Component, ContentChild, EventEmitter, Input, ViewEncapsulation, } from '@angular/core';
import { RdsTabLabelDirective } from './tab-label.directive';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/observers";
export class RdsTabComponent {
    constructor() {
        this._active = false;
        this.disabled = false;
        this.stateChanges = new EventEmitter();
    }
    set active(value) {
        const newValue = booleanAttribute(value);
        if (newValue !== this.active) {
            this._active = newValue;
            this.stateChanges.next();
        }
    }
    get active() {
        return this._active;
    }
    get templateLabel() {
        return this._templateLabel;
    }
    set templateLabel(value) {
        if (value) {
            this._templateLabel = value;
        }
    }
    ngOnChanges(changes) {
        if (changes['label'] || changes['disabled']) {
            this.stateChanges.next();
        }
    }
    ngOnDestroy() {
        this.stateChanges.complete();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTabComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "16.1.0", version: "17.3.1", type: RdsTabComponent, selector: "rds-tab", inputs: { label: "label", active: "active", disabled: ["disabled", "disabled", booleanAttribute] }, queries: [{ propertyName: "templateLabel", first: true, predicate: RdsTabLabelDirective, descendants: true }], usesOnChanges: true, ngImport: i0, template: `
    <div [hidden]="!active" (cdkObserveContent)="stateChanges.next()">
      <ng-content />
    </div>
  `, isInline: true, dependencies: [{ kind: "directive", type: i1.CdkObserveContent, selector: "[cdkObserveContent]", inputs: ["cdkObserveContentDisabled", "debounce"], outputs: ["cdkObserveContent"], exportAs: ["cdkObserveContent"] }], encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTabComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'rds-tab',
                    template: `
    <div [hidden]="!active" (cdkObserveContent)="stateChanges.next()">
      <ng-content />
    </div>
  `,
                    encapsulation: ViewEncapsulation.None,
                }]
        }], propDecorators: { label: [{
                type: Input
            }], active: [{
                type: Input
            }], disabled: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }], templateLabel: [{
                type: ContentChild,
                args: [RdsTabLabelDirective]
            }] } });
