// @rds/angular-components/esm2022/lib/progress-spinner/progress-spinner.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RdsProgressSpinnerComponent } from './progress-spinner.component';
import { LottieComponent } from 'ngx-lottie';
import * as i0 from "@angular/core";
export class RdsProgressSpinnerModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsProgressSpinnerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsProgressSpinnerModule, declarations: [RdsProgressSpinnerComponent], imports: [CommonModule, LottieComponent], exports: [RdsProgressSpinnerComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsProgressSpinnerModule, imports: [CommonModule, LottieComponent] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsProgressSpinnerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [RdsProgressSpinnerComponent],
                    exports: [RdsProgressSpinnerComponent],
                    imports: [CommonModule, LottieComponent],
                }]
        }] });
