// @rds/angular-components/esm2022/lib/table/cell.mjs
import { Directive, HostBinding, InjectionToken, Input } from '@angular/core';
import { CdkCell, CdkCellDef, CdkColumnDef, CdkFooterCell, CdkFooterCellDef, CdkHeaderCell, CdkHeaderCellDef, } from '@angular/cdk/table';
import * as i0 from "@angular/core";
export class RdsCellDefDirective extends CdkCellDef {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCellDefDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsCellDefDirective, selector: "[rds-cell-def], [rdsCellDef]", providers: [{ provide: CdkCellDef, useExisting: RdsCellDefDirective }], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCellDefDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rds-cell-def], [rdsCellDef]',
                    providers: [{ provide: CdkCellDef, useExisting: RdsCellDefDirective }],
                }]
        }] });
export class RdsHeaderCellDefDirective extends CdkHeaderCellDef {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderCellDefDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsHeaderCellDefDirective, selector: "[rds-header-cell-def], [rdsHeaderCellDef]", providers: [
            { provide: CdkHeaderCellDef, useExisting: RdsHeaderCellDefDirective },
        ], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderCellDefDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rds-header-cell-def], [rdsHeaderCellDef]',
                    providers: [
                        { provide: CdkHeaderCellDef, useExisting: RdsHeaderCellDefDirective },
                    ],
                }]
        }] });
export class RdsFooterCellDefDirective extends CdkFooterCellDef {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsFooterCellDefDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsFooterCellDefDirective, selector: "[rds-footer-cell-def], [rdsFooterCellDef]", providers: [
            { provide: CdkFooterCellDef, useExisting: RdsFooterCellDefDirective },
        ], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsFooterCellDefDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rds-footer-cell-def], [rdsFooterCellDef]',
                    providers: [
                        { provide: CdkFooterCellDef, useExisting: RdsFooterCellDefDirective },
                    ],
                }]
        }] });
export const RDS_SORT_HEADER_COLUMN_DEF = new InjectionToken('RDS_SORT_HEADER_COLUMN_DEF');
export class RdsColumnDefDirective extends CdkColumnDef {
    get name() {
        return this._name;
    }
    set name(name) {
        this._setNameInput(name);
    }
    _updateColumnCssClassName() {
        super._updateColumnCssClassName();
        this._columnCssClassName.push(`rds-column-${this.cssClassFriendlyName}`);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsColumnDefDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsColumnDefDirective, selector: "[rds-column-def], [rdsColumnDef]", inputs: { name: ["rdsColumnDef", "name"] }, providers: [
            { provide: CdkColumnDef, useExisting: RdsColumnDefDirective },
            {
                provide: RDS_SORT_HEADER_COLUMN_DEF,
                useExisting: RdsColumnDefDirective,
            },
        ], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsColumnDefDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rds-column-def], [rdsColumnDef]',
                    providers: [
                        { provide: CdkColumnDef, useExisting: RdsColumnDefDirective },
                        {
                            provide: RDS_SORT_HEADER_COLUMN_DEF,
                            useExisting: RdsColumnDefDirective,
                        },
                    ],
                }]
        }], propDecorators: { name: [{
                type: Input,
                args: ['rdsColumnDef']
            }] } });
export class RdsHeaderCellDirective extends CdkHeaderCell {
    constructor() {
        super(...arguments);
        this.class = 'rds-header-cell';
        this.role = 'columnheader';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderCellDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsHeaderCellDirective, selector: "rds-header-cell, th[rdsHeaderCell], th[rds-header-cell]", host: { properties: { "class": "this.class", "attr.role": "this.role" } }, usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderCellDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'rds-header-cell, th[rdsHeaderCell], th[rds-header-cell]',
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }], role: [{
                type: HostBinding,
                args: ['attr.role']
            }] } });
export class RdsFooterCellDirective extends CdkFooterCell {
    constructor() {
        super(...arguments);
        this.class = 'rds-footer-cell';
        this.role = 'gridcell';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsFooterCellDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsFooterCellDirective, selector: "rds-footer-cell, td[rdsFooterCell], td[rds-footer-cell]", host: { properties: { "class": "this.class", "attr.role": "this.role" } }, usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsFooterCellDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'rds-footer-cell, td[rdsFooterCell], td[rds-footer-cell]',
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }], role: [{
                type: HostBinding,
                args: ['attr.role']
            }] } });
export class RdsCellDirective extends CdkCell {
    constructor() {
        super(...arguments);
        this.class = 'rds-cell';
        this.role = 'gridcell';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCellDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsCellDirective, selector: "rds-cell, td[rdsCell], td[rds-cell]", host: { properties: { "class": "this.class", "attr.role": "this.role" } }, usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCellDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'rds-cell, td[rdsCell], td[rds-cell]',
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }], role: [{
                type: HostBinding,
                args: ['attr.role']
            }] } });
