// @rds/angular-components/esm2022/lib/table/row.mjs
import { CDK_ROW_TEMPLATE, CdkFooterRow, CdkFooterRowDef, CdkHeaderRow, CdkHeaderRowDef, CdkNoDataRow, CdkRow, CdkRowDef, } from '@angular/cdk/table';
import { booleanAttribute, ChangeDetectionStrategy, Component, Directive, HostBinding, Input, ViewEncapsulation, } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/table";
export class RdsHeaderRowDefDirective extends CdkHeaderRowDef {
    get sticky() {
        return super.sticky;
    }
    set sticky(v) {
        super.sticky = booleanAttribute(v);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderRowDefDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsHeaderRowDefDirective, selector: "[rds-header-row-def], [rdsHeaderRowDef]", inputs: { columns: ["rdsHeaderRowDef", "columns"], sticky: ["rdsHeaderRowDefSticky", "sticky"] }, providers: [
            { provide: CdkHeaderRowDef, useExisting: RdsHeaderRowDefDirective },
        ], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderRowDefDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rds-header-row-def], [rdsHeaderRowDef]',
                    providers: [
                        { provide: CdkHeaderRowDef, useExisting: RdsHeaderRowDefDirective },
                    ],
                }]
        }], propDecorators: { columns: [{
                type: Input,
                args: ['rdsHeaderRowDef']
            }], sticky: [{
                type: Input,
                args: ['rdsHeaderRowDefSticky']
            }] } });
export class RdsFooterRowDefDirective extends CdkFooterRowDef {
    set sticky(v) {
        super.sticky = booleanAttribute(v);
    }
    get sticky() {
        return super.sticky;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsFooterRowDefDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsFooterRowDefDirective, selector: "[rds-footer-row-def], [rdsFooterRowDef]", inputs: { columns: ["rdsFooterRowDef", "columns"], sticky: ["rdsFooterRowDefSticky", "sticky"] }, providers: [
            { provide: CdkFooterRowDef, useExisting: RdsFooterRowDefDirective },
        ], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsFooterRowDefDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rds-footer-row-def], [rdsFooterRowDef]',
                    providers: [
                        { provide: CdkFooterRowDef, useExisting: RdsFooterRowDefDirective },
                    ],
                }]
        }], propDecorators: { columns: [{
                type: Input,
                args: ['rdsFooterRowDef']
            }], sticky: [{
                type: Input,
                args: ['rdsFooterRowDefSticky']
            }] } });
export class RdsRowDefDirective extends CdkRowDef {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsRowDefDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsRowDefDirective, selector: "[rds-row-def], [rdsRowDef]", inputs: { columns: ["rdsRowDefColumns", "columns"], when: ["rdsRowDefWhen", "when"] }, providers: [{ provide: CdkRowDef, useExisting: RdsRowDefDirective }], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsRowDefDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rds-row-def], [rdsRowDef]',
                    providers: [{ provide: CdkRowDef, useExisting: RdsRowDefDirective }],
                }]
        }], propDecorators: { columns: [{
                type: Input,
                args: ['rdsRowDefColumns']
            }], when: [{
                type: Input,
                args: ['rdsRowDefWhen']
            }] } });
export class RdsHeaderRowComponent extends CdkHeaderRow {
    constructor() {
        super(...arguments);
        this.class = 'rds-header-row';
        this.role = 'row';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderRowComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsHeaderRowComponent, selector: "rds-header-row, tr[rds-header-row], tr[rdsHeaderRow]", host: { properties: { "class": "this.class", "attr.role": "this.role" } }, providers: [{ provide: CdkHeaderRow, useExisting: RdsHeaderRowComponent }], exportAs: ["rdsHeaderRow"], usesInheritance: true, ngImport: i0, template: "<ng-container cdkCellOutlet></ng-container>", isInline: true, dependencies: [{ kind: "directive", type: i1.CdkCellOutlet, selector: "[cdkCellOutlet]" }], changeDetection: i0.ChangeDetectionStrategy.Default, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderRowComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'rds-header-row, tr[rds-header-row], tr[rdsHeaderRow]',
                    template: CDK_ROW_TEMPLATE,
                    changeDetection: ChangeDetectionStrategy.Default,
                    encapsulation: ViewEncapsulation.None,
                    exportAs: 'rdsHeaderRow',
                    providers: [{ provide: CdkHeaderRow, useExisting: RdsHeaderRowComponent }],
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }], role: [{
                type: HostBinding,
                args: ['attr.role']
            }] } });
export class RdsFooterRowComponent extends CdkFooterRow {
    constructor() {
        super(...arguments);
        this.class = 'rds-footer-row';
        this.role = 'row';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsFooterRowComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsFooterRowComponent, selector: "rds-footer-row, tr[rds-footer-row], tr[rdsFooterRow]", host: { properties: { "class": "this.class", "attr.role": "this.role" } }, providers: [{ provide: CdkFooterRow, useExisting: RdsFooterRowComponent }], exportAs: ["rdsFooterRow"], usesInheritance: true, ngImport: i0, template: "<ng-container cdkCellOutlet></ng-container>", isInline: true, dependencies: [{ kind: "directive", type: i1.CdkCellOutlet, selector: "[cdkCellOutlet]" }], changeDetection: i0.ChangeDetectionStrategy.Default, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsFooterRowComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'rds-footer-row, tr[rds-footer-row], tr[rdsFooterRow]',
                    template: CDK_ROW_TEMPLATE,
                    changeDetection: ChangeDetectionStrategy.Default,
                    encapsulation: ViewEncapsulation.None,
                    exportAs: 'rdsFooterRow',
                    providers: [{ provide: CdkFooterRow, useExisting: RdsFooterRowComponent }],
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }], role: [{
                type: HostBinding,
                args: ['attr.role']
            }] } });
export class RdsRowComponent extends CdkRow {
    constructor() {
        super(...arguments);
        this.class = 'rds-row';
        this.role = 'row';
        this.selected = false;
        this.disabled = false;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsRowComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "16.1.0", version: "17.3.1", type: RdsRowComponent, selector: "rds-row, tr[rds-row], tr[rdsRow]", inputs: { selected: ["selected", "selected", booleanAttribute], disabled: ["disabled", "disabled", booleanAttribute] }, host: { properties: { "class": "this.class", "attr.role": "this.role", "class.selected": "this.selected", "class.disabled": "this.disabled" } }, providers: [{ provide: CdkRow, useExisting: RdsRowComponent }], exportAs: ["rdsRow"], usesInheritance: true, ngImport: i0, template: "<ng-container cdkCellOutlet></ng-container>", isInline: true, dependencies: [{ kind: "directive", type: i1.CdkCellOutlet, selector: "[cdkCellOutlet]" }], changeDetection: i0.ChangeDetectionStrategy.Default, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsRowComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'rds-row, tr[rds-row], tr[rdsRow]',
                    template: CDK_ROW_TEMPLATE,
                    changeDetection: ChangeDetectionStrategy.Default,
                    encapsulation: ViewEncapsulation.None,
                    exportAs: 'rdsRow',
                    providers: [{ provide: CdkRow, useExisting: RdsRowComponent }],
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }], role: [{
                type: HostBinding,
                args: ['attr.role']
            }], selected: [{
                type: HostBinding,
                args: ['class.selected']
            }, {
                type: Input,
                args: [{ transform: booleanAttribute }]
            }], disabled: [{
                type: HostBinding,
                args: ['class.disabled']
            }, {
                type: Input,
                args: [{ transform: booleanAttribute }]
            }] } });
export class RdsNoDataRowDirective extends CdkNoDataRow {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsNoDataRowDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsNoDataRowDirective, selector: "ng-template[rds-no-data-row], ng-template[rdsNoDataRow]", providers: [{ provide: CdkNoDataRow, useExisting: RdsNoDataRowDirective }], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsNoDataRowDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ng-template[rds-no-data-row], ng-template[rdsNoDataRow]',
                    providers: [{ provide: CdkNoDataRow, useExisting: RdsNoDataRowDirective }],
                }]
        }] });
