// @rds/angular-components/esm2022/lib/table/sort-header.mjs
import { booleanAttribute, ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, HostBinding, HostListener, inject, Input, ViewEncapsulation, } from '@angular/core';
import { FocusMonitor } from '@angular/cdk/a11y';
import { ENTER, SPACE } from '@angular/cdk/keycodes';
import { merge } from 'rxjs';
import { RdsSortDirective } from './sort.directive';
import { animate, animateChild, keyframes, query, state, style, transition, trigger, } from '@angular/animations';
import { RDS_SORT_HEADER_COLUMN_DEF } from './cell';
import * as i0 from "@angular/core";
import * as i1 from "../icons/icon.component";
const SORT_ANIMATION_TRANSITION = '225ms cubic-bezier(0.4,0.0,0.2,1)';
const rdsSortAnimations = {
    arrowOpacity: trigger('arrowOpacity', [
        state('desc-to-active, asc-to-active, active', style({ opacity: 1 })),
        state('desc-to-hint, asc-to-hint, hint', style({ opacity: 0.54 })),
        state('hint-to-desc, active-to-desc, desc, hint-to-asc, active-to-asc, asc, void', style({ opacity: 0 })),
        transition('* => asc, * => desc, * => active, * => hint, * => void', animate('0ms')),
        transition('* <=> *', animate(SORT_ANIMATION_TRANSITION)),
    ]),
    arrowPosition: trigger('arrowPosition', [
        transition('* => desc-to-hint, * => desc-to-active', animate(SORT_ANIMATION_TRANSITION, keyframes([
            style({ transform: 'translateY(-25%)' }),
            style({ transform: 'translateY(0)' }),
        ]))),
        transition('* => hint-to-desc, * => active-to-desc', animate(SORT_ANIMATION_TRANSITION, keyframes([
            style({ transform: 'translateY(0)' }),
            style({ transform: 'translateY(25%)' }),
        ]))),
        transition('* => asc-to-hint, * => asc-to-active', animate(SORT_ANIMATION_TRANSITION, keyframes([
            style({ transform: 'translateY(25%)' }),
            style({ transform: 'translateY(0)' }),
        ]))),
        transition('* => hint-to-asc, * => active-to-asc', animate(SORT_ANIMATION_TRANSITION, keyframes([
            style({ transform: 'translateY(0)' }),
            style({ transform: 'translateY(-25%)' }),
        ]))),
        state('desc-to-hint, asc-to-hint, hint, desc-to-active, asc-to-active, active', style({ transform: 'translateY(0)' })),
        state('hint-to-desc, active-to-desc, desc', style({ transform: 'translateY(-25%)' })),
        state('hint-to-asc, active-to-asc, asc', style({ transform: 'translateY(25%)' })),
    ]),
    allowChildren: trigger('allowChildren', [
        transition('* <=> *', [query('@*', animateChild(), { optional: true })]),
    ]),
    arrow: trigger('arrow', [
        state('asc', style({ transform: '' })),
        state('desc', style({ transform: 'scaleY(-1)' })),
        transition('asc => desc', [animate(SORT_ANIMATION_TRANSITION)]),
        transition('desc => asc', [animate(SORT_ANIMATION_TRANSITION)]),
    ]),
};
export class RdsSortHeaderComponent {
    get _disabledClass() {
        return this._isDisabled();
    }
    get ariaSortAttribute() {
        if (!this._isSorted()) {
            return 'none';
        }
        return this.sort.direction == 'asc' ? 'ascending' : 'descending';
    }
    constructor() {
        this.cdr = inject(ChangeDetectorRef);
        this.sort = inject(RdsSortDirective, {
            optional: true,
        });
        this.columnDef = inject(RDS_SORT_HEADER_COLUMN_DEF, {
            optional: true,
        });
        this.focusMonitor = inject(FocusMonitor);
        this.elementRef = inject(ElementRef);
        this._rerenderSubscription = null;
        this.class = 'rds-sort-header';
        this._showIndicatorHint = false;
        this._viewState = null;
        this._arrowDirection = '';
        this._disableViewStateAnimation = false;
        // eslint-disable-next-line @angular-eslint/no-input-rename
        this.id = '';
        this.arrowPosition = 'right';
        this.start = '';
        this.disableClear = false;
        this.disabled = false;
        if (!this.sort) {
            throw Error(`RdsSortHeader must be placed within a parent element with the RdsSort directive.`);
        }
        this._handleStateChanges();
    }
    ngOnInit() {
        if (!this.id && this.columnDef) {
            this.id = this.columnDef.name;
        }
        this._updateArrowDirection();
        this._setAnimationTransitionState({
            toState: this._isSorted() ? 'active' : this._arrowDirection,
        });
        this.sort.register(this);
    }
    ngAfterViewInit() {
        this.focusMonitor.monitor(this.elementRef, true).subscribe((origin) => {
            const newState = !!origin;
            if (newState !== this._showIndicatorHint) {
                this._setIndicatorHintVisible(newState);
                this.cdr.markForCheck();
            }
        });
    }
    ngOnDestroy() {
        this.focusMonitor.stopMonitoring(this.elementRef);
        this.sort.deregister(this);
        this._rerenderSubscription?.unsubscribe();
    }
    onMouseenter() {
        this._setIndicatorHintVisible(true);
    }
    onMouseleave() {
        this._setIndicatorHintVisible(false);
    }
    _handleClick() {
        if (!this._isDisabled()) {
            this.sort.sort(this);
        }
    }
    _handleKeydown(event) {
        if (!this._isDisabled() &&
            (event.keyCode === SPACE || event.keyCode === ENTER)) {
            event.preventDefault();
            this._toggleOnInteraction();
        }
    }
    _isSorted() {
        return (this.sort.active == this.id &&
            (this.sort.direction === 'asc' || this.sort.direction === 'desc'));
    }
    _isDisabled() {
        return this.sort.disabled || this.disabled;
    }
    _renderArrow() {
        return !this._isDisabled() || this._isSorted();
    }
    _getArrowDirectionState() {
        return `${this._isSorted() ? 'active-' : ''}${this._arrowDirection}`;
    }
    _getArrowViewState() {
        const fromState = this._viewState?.fromState;
        return (fromState ? `${fromState}-to-` : '') + this._viewState?.toState;
    }
    _updateArrowDirection() {
        this._arrowDirection = this._isSorted()
            ? this.sort.direction
            : this.start || this.sort.start;
    }
    _handleStateChanges() {
        this._rerenderSubscription = merge(this.sort.rdsSortChange, this.sort._stateChanges).subscribe(() => {
            if (this._isSorted()) {
                this._updateArrowDirection();
                if (this._viewState?.toState === 'hint' ||
                    this._viewState?.toState === 'active') {
                    this._disableViewStateAnimation = true;
                }
                this._setAnimationTransitionState({
                    fromState: this._arrowDirection,
                    toState: 'active',
                });
                this._showIndicatorHint = false;
            }
            if (!this._isSorted() &&
                this._viewState &&
                this._viewState.toState === 'active') {
                this._disableViewStateAnimation = false;
                this._setAnimationTransitionState({
                    fromState: 'active',
                    toState: this._arrowDirection,
                });
            }
            this.cdr.markForCheck();
        });
    }
    _setIndicatorHintVisible(visible) {
        if (this._isDisabled() && visible) {
            return;
        }
        this._showIndicatorHint = visible;
        if (!this._isSorted()) {
            this._updateArrowDirection();
            if (this._showIndicatorHint) {
                this._setAnimationTransitionState({
                    fromState: this._arrowDirection,
                    toState: 'hint',
                });
            }
            else {
                this._setAnimationTransitionState({
                    fromState: 'hint',
                    toState: this._arrowDirection,
                });
            }
        }
    }
    _setAnimationTransitionState(viewState) {
        this._viewState = viewState;
        if (this._disableViewStateAnimation) {
            this._viewState = { toState: viewState.toState };
        }
    }
    _toggleOnInteraction() {
        this.sort.sort(this);
        if (this._viewState?.toState === 'hint' ||
            this._viewState?.toState === 'active') {
            this._disableViewStateAnimation = true;
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSortHeaderComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "17.3.1", type: RdsSortHeaderComponent, selector: "[rds-sort-header], [rdsSortHeader]", inputs: { id: ["rds-sort-header", "id"], arrowPosition: "arrowPosition", start: "start", disableClear: ["disableClear", "disableClear", booleanAttribute], disabled: ["disabled", "disabled", booleanAttribute] }, host: { listeners: { "mouseenter": "onMouseenter()", "mouseleave": "onMouseleave()", "click": "_handleClick()", "keydown": "_handleKeydown($event)" }, properties: { "class": "this.class", "class.rds-sort-header-disabled": "this._disabledClass", "attr.aria-sort": "this.ariaSortAttribute" } }, exportAs: ["rdsSortHeader"], ngImport: i0, template: "<div\n  class=\"rds-sort-header-container rds-focus-indicator\"\n  [class.rds-sort-header-sorted]=\"_isSorted()\"\n  [class.rds-sort-header-position-left]=\"arrowPosition === 'left'\"\n  [attr.tabindex]=\"_isDisabled() ? null : 0\"\n  role=\"button\"\n>\n  <div class=\"rds-sort-header-content\">\n    <ng-content />\n  </div>\n  @if (_renderArrow()) {\n  <div\n    class=\"rds-sort-header-arrow\"\n    [@arrowOpacity]=\"_getArrowViewState()\"\n    [@arrowPosition]=\"_getArrowViewState()\"\n    [@allowChildren]=\"_getArrowDirectionState()\"\n    (@arrowPosition.start)=\"_disableViewStateAnimation = true\"\n    (@arrowPosition.done)=\"_disableViewStateAnimation = false\"\n  >\n    <rds-icon [@arrow]=\"_arrowDirection\" namespace=\"outlined\" icon=\"arrow_up\" />\n  </div>\n  }\n</div>\n", styles: ["[rds-sort-header]:hover,[rds-sort-header]:focus{outline:none}.rds-sort-header-container{display:flex;cursor:pointer;align-items:center;outline:0}.rds-sort-header-disabled .rds-sort-header-container{cursor:default}.rds-sort-header-content{text-align:center;display:flex;align-items:center;width:100%}.rds-sort-header-position-left{flex-direction:row-reverse}.rds-sort-header-arrow{position:relative;display:flex;opacity:0;margin:0 0 0 8px}.rds-sort-header-arrow rds-icon{height:18px;width:18px}.rds-sort-header-position-left .rds-sort-header-arrow{margin:0 8px 0 0}\n"], dependencies: [{ kind: "component", type: i1.RdsIconComponent, selector: "rds-icon", inputs: ["namespace", "icon"] }], animations: [
            rdsSortAnimations.arrowOpacity,
            rdsSortAnimations.arrowPosition,
            rdsSortAnimations.allowChildren,
            rdsSortAnimations.arrow,
        ], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSortHeaderComponent, decorators: [{
            type: Component,
            args: [{ selector: '[rds-sort-header], [rdsSortHeader]', exportAs: 'rdsSortHeader', encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, animations: [
                        rdsSortAnimations.arrowOpacity,
                        rdsSortAnimations.arrowPosition,
                        rdsSortAnimations.allowChildren,
                        rdsSortAnimations.arrow,
                    ], template: "<div\n  class=\"rds-sort-header-container rds-focus-indicator\"\n  [class.rds-sort-header-sorted]=\"_isSorted()\"\n  [class.rds-sort-header-position-left]=\"arrowPosition === 'left'\"\n  [attr.tabindex]=\"_isDisabled() ? null : 0\"\n  role=\"button\"\n>\n  <div class=\"rds-sort-header-content\">\n    <ng-content />\n  </div>\n  @if (_renderArrow()) {\n  <div\n    class=\"rds-sort-header-arrow\"\n    [@arrowOpacity]=\"_getArrowViewState()\"\n    [@arrowPosition]=\"_getArrowViewState()\"\n    [@allowChildren]=\"_getArrowDirectionState()\"\n    (@arrowPosition.start)=\"_disableViewStateAnimation = true\"\n    (@arrowPosition.done)=\"_disableViewStateAnimation = false\"\n  >\n    <rds-icon [@arrow]=\"_arrowDirection\" namespace=\"outlined\" icon=\"arrow_up\" />\n  </div>\n  }\n</div>\n", styles: ["[rds-sort-header]:hover,[rds-sort-header]:focus{outline:none}.rds-sort-header-container{display:flex;cursor:pointer;align-items:center;outline:0}.rds-sort-header-disabled .rds-sort-header-container{cursor:default}.rds-sort-header-content{text-align:center;display:flex;align-items:center;width:100%}.rds-sort-header-position-left{flex-direction:row-reverse}.rds-sort-header-arrow{position:relative;display:flex;opacity:0;margin:0 0 0 8px}.rds-sort-header-arrow rds-icon{height:18px;width:18px}.rds-sort-header-position-left .rds-sort-header-arrow{margin:0 8px 0 0}\n"] }]
        }], ctorParameters: () => [], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }], _disabledClass: [{
                type: HostBinding,
                args: ['class.rds-sort-header-disabled']
            }], ariaSortAttribute: [{
                type: HostBinding,
                args: ['attr.aria-sort']
            }], id: [{
                type: Input,
                args: ['rds-sort-header']
            }], arrowPosition: [{
                type: Input
            }], start: [{
                type: Input
            }], disableClear: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }], disabled: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }], onMouseenter: [{
                type: HostListener,
                args: ['mouseenter']
            }], onMouseleave: [{
                type: HostListener,
                args: ['mouseleave']
            }], _handleClick: [{
                type: HostListener,
                args: ['click']
            }], _handleKeydown: [{
                type: HostListener,
                args: ['keydown', ['$event']]
            }] } });
