// @rds/angular-components/esm2022/lib/table/table.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RdsTableComponent } from './table.component';
import { CdkTableModule } from '@angular/cdk/table';
import { RdsSortDirective } from './sort.directive';
import { RdsCellDefDirective, RdsCellDirective, RdsColumnDefDirective, RdsFooterCellDefDirective, RdsFooterCellDirective, RdsHeaderCellDefDirective, RdsHeaderCellDirective, } from './cell';
import { RdsFooterRowComponent, RdsFooterRowDefDirective, RdsHeaderRowComponent, RdsHeaderRowDefDirective, RdsNoDataRowDirective, RdsRowComponent, RdsRowDefDirective, } from './row';
import { RdsSortHeaderComponent } from './sort-header';
import { RdsTextColumnComponent } from './text-column.component';
import { RdsIconComponent } from '../icons';
import { RdsTableDraggableDirective } from './table-draggable.directive';
import { RdsTableDragHandleDirective } from './table-drag-handle.directive';
import * as i0 from "@angular/core";
const EXPORTS = [
    RdsTableComponent,
    RdsSortDirective,
    RdsSortHeaderComponent,
    RdsCellDefDirective,
    RdsHeaderCellDefDirective,
    RdsFooterCellDefDirective,
    RdsColumnDefDirective,
    RdsHeaderCellDirective,
    RdsFooterCellDirective,
    RdsCellDirective,
    RdsHeaderRowDefDirective,
    RdsFooterRowDefDirective,
    RdsRowDefDirective,
    RdsHeaderRowComponent,
    RdsFooterRowComponent,
    RdsRowComponent,
    RdsNoDataRowDirective,
    RdsTextColumnComponent,
    RdsTableDraggableDirective,
    RdsTableDragHandleDirective,
];
export class RdsTableModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTableModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsTableModule, declarations: [RdsTableComponent,
            RdsSortDirective,
            RdsSortHeaderComponent,
            RdsCellDefDirective,
            RdsHeaderCellDefDirective,
            RdsFooterCellDefDirective,
            RdsColumnDefDirective,
            RdsHeaderCellDirective,
            RdsFooterCellDirective,
            RdsCellDirective,
            RdsHeaderRowDefDirective,
            RdsFooterRowDefDirective,
            RdsRowDefDirective,
            RdsHeaderRowComponent,
            RdsFooterRowComponent,
            RdsRowComponent,
            RdsNoDataRowDirective,
            RdsTextColumnComponent,
            RdsTableDraggableDirective,
            RdsTableDragHandleDirective], imports: [CommonModule, CdkTableModule, RdsIconComponent], exports: [RdsTableComponent,
            RdsSortDirective,
            RdsSortHeaderComponent,
            RdsCellDefDirective,
            RdsHeaderCellDefDirective,
            RdsFooterCellDefDirective,
            RdsColumnDefDirective,
            RdsHeaderCellDirective,
            RdsFooterCellDirective,
            RdsCellDirective,
            RdsHeaderRowDefDirective,
            RdsFooterRowDefDirective,
            RdsRowDefDirective,
            RdsHeaderRowComponent,
            RdsFooterRowComponent,
            RdsRowComponent,
            RdsNoDataRowDirective,
            RdsTextColumnComponent,
            RdsTableDraggableDirective,
            RdsTableDragHandleDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTableModule, imports: [CommonModule, CdkTableModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTableModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [...EXPORTS],
                    exports: [...EXPORTS],
                    imports: [CommonModule, CdkTableModule, RdsIconComponent],
                }]
        }] });
