// @rds/angular-components/esm2022/lib/table/table-drag-handle.directive.mjs
import { Directive, ElementRef, HostBinding, inject } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsTableDragHandleDirective {
    constructor() {
        this.nativeElement = inject(ElementRef).nativeElement;
        this.classes = 'rds-table-drag-handle';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTableDragHandleDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsTableDragHandleDirective, selector: "[rdsTableDragHandle], [rds-table-drag-handle]", host: { properties: { "class": "this.classes" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTableDragHandleDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rdsTableDragHandle], [rds-table-drag-handle]',
                }]
        }], propDecorators: { classes: [{
                type: HostBinding,
                args: ['class']
            }] } });
