// @rds/angular-components/esm2022/lib/table/text-column.component.mjs
import { CdkTextColumn } from '@angular/cdk/table';
import { ChangeDetectionStrategy, Component, ViewEncapsulation, } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./cell";
export class RdsTextColumnComponent extends CdkTextColumn {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTextColumnComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsTextColumnComponent, selector: "rds-text-column", usesInheritance: true, ngImport: i0, template: `
    <ng-container rdsColumnDef>
      <th rds-header-cell *rdsHeaderCellDef [style.text-align]="justify">
        {{ headerText }}
      </th>
      <td rds-cell *rdsCellDef="let data" [style.text-align]="justify">
        {{ dataAccessor(data, name) }}
      </td>
    </ng-container>
  `, isInline: true, dependencies: [{ kind: "directive", type: i1.RdsCellDefDirective, selector: "[rds-cell-def], [rdsCellDef]" }, { kind: "directive", type: i1.RdsHeaderCellDefDirective, selector: "[rds-header-cell-def], [rdsHeaderCellDef]" }, { kind: "directive", type: i1.RdsColumnDefDirective, selector: "[rds-column-def], [rdsColumnDef]", inputs: ["rdsColumnDef"] }, { kind: "directive", type: i1.RdsHeaderCellDirective, selector: "rds-header-cell, th[rdsHeaderCell], th[rds-header-cell]" }, { kind: "directive", type: i1.RdsCellDirective, selector: "rds-cell, td[rdsCell], td[rds-cell]" }], changeDetection: i0.ChangeDetectionStrategy.Default, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTextColumnComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'rds-text-column',
                    template: `
    <ng-container rdsColumnDef>
      <th rds-header-cell *rdsHeaderCellDef [style.text-align]="justify">
        {{ headerText }}
      </th>
      <td rds-cell *rdsCellDef="let data" [style.text-align]="justify">
        {{ dataAccessor(data, name) }}
      </td>
    </ng-container>
  `,
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.Default,
                }]
        }] });
