// @rds/angular-components/esm2022/lib/card/card-overline.directive.mjs
import { Directive, HostBinding } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsCardOverlineDirective {
    constructor() {
        this.class = 'rds-card-overline';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCardOverlineDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsCardOverlineDirective, isStandalone: true, selector: "rds-card-overline, [rds-card-overline], [rdsCardOverline]", host: { properties: { "class": "this.class" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCardOverlineDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'rds-card-overline, [rds-card-overline], [rdsCardOverline]',
                    standalone: true,
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }] } });
