// @rds/angular-components/esm2022/lib/card/card-bottom-text.directive.mjs
import { Directive, HostBinding } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsCardBottomTextDirective {
    constructor() {
        this.class = 'rds-card-bottom-text';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCardBottomTextDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsCardBottomTextDirective, isStandalone: true, selector: "rds-card-bottom-text, [rds-card-bottom-text], [rdsCardBottomText]", host: { properties: { "class": "this.class" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCardBottomTextDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'rds-card-bottom-text, [rds-card-bottom-text], [rdsCardBottomText]',
                    standalone: true,
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }] } });
