// @rds/angular-components/esm2022/lib/card/card-label.directive.mjs
import { Directive, ElementRef, HostBinding, inject, Input, } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsCardLabelDirective {
    constructor() {
        this.nativeElement = inject(ElementRef).nativeElement;
        this.class = 'rds-card-label';
        this.color = 'blue';
    }
    getTextContent() {
        return this.nativeElement.innerText;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCardLabelDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsCardLabelDirective, isStandalone: true, selector: "rds-card-label, [rds-card-label], [rdsCardLabel]", inputs: { color: "color" }, host: { properties: { "class": "this.class" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCardLabelDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'rds-card-label, [rds-card-label], [rdsCardLabel]',
                    standalone: true,
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }], color: [{
                type: Input
            }] } });
