// @rds/angular-components/esm2022/lib/card/card-title.directive.mjs
import { Directive, HostBinding } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Title of a card, needed as it's used as a selector in the API.
 * @docs-private
 */
export class RdsCardTitleDirective {
    constructor() {
        this.class = 'rds-card-title';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCardTitleDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsCardTitleDirective, isStandalone: true, selector: "rds-card-title, [rds-card-title], [rdsCardTitle]", host: { properties: { "class": "this.class" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCardTitleDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: `rds-card-title, [rds-card-title], [rdsCardTitle]`,
                    standalone: true,
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }] } });
