// @rds/angular-components/esm2022/lib/card/card-selectable.directive.mjs
import { booleanAttribute, Directive, HostBinding, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsCardSelectableDirective {
    constructor() {
        this.class = 'rds-card-selectable';
        this.isSelected = false;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCardSelectableDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "16.1.0", version: "17.3.1", type: RdsCardSelectableDirective, isStandalone: true, selector: "rds-card-selectable, [rds-card-selectable], [rdsCardSelectable]", inputs: { isSelected: ["isSelected", "isSelected", booleanAttribute] }, host: { properties: { "class": "this.class", "class.rds-card-selected": "this.isSelected", "attr.aria-checked": "this.isSelected" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCardSelectableDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: `rds-card-selectable, [rds-card-selectable], [rdsCardSelectable]`,
                    standalone: true,
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }], isSelected: [{
                type: HostBinding,
                args: ['class.rds-card-selected']
            }, {
                type: HostBinding,
                args: ['attr.aria-checked']
            }, {
                type: Input,
                args: [{ transform: booleanAttribute }]
            }] } });
