// @rds/angular-components/esm2022/lib/card/card-image.directive.mjs
import { Directive, HostBinding } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Image used in a card, needed to add the rds- CSS styling.
 * @docs-private
 */
export class RdsCardImageDirective {
    constructor() {
        this.class = 'rds-card-image';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCardImageDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsCardImageDirective, isStandalone: true, selector: "[rds-card-image], [rdsCardImage]", host: { properties: { "class": "this.class" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCardImageDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rds-card-image], [rdsCardImage]',
                    standalone: true,
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }] } });
