// @rds/angular-components/esm2022/lib/anchor-menu/anchor-menu.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RdsAnchorMenuComponent } from './anchor-menu.component';
import { RdsAnchorMenuItemComponent } from './anchor-menu-item/anchor-menu-item.component';
import { RdsAnchorMenuSectionDirective } from './anchor-menu-section.directive';
import { RdsAnchorMenuSectionsContainerDirective } from './anchor-menu-sections-container.directive';
import { RouterLink, RouterLinkActive } from '@angular/router';
import { RdsAnchorMenuTitleDirective } from './anchor-menu-title.directive';
import * as i0 from "@angular/core";
const ELEMENTS = [
    RdsAnchorMenuComponent,
    RdsAnchorMenuItemComponent,
    RdsAnchorMenuSectionDirective,
    RdsAnchorMenuSectionsContainerDirective,
    RdsAnchorMenuTitleDirective,
];
export class RdsAnchorMenuModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAnchorMenuModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsAnchorMenuModule, declarations: [RdsAnchorMenuComponent,
            RdsAnchorMenuItemComponent,
            RdsAnchorMenuSectionDirective,
            RdsAnchorMenuSectionsContainerDirective,
            RdsAnchorMenuTitleDirective], imports: [CommonModule, RouterLink, RouterLinkActive], exports: [RdsAnchorMenuComponent,
            RdsAnchorMenuItemComponent,
            RdsAnchorMenuSectionDirective,
            RdsAnchorMenuSectionsContainerDirective,
            RdsAnchorMenuTitleDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAnchorMenuModule, imports: [CommonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAnchorMenuModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, RouterLink, RouterLinkActive],
                    declarations: [...ELEMENTS],
                    exports: [...ELEMENTS],
                }]
        }] });
