// @rds/angular-components/esm2022/lib/form-field/abstract-form-field-control.mjs
import { booleanAttribute, EventEmitter, inject } from '@angular/core';
import { FormGroupDirective, NgControl, NgForm, Validators, } from '@angular/forms';
export class RdsAbstractFormFieldControl {
    constructor() {
        this.ngControl = inject(NgControl, { optional: true, self: true });
        this.parentForm = inject(NgForm, { optional: true });
        this.parentFormGroup = inject(FormGroupDirective, {
            optional: true,
        });
        this.stateChanges = new EventEmitter();
        this._disabled = false;
        this._focused = false;
        this._errorState = false;
    }
    get focused() {
        return this._focused;
    }
    get value() {
        return this._value;
    }
    set value(newId) {
        this._value = newId;
    }
    get disabled() {
        return this._disabled;
    }
    set disabled(isDisabled) {
        this._disabled = booleanAttribute(isDisabled);
    }
    get required() {
        return (this._required ??
            this.ngControl?.control?.hasValidator(Validators.required) ??
            false);
    }
    set required(isRequired) {
        this._required = booleanAttribute(isRequired);
    }
    get id() {
        return this._id;
    }
    set id(newId) {
        this._id = newId;
    }
}
