// @rds/angular-components/esm2022/lib/form-field/error.directive.mjs
import { Directive, HostBinding, Input } from '@angular/core';
import * as i0 from "@angular/core";
let nextUniqueId = 0;
export class RdsErrorDirective {
    constructor() {
        this._id = `rds-control-error-${nextUniqueId++}`;
        this.hostRole = 'alert';
    }
    get id() {
        return this._id;
    }
    set id(newId) {
        this._id = newId;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsErrorDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsErrorDirective, selector: "rds-control-error", inputs: { id: "id" }, host: { properties: { "attr.role": "this.hostRole", "attr.id": "this.id" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsErrorDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'rds-control-error',
                }]
        }], propDecorators: { hostRole: [{
                type: HostBinding,
                args: ['attr.role']
            }], id: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['attr.id']
            }] } });
