// @rds/angular-components/esm2022/lib/form-field/prefix-separator.component.mjs
import { ChangeDetectionStrategy, Component, HostBinding } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsPrefixSeparatorComponent {
    constructor() {
        this.hostClass = 'rds-form-field__control-prefix-separator-container';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsPrefixSeparatorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsPrefixSeparatorComponent, selector: "rds-control-prefix-separator", host: { properties: { "class": "this.hostClass" } }, ngImport: i0, template: ` <div class="rds-form-field__control-prefix-separator"></div>`, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsPrefixSeparatorComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'rds-control-prefix-separator',
                    template: ` <div class="rds-form-field__control-prefix-separator"></div>`,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], propDecorators: { hostClass: [{
                type: HostBinding,
                args: ['class']
            }] } });
