// @rds/angular-components/esm2022/lib/form-field/textarea-resize.directive.mjs
import { Directive, HostBinding, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsTextareaResizeDirective {
    constructor() {
        this.resize = 'vertical';
    }
    get class() {
        return `rds-textarea-resize-${this.resize || 'vertical'}`;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTextareaResizeDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsTextareaResizeDirective, selector: "textarea[resize]", inputs: { resize: "resize" }, host: { properties: { "class": "this.class" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTextareaResizeDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: `textarea[resize]`,
                }]
        }], propDecorators: { resize: [{
                type: Input
            }], class: [{
                type: HostBinding,
                args: ['class']
            }] } });
