// @rds/angular-components/esm2022/lib/menu/menu.module.mjs
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { CdkScrollableModule } from '@angular/cdk/scrolling';
import { RdsMenuComponent } from './menu.component';
import { RdsMenuContentDirective } from './menu-content.directive';
import { RdsMenuItemComponent } from './menu-item.component';
import { BidiModule } from '@angular/cdk/bidi';
import { RDS_MENU_SCROLL_STRATEGY_FACTORY_PROVIDER, RdsMenuTriggerDirective, } from './menu-trigger.directive';
import { RdsIconComponent } from '../icons';
import { RdsMenuSeparatorDirective } from './menu-separator.directive';
import { RdsMenuHeaderDirective } from './menu-header.directive';
import { RdsListItemModule } from '../list-item';
import * as i0 from "@angular/core";
const ELEMENTS = [
    RdsMenuComponent,
    RdsMenuItemComponent,
    RdsMenuTriggerDirective,
    RdsMenuContentDirective,
    RdsMenuHeaderDirective,
    RdsMenuSeparatorDirective,
];
export class RdsMenuModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsMenuModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsMenuModule, declarations: [RdsMenuComponent,
            RdsMenuItemComponent,
            RdsMenuTriggerDirective,
            RdsMenuContentDirective,
            RdsMenuHeaderDirective,
            RdsMenuSeparatorDirective], imports: [CommonModule,
            BidiModule,
            OverlayModule,
            RdsIconComponent,
            RdsListItemModule], exports: [CdkScrollableModule, RdsMenuComponent,
            RdsMenuItemComponent,
            RdsMenuTriggerDirective,
            RdsMenuContentDirective,
            RdsMenuHeaderDirective,
            RdsMenuSeparatorDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsMenuModule, providers: [RDS_MENU_SCROLL_STRATEGY_FACTORY_PROVIDER], imports: [CommonModule,
            BidiModule,
            OverlayModule,
            RdsListItemModule, CdkScrollableModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsMenuModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        BidiModule,
                        OverlayModule,
                        RdsIconComponent,
                        RdsListItemModule,
                    ],
                    declarations: [...ELEMENTS],
                    exports: [CdkScrollableModule, ...ELEMENTS],
                    providers: [RDS_MENU_SCROLL_STRATEGY_FACTORY_PROVIDER],
                }]
        }] });
