// @rds/angular-components/esm2022/lib/menu/menu-content.directive.mjs
import { DomPortalOutlet, TemplatePortal } from '@angular/cdk/portal';
import { DOCUMENT } from '@angular/common';
import { ApplicationRef, ChangeDetectorRef, ComponentFactoryResolver, Directive, inject, InjectionToken, Injector, TemplateRef, ViewContainerRef, } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
export const RDS_MENU_CONTENT = new InjectionToken('RDS_MENU_CONTENT');
export class RdsMenuContentDirective {
    constructor() {
        this.template = inject(TemplateRef);
        this.componentFactoryResolver = inject(ComponentFactoryResolver);
        this.appRef = inject(ApplicationRef);
        this.injector = inject(Injector);
        this.viewContainerRef = inject(ViewContainerRef);
        this.document = inject(DOCUMENT);
        this.cdr = inject(ChangeDetectorRef, { optional: true });
        this._attached = new Subject();
    }
    attach(context = {}) {
        if (!this._portal) {
            this._portal = new TemplatePortal(this.template, this.viewContainerRef);
        }
        this.detach();
        if (!this._outlet) {
            this._outlet = new DomPortalOutlet(this.document.createElement('div'), this.componentFactoryResolver, this.appRef, this.injector);
        }
        const element = this.template.elementRef.nativeElement;
        if (element.parentNode) {
            element.parentNode.insertBefore(this._outlet.outletElement, element);
        }
        if (this.cdr) {
            this.cdr.markForCheck();
        }
        this._portal.attach(this._outlet, context);
        this._attached.next();
    }
    detach() {
        if (this._portal.isAttached) {
            this._portal.detach();
        }
    }
    ngOnDestroy() {
        if (this._outlet) {
            this._outlet.dispose();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsMenuContentDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsMenuContentDirective, selector: "ng-template[rds-menu-content], ng-template[rdsMenuContent]", providers: [
            { provide: RDS_MENU_CONTENT, useExisting: RdsMenuContentDirective },
        ], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsMenuContentDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ng-template[rds-menu-content], ng-template[rdsMenuContent]',
                    providers: [
                        { provide: RDS_MENU_CONTENT, useExisting: RdsMenuContentDirective },
                    ],
                }]
        }] });
