// @rds/angular-components/esm2022/lib/menu/menu-item.component.mjs
import { FocusMonitor } from '@angular/cdk/a11y';
import { ChangeDetectionStrategy, Component, HostBinding, HostListener, inject, ViewEncapsulation, } from '@angular/core';
import { Subject } from 'rxjs';
import { RdsNestableListItemComponent, } from '../list-item';
import { RDS_NESTABLE_LIST_ITEM_TEMPLATE } from '../list-item';
import { RDS_MENU_PANEL } from './menu-panel';
import * as i0 from "@angular/core";
import * as i1 from "../icons/icon.component";
export class RdsMenuItemComponent extends RdsNestableListItemComponent {
    constructor() {
        super(...arguments);
        this._parentMenu = inject(RDS_MENU_PANEL, { optional: true });
        this.focusMonitor = inject(FocusMonitor);
        this.rdsMenuItemClass = true;
        this._hovered = new Subject();
    }
    ngAfterViewInit() {
        this.focusMonitor.monitor(this.elementRef, false);
    }
    ngOnDestroy() {
        this._hovered.complete();
        this._focused.complete();
        this.focusMonitor.stopMonitoring(this.elementRef);
    }
    _getHostElement() {
        return this.elementRef.nativeElement;
    }
    handleMouseEnter() {
        this._hovered.next(this);
    }
    focus(origin, options) {
        if (this.focusMonitor && origin) {
            this.focusMonitor.focusVia(this.elementRef.nativeElement, origin, options);
        }
        else {
            super.focus();
        }
        this._focused.next(this);
    }
    blur() {
        this.markAsInactive();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsMenuItemComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "17.3.1", type: RdsMenuItemComponent, selector: "[rds-menu-item], [rdsMenuItem]", host: { listeners: { "mouseenter": "handleMouseEnter()", "blur": "blur()" }, properties: { "class.rds-menu-item": "this.rdsMenuItemClass" } }, exportAs: ["rdsMenuItem"], usesInheritance: true, ngImport: i0, template: "<ng-content select=\"rds-icon\"></ng-content>\n<span class=\"rds-list-item-text\">\n  <ng-content />\n</span>\n\n@if (hasChildItems) {\n  <rds-icon\n    namespace=\"outlined\"\n    icon=\"chevron_right\"\n    class=\"rds-list-item__has-child-items\"\n  />\n}", isInline: true, dependencies: [{ kind: "component", type: i1.RdsIconComponent, selector: "rds-icon", inputs: ["namespace", "icon"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsMenuItemComponent, decorators: [{
            type: Component,
            args: [{
                    selector: '[rds-menu-item], [rdsMenuItem]',
                    exportAs: 'rdsMenuItem',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    template: RDS_NESTABLE_LIST_ITEM_TEMPLATE,
                }]
        }], propDecorators: { rdsMenuItemClass: [{
                type: HostBinding,
                args: ['class.rds-menu-item']
            }], handleMouseEnter: [{
                type: HostListener,
                args: ['mouseenter']
            }], blur: [{
                type: HostListener,
                args: ['blur']
            }] } });
