// @rds/angular-components/esm2022/lib/slider/slider.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RdsSliderComponent } from './slider.component';
import * as i0 from "@angular/core";
export class RdsSliderModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSliderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsSliderModule, declarations: [RdsSliderComponent], imports: [CommonModule], exports: [RdsSliderComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSliderModule, imports: [CommonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSliderModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    exports: [RdsSliderComponent],
                    declarations: [RdsSliderComponent],
                }]
        }] });
