// @rds/angular-components/esm2022/lib/breadcrumbs/breadcrumb-item.component.mjs
import { FocusMonitor } from '@angular/cdk/a11y';
import { Attribute, ChangeDetectionStrategy, Component, ElementRef, EventEmitter, HostBinding, HostListener, inject, Input, numberAttribute, Output, } from '@angular/core';
import { TAB_INDEX_DEFAULT } from '../utils';
import * as i0 from "@angular/core";
export class RdsBreadcrumbItemComponent {
    constructor(tabIndex) {
        this.focusMonitor = inject(FocusMonitor);
        this.elementRef = inject(ElementRef);
        this.class = 'rds-breadcrumb-item rds-focus-indicator';
        this.tabIndex = TAB_INDEX_DEFAULT;
        this.itemClick = new EventEmitter();
        this.tabIndex = numberAttribute(tabIndex, TAB_INDEX_DEFAULT);
    }
    ngAfterViewInit() {
        this.focusMonitor.monitor(this.elementRef, false);
    }
    ngOnDestroy() {
        this.focusMonitor.stopMonitoring(this.elementRef);
    }
    onClickItem() {
        this.itemClick.emit(this.item);
    }
    focus(origin = 'program', options) {
        this.focusMonitor.focusVia(this.elementRef.nativeElement, origin, options);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsBreadcrumbItemComponent, deps: [{ token: 'tabindex', attribute: true }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "16.1.0", version: "17.3.1", type: RdsBreadcrumbItemComponent, isStandalone: true, selector: "rds-breadcrumb-item", inputs: { tabIndex: ["tabIndex", "tabIndex", (v) => numberAttribute(v, TAB_INDEX_DEFAULT)], item: "item" }, outputs: { itemClick: "itemClick" }, host: { listeners: { "click": "onClickItem()", "keydown.enter": "onClickItem()", "keydown.space": "onClickItem()" }, properties: { "class": "this.class", "tabIndex": "this.tabIndex" } }, ngImport: i0, template: `<span [title]="item.title ?? ''">{{ item.text }}</span>`, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsBreadcrumbItemComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'rds-breadcrumb-item',
                    template: `<span [title]="item.title ?? ''">{{ item.text }}</span>`,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    standalone: true,
                }]
        }], ctorParameters: () => [{ type: undefined, decorators: [{
                    type: Attribute,
                    args: ['tabindex']
                }] }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }], tabIndex: [{
                type: HostBinding,
                args: ['tabIndex']
            }, {
                type: Input,
                args: [{ transform: (v) => numberAttribute(v, TAB_INDEX_DEFAULT) }]
            }], item: [{
                type: Input
            }], itemClick: [{
                type: Output
            }], onClickItem: [{
                type: HostListener,
                args: ['click']
            }, {
                type: HostListener,
                args: ['keydown.enter']
            }, {
                type: HostListener,
                args: ['keydown.space']
            }] } });
