// @rds/angular-components/esm2022/lib/breadcrumbs/breadcrumbs.module.mjs
import { NgModule } from '@angular/core';
import { RdsBreadcrumbItemComponent } from './breadcrumb-item.component';
import { RdsBreadcrumbsComponent } from './breadcrumbs.component';
import * as i0 from "@angular/core";
export const RDS_BREADCRUMBS_DIRECTIVES = [
    RdsBreadcrumbsComponent,
    RdsBreadcrumbItemComponent,
];
export class RdsBreadcrumbsModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsBreadcrumbsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsBreadcrumbsModule, imports: [RdsBreadcrumbsComponent,
            RdsBreadcrumbItemComponent], exports: [RdsBreadcrumbsComponent,
            RdsBreadcrumbItemComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsBreadcrumbsModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsBreadcrumbsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...RDS_BREADCRUMBS_DIRECTIVES],
                    exports: [...RDS_BREADCRUMBS_DIRECTIVES],
                }]
        }] });
