// @rds/angular-components/esm2022/lib/badge/badge.module.mjs
import { NgModule } from '@angular/core';
import { RdsBadgeComponent } from './badge.component';
import { RdsBadgeDirective } from './badge.directive';
import * as i0 from "@angular/core";
export const RDS_BADGE_DIRECTIVES = [
    RdsBadgeComponent,
    RdsBadgeDirective,
];
export class RdsBadgeModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsBadgeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsBadgeModule, imports: [RdsBadgeComponent,
            RdsBadgeDirective], exports: [RdsBadgeComponent,
            RdsBadgeDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsBadgeModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsBadgeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...RDS_BADGE_DIRECTIVES],
                    exports: [...RDS_BADGE_DIRECTIVES],
                }]
        }] });
