// @rds/angular-components/esm2022/lib/upload/file-item.class.mjs
import { BehaviorSubject } from 'rxjs';
let id = 0;
export class RdsFileItem {
    get errorIntlToken() {
        return this._errorIntlToken;
    }
    get state() {
        return this._state$.value;
    }
    set state(v) {
        if (v !== this.state) {
            this._state$.next(v);
        }
    }
    get isUploading() {
        return this.state === 'uploading';
    }
    constructor(file, failedFilters) {
        this.file = file;
        this.failedFilters = failedFilters;
        this._errorIntlToken = 'EMPTY';
        this._progress$ = new BehaviorSubject(0);
        this._state$ = new BehaviorSubject('ready');
        this.progress$ = this._progress$.asObservable();
        this.state$ = this._state$.asObservable();
        this.id = id++;
        this.hasFailedFilters = !!this.failedFilters.length;
        if (this.hasFailedFilters) {
            this._errorIntlToken = this.failedFilters[0].descriptionIntlToken;
            this.state = 'error';
        }
    }
    onBeforeUpload() {
        this._errorIntlToken = 'EMPTY';
        this.state = 'uploading';
        this._progress$.next(0);
    }
    onProgress(progress) {
        this._progress$.next(progress);
    }
    onSuccess() {
        this.state = 'success';
        this._progress$.next(100);
    }
    onError() {
        this._errorIntlToken = 'uploadFailed';
        this.state = 'error';
    }
    onAbort() {
        this.state = 'aborted';
    }
}
