// @rds/angular-components/esm2022/lib/upload/http-client-wrapper.mjs
import { HttpEventType, HttpRequest, } from '@angular/common/http';
export class RdsHttpClientWrapper {
    constructor(item, config, client) {
        this.item = item;
        this.config = config;
        this.client = client;
        this.form = new FormData();
        this.form.append('file', item.file, item.file.name);
        const method = config.method ?? 'POST';
        this.request = new HttpRequest(method, config.url, this.form, {
            headers: config.headers,
            reportProgress: true,
            withCredentials: !!config.authToken,
            responseType: config.responseType,
        });
        if (config.authToken) {
            this.request = this.request.clone({
                headers: this.request.headers.set('Authorization', config.authToken),
            });
        }
        this.response$ = this.client.request(this.request);
    }
    abort() {
        this.responseSubscription?.unsubscribe();
        this.responseSubscription = undefined;
        this.onAbort();
    }
    send() {
        this.responseSubscription = this.response$.subscribe({
            error: (e) => this.onError(e),
            next: (e) => {
                if (e.type === HttpEventType.UploadProgress) {
                    this.onProgress(e);
                }
                else if (e.type === HttpEventType.Response) {
                    this.onSuccess(e);
                }
            },
        });
    }
    onSuccess(event) {
        const change = {
            item: this.item,
            event,
        };
        this.success && this.success(change);
        this.complete && this.complete(change);
    }
    onError(event) {
        const change = {
            item: this.item,
            event,
        };
        this.error && this.error(change);
        this.complete && this.complete(change);
    }
    onProgress(event) {
        const progress = event.total != undefined
            ? Math.round((event.loaded * 100) / event.total)
            : null;
        this.progress && this.progress({ item: this.item, progress });
    }
    onAbort() {
        const change = {
            item: this.item,
        };
        this.cancel && this.cancel(change);
        this.complete && this.complete(change);
    }
}
