// @rds/angular-components/esm2022/lib/upload/rds-upload-intl.service.mjs
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { FILTER_FILE_SIZE_DESCRIPTION_INTL_TOKEN, FILTER_FILE_TYPE_DESCRIPTION_INTL_TOKEN, FILTER_QUEUE_LIMIT_DESCRIPTION_INTL_TOKEN, } from './upload.filters';
import * as i0 from "@angular/core";
export class RdsUploadIntlService {
    constructor() {
        this._change$ = new BehaviorSubject(RdsUploadIntlService.translations);
        this.change$ = this._change$.asObservable();
        this.fileItemUploadSuccessMessageFactory = (fileItem) => {
            return fileItem.file.name + ' upload completed successfully';
        };
        this.fileItemUploadErrorMessageFactory = (fileItem) => {
            return (fileItem.file.name +
                ' error: ' +
                RdsUploadIntlService.translations[fileItem.errorIntlToken]);
        };
        this.fileItemFilterFailedMessageFactory = (fileItem) => {
            return (fileItem.file.name +
                ' error: ' +
                RdsUploadIntlService.translations[fileItem.errorIntlToken]);
        };
        this.fileItemProgressMessageFactory = (progress) => {
            return progress == null ? '' : `${progress}% Uploaded`;
        };
    }
    static { this.translations = {
        EMPTY: '',
        chooseFilesButtonLabel: 'Upload',
        dropZoneChooseFiles: 'Upload',
        dropZoneOrDropFiles: 'or drop files here',
        uploadListAriaLabel: 'File upload queue',
        startUploadButtonLabel: 'Upload',
        retryButtonAriaLabel: 'Retry',
        cancelButtonAriaLabel: 'Cancel',
        uploadFailed: 'Upload failed.',
        [FILTER_QUEUE_LIMIT_DESCRIPTION_INTL_TOKEN]: '',
        [FILTER_FILE_SIZE_DESCRIPTION_INTL_TOKEN]: 'File exceeds size limit.',
        [FILTER_FILE_TYPE_DESCRIPTION_INTL_TOKEN]: 'Invalid file type.',
    }; }
    setTranslation(key, value) {
        RdsUploadIntlService.translations[key] = value;
        this._change$.next({ ...RdsUploadIntlService.translations });
    }
    setFileItemUploadSuccessMessageFactory(fn) {
        this.fileItemUploadSuccessMessageFactory = fn;
    }
    setFileItemUploadErrorMessageFactory(fn) {
        this.fileItemUploadErrorMessageFactory = fn;
    }
    setFileItemFilterFailedMessageFactory(fn) {
        this.fileItemFilterFailedMessageFactory = fn;
    }
    setFileItemProgressMessageFactory(fn) {
        this.fileItemProgressMessageFactory = fn;
    }
    getFileItemUploadSuccessMessage(fileItem) {
        return this.fileItemUploadSuccessMessageFactory(fileItem);
    }
    getFileItemUploadErrorMessage(fileItem) {
        return this.fileItemUploadErrorMessageFactory(fileItem);
    }
    getFileItemFilterFailedMessage(fileItem) {
        return this.fileItemFilterFailedMessageFactory(fileItem);
    }
    getFileItemProgressMessage(progress) {
        return this.fileItemProgressMessageFactory(progress);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsUploadIntlService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsUploadIntlService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsUploadIntlService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
