// @rds/angular-components/esm2022/lib/upload/upload-drop-zone.directive.mjs
import { booleanAttribute, Directive, ElementRef, EventEmitter, HostListener, inject, Input, Output, Renderer2, } from '@angular/core';
import * as i0 from "@angular/core";
const RDS_DROPZONE_DRAGOVER_CLASS_TOKEN = 'rds-upload-drop-zone-dragover';
const RDS_DROPZONE_DISABLED_CLASS_TOKEN = 'rds-upload-drop-zone-disabled';
export class RdsUploadDropZoneDirective {
    constructor() {
        this.nativeElement = inject(ElementRef).nativeElement;
        this.renderer = inject(Renderer2);
        this._disabled = false;
        this.fileOver = new EventEmitter();
        this.fileDropped = new EventEmitter();
    }
    set disabled(v) {
        this._disabled = booleanAttribute(v);
        this.onDisabledChange();
    }
    get disabled() {
        return this._disabled;
    }
    onDragenter(e) {
        if (!this.hasFiles(e)) {
            return;
        }
        this.fileOver.emit(true);
        this.toggleDragoverClass(true);
    }
    onDragover(e) {
        if (!this.hasFiles(e)) {
            return;
        }
        e.dataTransfer.dropEffect = this.disabled ? 'none' : 'copy';
        // without this 'drop' event won't be fired in chrome
        e.preventDefault();
    }
    onDragleave() {
        this.fileOver.emit(false);
        this.toggleDragoverClass(false);
    }
    onDrop(e) {
        if (!this.hasFiles(e) || this.disabled) {
            return;
        }
        e.preventDefault();
        this.fileOver.emit(false);
        this.toggleDragoverClass(false);
        this.fileDropped.emit(Array.from(e.dataTransfer.files));
    }
    hasFiles(event) {
        return !!event.dataTransfer?.types?.includes('Files');
    }
    toggleDragoverClass(hasDragover) {
        if (hasDragover) {
            this.renderer.addClass(this.nativeElement, RDS_DROPZONE_DRAGOVER_CLASS_TOKEN);
        }
        else {
            this.renderer.removeClass(this.nativeElement, RDS_DROPZONE_DRAGOVER_CLASS_TOKEN);
        }
    }
    onDisabledChange() {
        if (this.disabled) {
            this.renderer.setAttribute(this.nativeElement, 'disabled', '');
            this.renderer.addClass(this.nativeElement, RDS_DROPZONE_DISABLED_CLASS_TOKEN);
        }
        else {
            this.renderer.removeAttribute(this.nativeElement, 'disabled');
            this.renderer.removeClass(this.nativeElement, RDS_DROPZONE_DISABLED_CLASS_TOKEN);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsUploadDropZoneDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsUploadDropZoneDirective, selector: "[rds-upload-drop-zone], [rdsUploadDropZone]", inputs: { disabled: "disabled" }, outputs: { fileOver: "fileOver", fileDropped: "fileDropped" }, host: { listeners: { "dragenter": "onDragenter($event)", "dragover": "onDragover($event)", "dragleave": "onDragleave()", "drop": "onDrop($event)" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsUploadDropZoneDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rds-upload-drop-zone], [rdsUploadDropZone]',
                }]
        }], propDecorators: { disabled: [{
                type: Input
            }], fileOver: [{
                type: Output
            }], fileDropped: [{
                type: Output
            }], onDragenter: [{
                type: HostListener,
                args: ['dragenter', ['$event']]
            }], onDragover: [{
                type: HostListener,
                args: ['dragover', ['$event']]
            }], onDragleave: [{
                type: HostListener,
                args: ['dragleave']
            }], onDrop: [{
                type: HostListener,
                args: ['drop', ['$event']]
            }] } });
