// @rds/angular-components/esm2022/lib/upload/upload.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RdsIconComponent } from '../icons';
import { RdsButtonModule } from '../button';
import { RdsUploadComponent } from './upload.component';
import { RdsUploadDirective } from './upload.directive';
import { RdsUploadDropZoneDirective } from './upload-drop-zone.directive';
import { RdsProgressBarModule } from '../progress-bar';
import { RdsFileSizePipe } from './file-size.pipe';
import { RdsUploadListItemComponent } from './upload-list-item/upload-list-item.component';
import { RdsUploadErrorDirective } from './upload-error.directive';
import * as i0 from "@angular/core";
const publicElements = [
    RdsUploadComponent,
    RdsUploadDirective,
    RdsUploadDropZoneDirective,
    RdsFileSizePipe,
    RdsUploadListItemComponent,
    RdsUploadErrorDirective,
];
export class RdsUploadModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsUploadModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsUploadModule, declarations: [RdsUploadComponent,
            RdsUploadDirective,
            RdsUploadDropZoneDirective,
            RdsFileSizePipe,
            RdsUploadListItemComponent,
            RdsUploadErrorDirective], imports: [CommonModule,
            RdsIconComponent,
            RdsButtonModule,
            RdsProgressBarModule], exports: [RdsUploadComponent,
            RdsUploadDirective,
            RdsUploadDropZoneDirective,
            RdsFileSizePipe,
            RdsUploadListItemComponent,
            RdsUploadErrorDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsUploadModule, imports: [CommonModule,
            RdsButtonModule,
            RdsProgressBarModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsUploadModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...publicElements],
                    declarations: [...publicElements],
                    imports: [
                        CommonModule,
                        RdsIconComponent,
                        RdsButtonModule,
                        RdsProgressBarModule,
                    ],
                }]
        }] });
