// @rds/angular-components/esm2022/lib/upload/upload-error.directive.mjs
import { Directive, ElementRef, inject, Renderer2 } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsUploadErrorDirective {
    constructor() {
        this.nativeElement = inject((ElementRef))
            .nativeElement;
        this.renderer = inject(Renderer2);
    }
    setId(id) {
        this.renderer.setAttribute(this.nativeElement, 'id', id);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsUploadErrorDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsUploadErrorDirective, selector: "rds-upload-error, [rds-upload-error], [rdsUploadError]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsUploadErrorDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: `rds-upload-error, [rds-upload-error], [rdsUploadError]`,
                }]
        }] });
