// @rds/angular-components/esm2022/lib/upload/upload.directive.mjs
import { Directive, ElementRef, EventEmitter, HostBinding, HostListener, inject, Input, Output, Renderer2, } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsUploadDirective {
    constructor() {
        this.renderer = inject(Renderer2);
        this.nativeElement = inject(ElementRef).nativeElement;
        this.fileSelected = new EventEmitter();
        this.display = 'none';
        this.type = 'file';
    }
    set config(v) {
        this._config = v;
        this.onConfigChange();
    }
    onChange() {
        const files = Array.from(this.nativeElement.files ?? []);
        this.fileSelected.emit(files);
        this.nativeElement.value = '';
    }
    ngOnInit() {
        if (ngDevMode && !this._config) {
            throw new Error(`The "config" input is required, received "${JSON.stringify(this._config)}"`);
        }
        this.onConfigChange();
    }
    onConfigChange() {
        this._config.multiple
            ? this.renderer.setAttribute(this.nativeElement, 'multiple', '')
            : this.renderer.removeAttribute(this.nativeElement, 'multiple');
        const accept = this._config.accept?.join(',');
        accept
            ? this.renderer.setAttribute(this.nativeElement, 'accept', accept)
            : this.renderer.removeAttribute(this.nativeElement, 'accept');
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsUploadDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsUploadDirective, selector: "input[rds-upload], input[rdsUpload]", inputs: { config: "config" }, outputs: { fileSelected: "fileSelected" }, host: { listeners: { "change": "onChange()" }, properties: { "style.display": "this.display", "type": "this.type" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsUploadDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'input[rds-upload], input[rdsUpload]',
                }]
        }], propDecorators: { config: [{
                type: Input
            }], fileSelected: [{
                type: Output
            }], display: [{
                type: HostBinding,
                args: ['style.display']
            }], type: [{
                type: HostBinding,
                args: ['type']
            }], onChange: [{
                type: HostListener,
                args: ['change']
            }] } });
