// @rds/angular-components/esm2022/lib/upload/upload.filters.mjs
// those consts exist to avoid circular reference in type definition
export const FILTER_QUEUE_LIMIT_DESCRIPTION_INTL_TOKEN = 'queueLimitFilterDescription';
export const FILTER_FILE_SIZE_DESCRIPTION_INTL_TOKEN = 'fileSizeFilterDescription';
export const FILTER_FILE_TYPE_DESCRIPTION_INTL_TOKEN = 'fileTypeFilterDescription';
export class RdsUploadFilters {
    static { this.queueLimit = {
        name: 'queueLimit',
        descriptionIntlToken: FILTER_QUEUE_LIMIT_DESCRIPTION_INTL_TOKEN,
        preventsAddingToQueue: true,
        fn: (item, config, queue) => !config.queueLimit || queue.length < config.queueLimit,
    }; }
    static { this.fileSize = {
        name: 'fileSize',
        descriptionIntlToken: FILTER_FILE_SIZE_DESCRIPTION_INTL_TOKEN,
        fn: (item, config) => !(config.maxFileSize && item.size > config.maxFileSize),
    }; }
    static { this.fileType = {
        name: 'fileType',
        descriptionIntlToken: FILTER_FILE_TYPE_DESCRIPTION_INTL_TOKEN,
        fn: verifyAccept,
    }; }
    static compose(currentFilters, config) {
        const filters = new Set([...currentFilters, RdsUploadFilters.queueLimit]);
        config.maxFileSize && filters.add(RdsUploadFilters.fileSize);
        config.accept && filters.add(RdsUploadFilters.fileType);
        return Array.from(filters);
    }
}
function verifyAccept(file, config) {
    const accept = config.accept?.filter((v) => !!v);
    if (!accept || !accept.length) {
        return true;
    }
    const allowed = parseAccept(accept);
    const extension = extractExtension(file.name);
    return (allowed.extension.includes(extension) ||
        allowed.mime.some((regex) => regex.test(file.type)));
}
function parseAccept(accept) {
    return accept.reduce((acc, curr) => {
        curr = curr.trim().toLowerCase();
        if (curr.startsWith('.')) {
            acc.extension.push(...unifyJpgExtensionsIfNeeded(curr));
        }
        else {
            acc.mime.push(new RegExp(curr.replace('*', '.*')));
        }
        return acc;
    }, { extension: [], mime: [] });
}
function unifyJpgExtensionsIfNeeded(extension) {
    if (extension === '.jpg' || extension === '.jpeg') {
        return ['.jpeg', '.jpg'];
    }
    else {
        return [extension];
    }
}
function extractExtension(filename) {
    const splitName = filename.split('.');
    return '.' + splitName[splitName.length - 1].toLowerCase();
}
