// @rds/angular-components/esm2022/lib/form-field/select/multi-select-option.component.mjs
import { Component, HostBinding, Input } from '@angular/core';
import { Subject } from 'rxjs';
import { RdsListItemComponent, RdsMultiSelectListItemComponent, } from '../../list-item';
import { RDS_MULTI_SELECT_LIST_ITEM_TEMPLATE } from '../../list-item';
import { AbstractFocusableOption } from './abstract-focusable-option';
import { RdsMultiSelectOptionSelectionChangedEvent } from './select-option-selection-changed.event';
import * as i0 from "@angular/core";
import * as i1 from "../../checkbox/pseudo-checkbox.component";
let nextId = 0;
export class RdsMultiSelectOptionComponent extends RdsMultiSelectListItemComponent {
    constructor() {
        super(...arguments);
        this._currentViewValue = '';
        this.hostClass = 'rds-select__option';
        this.id = `rds-option-${nextId++}`;
        this.selectionChange$ = new Subject();
    }
    get viewValue() {
        return (this.elementRef.nativeElement.textContent || '').trim();
    }
    ngAfterViewChecked() {
        if (this.checked) {
            const viewValue = this.viewValue;
            if (viewValue !== this._currentViewValue) {
                this._currentViewValue = viewValue;
            }
        }
    }
    rect() {
        return this.elementRef.nativeElement.getBoundingClientRect();
    }
    isVisible() {
        return !this.elementRef.nativeElement.classList.contains('hide');
    }
    toggleVisibility(value) {
        if (value === true) {
            this.show();
        }
        else if (value === false) {
            this.hide();
        }
        else {
            if (this.isVisible()) {
                this.hide();
            }
            else if (!this.isVisible()) {
                this.show();
            }
        }
    }
    hide() {
        this.renderer.addClass(this.elementRef.nativeElement, 'hide');
    }
    show() {
        this.renderer.removeClass(this.elementRef.nativeElement, 'hide');
    }
    toggle() {
        super.toggle();
        this.selectionChange$.next(new RdsMultiSelectOptionSelectionChangedEvent(this, true));
    }
    handleClick() {
        this.toggle();
    }
    setActiveStyles() {
        this.focus('keyboard');
    }
    setInactiveStyles() {
        if (this._focused) {
            this.blur();
            this.cdr.markForCheck();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsMultiSelectOptionComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "17.3.1", type: RdsMultiSelectOptionComponent, selector: "rds-multi-select-option", inputs: { id: "id", value: "value" }, host: { properties: { "class": "this.hostClass", "attr.id": "this.id" } }, providers: [
            {
                provide: RdsListItemComponent,
                useExisting: RdsMultiSelectOptionComponent,
            },
            {
                provide: AbstractFocusableOption,
                useExisting: RdsMultiSelectOptionComponent,
            },
        ], usesInheritance: true, ngImport: i0, template: "<ng-content select=\"rds-avatar\"/>\n<ng-content select=\"rds-icon\"/>\n@if (!avatar && !icon) {\n    <rds-pseudo-checkbox\n      noSpacing\n      [checked]=\"checked\"\n      [disabled]=\"disabled\"\n    />\n}\n<span class=\"rds-list-item-text\">\n  <ng-content/>\n</span>\n@if (avatar || icon) {\n    <rds-pseudo-checkbox\n      noSpacing\n      [checked]=\"checked\"\n      [disabled]=\"disabled\"\n      class=\"rds-list-item__checkbox-right\"\n    />\n}\n", isInline: true, dependencies: [{ kind: "component", type: i1.RdsPseudoCheckboxComponent, selector: "rds-pseudo-checkbox", inputs: ["checked", "disabled", "indeterminate", "noSpacing"] }] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsMultiSelectOptionComponent, decorators: [{
            type: Component,
            args: [{
                    template: RDS_MULTI_SELECT_LIST_ITEM_TEMPLATE,
                    selector: 'rds-multi-select-option',
                    providers: [
                        {
                            provide: RdsListItemComponent,
                            useExisting: RdsMultiSelectOptionComponent,
                        },
                        {
                            provide: AbstractFocusableOption,
                            useExisting: RdsMultiSelectOptionComponent,
                        },
                    ],
                }]
        }], propDecorators: { hostClass: [{
                type: HostBinding,
                args: ['class']
            }], id: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['attr.id']
            }], value: [{
                type: Input
            }] } });
