// @rds/angular-components/esm2022/lib/form-field/select/multi-select.component.mjs
import { A, DOWN_ARROW, ENTER, hasModifierKey, LEFT_ARROW, RIGHT_ARROW, SPACE, UP_ARROW, } from '@angular/cdk/keycodes';
import { booleanAttribute, ChangeDetectionStrategy, Component, ContentChildren, inject, Input, QueryList, ViewChildren, ViewEncapsulation, } from '@angular/core';
import { merge } from 'rxjs';
import { debounceTime, filter, map, startWith, switchMap, takeUntil, tap, } from 'rxjs/operators';
import { RdsAbstractFormFieldControl } from '../abstract-form-field-control';
import { SelectAnimations } from './select.animations';
import { RdsAbstractSelectComponent } from './abstract-select.component';
import { SelectionModel } from '@angular/cdk/collections';
import { RdsSelectAllOptionComponent } from './select-all-option.component';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { RdsChipComponent } from '../../chips';
import { FocusKeyManager } from '@angular/cdk/a11y';
import { Directionality } from '@angular/cdk/bidi';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/overlay";
import * as i2 from "../../icons/icon.component";
import * as i3 from "../../button/icon-button.component";
import * as i4 from "../../chips/chip-list/chip-list.component";
import * as i5 from "../../chips/chip/chip.component";
import * as i6 from "../../badge/badge.component";
import * as i7 from "./trigger-values.component";
import * as i8 from "@angular/common";
export class RdsMultiSelectComponent extends RdsAbstractSelectComponent {
    constructor() {
        super(...arguments);
        this.dir = inject(Directionality, { optional: true });
        this._name = 'rds-multi-select';
        this.allOptionsSelected = false;
        this._withChips = false;
        this.focusedChip = null;
        this.chipActions = new QueryList();
    }
    get selected() {
        return this.selectionModel.selected;
    }
    get triggerValues() {
        if (this.empty) {
            return [''];
        }
        return this.selectionModel.selected.map((option) => option.viewValue);
    }
    get controlType() {
        if (this.withChips) {
            return 'select-with-chips';
        }
        else {
            return 'select';
        }
    }
    set withChips(value) {
        const newValue = booleanAttribute(value);
        if (this._withChips !== newValue) {
            this._withChips = newValue;
            this.stateChanges.emit();
        }
    }
    get withChips() {
        return this._withChips;
    }
    ngOnInit() {
        super.ngOnInit();
        this.selectionModel = new SelectionModel(true);
    }
    ngAfterContentInit() {
        super.ngAfterContentInit();
        this.selectionModel.changed
            .pipe(takeUntilDestroyed(this.destroyRef))
            .subscribe(() => {
            // check it
            this.calculateAllOptionsSelected();
            this.refreshSelectAllCheck();
            this.cdr.markForCheck();
        });
        this.calculateAllOptionsSelected();
        this.selectAllElements.changes
            .pipe(startWith(this.selectAllElements), tap(() => this.selectAllDestroy$.next()), filter(() => this.selectAllElements.length > 0), switchMap(() => merge(...this.selectAllElements.map((all) => all.selectAllChanged)).pipe(takeUntil(this.selectAllDestroy$), takeUntilDestroyed(this.destroyRef))), takeUntilDestroyed(this.destroyRef))
            .subscribe(() => {
            this.toggleAllOptions();
        });
    }
    ngAfterViewInit() {
        this.setUpFocusManagement();
    }
    close() {
        super.close();
        if (this._panelOpen) {
            this.selectAllElements.forEach((all) => all.setInactiveStyles());
        }
    }
    toggleAllOptions(selectAll = !this.allOptionsSelected) {
        if (this.allOptionsSelected !== selectAll) {
            this.selectionModel.clear();
            this.allOptionsSelected = selectAll;
            this.selectionModel.select(...this.options.filter((item) => this.allOptionsSelected
                ? item.checked || !item.disabled
                : item.checked && item.disabled));
            this.sortValues();
            this.propagateChanges();
        }
    }
    selectAllOptions() {
        this.toggleAllOptions(true);
    }
    deselectAllOptions() {
        this.toggleAllOptions(false);
    }
    clearSelectedOptions(event) {
        event.preventDefault();
        event.stopPropagation();
        this.selectionModel.clear();
        this.propagateChanges();
    }
    handleClosedKeydown(event) {
        if (this.withChips) {
            return;
        }
        const keyCode = event.keyCode;
        const isArrowKey = keyCode === DOWN_ARROW ||
            keyCode === UP_ARROW ||
            keyCode === LEFT_ARROW ||
            keyCode === RIGHT_ARROW;
        const isOpenKey = keyCode === ENTER || keyCode === SPACE;
        const manager = this.keyManager;
        // Open the select on ALT + arrow key to match the native <select>
        if ((!manager.isTyping() && isOpenKey && !hasModifierKey(event)) ||
            isArrowKey) {
            event.preventDefault(); // prevents the page from scrolling down when pressing space
            this.open();
        }
    }
    handleOpenKeydown(event) {
        const manager = this.keyManager;
        const keyCode = event.keyCode;
        const isArrowKey = keyCode === DOWN_ARROW || keyCode === UP_ARROW;
        const isTyping = manager.isTyping();
        if (isArrowKey && event.altKey) {
            // Close the select on ALT + arrow key to match the native <select>
            event.preventDefault();
            this.close();
            // Don't do anything in this case if the user is typing,
            // because the typing sequence can include the space key.
        }
        else if (!isTyping &&
            (keyCode === ENTER || keyCode === SPACE) &&
            manager.activeItem &&
            !hasModifierKey(event)) {
            event.preventDefault();
            manager.activeItem.handleClick();
        }
        else if (!isTyping && keyCode === A && event.ctrlKey) {
            event.preventDefault();
            this.toggleAllOptions();
            this.selectAllElements.forEach((option) => option.markForCheck());
        }
        else {
            const previouslyFocusedIndex = manager.activeItemIndex;
            manager.onKeydown(event);
            if (isArrowKey &&
                event.shiftKey &&
                manager.activeItem &&
                manager.activeItemIndex !== previouslyFocusedIndex) {
                manager.activeItem.handleClick();
            }
        }
    }
    setSelectionByValue(value) {
        if (value) {
            this.selectionModel.clear();
            value.forEach((currentValue) => this.selectValue(currentValue));
            this.sortValues();
        }
        this.cdr.markForCheck();
    }
    resetOptions() {
        super.resetOptions();
        this.optionSelectionChanges
            .pipe(takeUntil(this.options.changes), takeUntilDestroyed(this.destroyRef))
            .subscribe((event) => this.onSelect(event.source));
        // Listen to changes in the internal state of the options and react accordingly.
        // Handles cases like the labels of the selected options changing.
        merge(...this.options.map((option) => option.checkedChanged))
            .pipe(
        // should emit one time if states changes for many options
        debounceTime(1), takeUntil(this.options.changes), takeUntilDestroyed(this.destroyRef))
            .subscribe(() => {
            this.calculateAllOptionsSelected();
            this.refreshSelectAllCheck();
            this.selectAllElements.forEach((option) => option.markForCheck());
            this.stateChanges.next();
        });
        this.calculateAllOptionsSelected();
        this.refreshSelectAllCheck();
    }
    onSelect(option) {
        super.onSelect(option);
        this.sortValues();
        this.stateChanges.next();
    }
    sortValues() {
        const options = this.options.toArray();
        this.selectionModel.sort((a, b) => {
            return this.sortComparator
                ? this.sortComparator(a, b, options)
                : options.indexOf(a) - options.indexOf(b);
        });
        this.stateChanges.next();
    }
    getValueToEmit() {
        return this.selected.map((s) => s.value);
    }
    setUpFocusManagement() {
        this.chips.changes
            .pipe(startWith(this.chips), takeUntilDestroyed(this.destroyRef))
            .subscribe((chips) => {
            const actions = [];
            chips.forEach((chip) => chip.getActions().forEach((action) => actions.push(action)));
            this.chipActions.reset(actions);
            this.chipActions.notifyOnChanges();
        });
        this.keyManagerChips = new FocusKeyManager(this.chipActions)
            .withVerticalOrientation()
            .withHorizontalOrientation(this.dir ? this.dir.value : 'ltr')
            .withHomeAndEnd()
            .skipPredicate((action) => action.disabled);
        this.chips.changes
            .pipe(startWith(null), switchMap(() => merge(...this.chips.map((c) => c.focus$))), takeUntilDestroyed(this.destroyRef))
            .subscribe(({ source: chip }) => {
            const action = chip.getSourceAction(document.activeElement);
            if (action) {
                this.keyManagerChips.updateActiveItem(action);
            }
        });
        this.chips.changes
            .pipe(startWith(null), switchMap(() => merge(...this.chips.map((c) => c.focus$.pipe(map((e) => ({ focused: true, source: e.source })))), ...this.chips.map((c) => c.blur$.pipe(map((e) => ({ focused: false, source: e.source })))))), takeUntilDestroyed(this.destroyRef))
            .subscribe((chip) => {
            if (chip.focused) {
                this.focusedChip = chip.source;
            }
            else if (this.focusedChip === chip.source) {
                this.focusedChip = null;
            }
            this.cdr.markForCheck();
        });
        this.dir?.change
            .pipe(takeUntilDestroyed(this.destroyRef))
            .subscribe((direction) => this.keyManagerChips.withHorizontalOrientation(direction));
    }
    refreshSelectAllCheck() {
        this.selectAllElements.forEach((all) => (all.checked = this.allOptionsSelected));
        this.cdr.markForCheck();
    }
    calculateAllOptionsSelected() {
        this.allOptionsSelected = this.options
            .toArray()
            .filter((item) => !item.disabled)
            .every((item) => item.checked);
        this.cdr.markForCheck();
    }
    _setSize(size) {
        super._setSize(size);
        this.selectAllElements?.forEach((element) => (element.size = size));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsMultiSelectComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "17.3.1", type: RdsMultiSelectComponent, selector: "rds-multi-select", inputs: { withChips: "withChips" }, providers: [
            {
                provide: RdsAbstractFormFieldControl,
                useExisting: RdsMultiSelectComponent,
            },
        ], queries: [{ propertyName: "selectAllElements", predicate: RdsSelectAllOptionComponent }], viewQueries: [{ propertyName: "chips", predicate: RdsChipComponent, descendants: true }], usesInheritance: true, ngImport: i0, template: `
    <ng-container aria-hidden="true" (click)="toggle()">
      @if (shouldShowValue()) {
        <div class="rds-form-field__select-value" #wrapperElement>
          @if (withChips) {
            @if (triggerValues.length && triggerValues[0]) {
              <rds-chip-list size="s">
                @for (value of triggerValues; track value; let i = $index) {
                  <rds-chip
                    closable
                    (closeChip)="removeValue(i)"
                    #valueContainer
                    [disabled]="disabled"
                    [tabindex]="
                      !focusedChip || focusedChip === valueContainer ? 0 : -1
                    "
                  >
                    {{ value }}
                  </rds-chip>
                }
              </rds-chip-list>
            }
          } @else {
            <span
              class="rds-form-field__select-trigger-value"
              rds-trigger-values
              multiple
              #rdsTriggerValues="rdsTriggerValues"
              [triggerValues]="triggerValues"
              [wrapperElement]="wrapperElement"
            ></span>
            @if (((rdsTriggerValues.hiddenValues | async) || 0) > 0) {
              <span
                class="rds-form-field__select-hidden-trigger-values-placeholder"
              >
                <rds-badge
                  type="minimal"
                  size="s"
                  [color]="disabled ? 'black' : 'blue'"
                  [label]="'+' + (rdsTriggerValues.hiddenValues | async)"
                />
              </span>
            }
          }
          @if (withChips && triggerValues.length && triggerValues[0]) {
            <div class="rds-form-field__select-x-wrapper">
              <button
                rds-icon-button
                size="m"
                [disabled]="disabled"
                (click)="clearSelectedOptions($event)"
              >
                <rds-icon namespace="filled" icon="close" />
              </button>
            </div>
          }
        </div>
      } @else {
        @if (shouldShowPlaceholder()) {
          <div class="rds-form-field__select-value">
            <div
              class="rds-form-field__select-trigger-value rds-form-field__select-trigger-value-placeholder"
            >
              {{ placeholder }}
            </div>
          </div>
        }
      }
      <div class="rds-form-field__select-arrow-wrapper">
        <div
          class="rds-form-field__select-arrow"
          [class.rds-form-field__select-arrow--rotate]="panelOpen"
        ></div>
      </div>
    </ng-container>

    <ng-template
      cdk-connected-overlay
      cdkConnectedOverlayLockPosition
      cdkConnectedOverlayHasBackdrop
      cdkConnectedOverlayPanelClass="test"
      cdkConnectedOverlayBackdropClass="cdk-overlay-transparent-backdrop"
      [cdkConnectedOverlayScrollStrategy]="scrollStrategy"
      [cdkConnectedOverlayOrigin]="origin"
      [cdkConnectedOverlayOpen]="panelOpen"
      [cdkConnectedOverlayPositions]="positions"
      [cdkConnectedOverlayMinWidth]="minWidth"
      (backdropClick)="close(); onBlur()"
      (attach)="onAttached()"
      (detach)="close()"
    >
      <div class="rds-select__panel-wrapper" [@transformPanelWrap]="null">
        <div
          #panel
          class="rds-select__panel"
          aria-multiselectable="true"
          [attr.id]="id + '-panel'"
          [@transformPanel]="'showing'"
          (@transformPanel.done)="panelDoneAnimatingStream.next($event.toState)"
          [style.transformOrigin]="transformOrigin"
          (keydown)="handleKeydown($event)"
        >
          <ng-content select="rds-select-search-input" />
          <ng-content select="rds-select-all-option" />
          <ng-content select="rds-multi-select-option" />
        </div>
      </div>
    </ng-template>
  `, isInline: true, dependencies: [{ kind: "directive", type: i1.CdkConnectedOverlay, selector: "[cdk-connected-overlay], [connected-overlay], [cdkConnectedOverlay]", inputs: ["cdkConnectedOverlayOrigin", "cdkConnectedOverlayPositions", "cdkConnectedOverlayPositionStrategy", "cdkConnectedOverlayOffsetX", "cdkConnectedOverlayOffsetY", "cdkConnectedOverlayWidth", "cdkConnectedOverlayHeight", "cdkConnectedOverlayMinWidth", "cdkConnectedOverlayMinHeight", "cdkConnectedOverlayBackdropClass", "cdkConnectedOverlayPanelClass", "cdkConnectedOverlayViewportMargin", "cdkConnectedOverlayScrollStrategy", "cdkConnectedOverlayOpen", "cdkConnectedOverlayDisableClose", "cdkConnectedOverlayTransformOriginOn", "cdkConnectedOverlayHasBackdrop", "cdkConnectedOverlayLockPosition", "cdkConnectedOverlayFlexibleDimensions", "cdkConnectedOverlayGrowAfterOpen", "cdkConnectedOverlayPush", "cdkConnectedOverlayDisposeOnNavigation"], outputs: ["backdropClick", "positionChange", "attach", "detach", "overlayKeydown", "overlayOutsideClick"], exportAs: ["cdkConnectedOverlay"] }, { kind: "component", type: i2.RdsIconComponent, selector: "rds-icon", inputs: ["namespace", "icon"] }, { kind: "component", type: i3.RdsIconButtonComponent, selector: "\n    button[rds-icon-button],\n    button[rdsIconButton],\n    button[rds-icon-warning-button],\n    button[rdsIconWarningButton],\n  ", inputs: ["size"] }, { kind: "component", type: i4.RdsChipListComponent, selector: "rds-chip-list", inputs: ["tabIndex", "multiple", "selectable", "size", "compareFn", "value", "disabled"], outputs: ["valueChange"] }, { kind: "component", type: i5.RdsChipComponent, selector: "rds-chip, button[rds-chip], button[rdsChip]", inputs: ["ariaLabelCloseButton", "size", "type", "role", "disabled", "closable", "value"], outputs: ["closeChip", "destroyed"] }, { kind: "component", type: i6.RdsBadgeComponent, selector: "rds-badge", inputs: ["size", "label", "color", "type", "ariaDescribedBy"] }, { kind: "component", type: i7.RdsTriggerValuesComponent, selector: "[rds-trigger-values]", inputs: ["triggerValues", "wrapperElement", "multiple"], outputs: ["hiddenValues"], exportAs: ["rdsTriggerValues"] }, { kind: "pipe", type: i8.AsyncPipe, name: "async" }], animations: SelectAnimations, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsMultiSelectComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'rds-multi-select',
                    template: `
    <ng-container aria-hidden="true" (click)="toggle()">
      @if (shouldShowValue()) {
        <div class="rds-form-field__select-value" #wrapperElement>
          @if (withChips) {
            @if (triggerValues.length && triggerValues[0]) {
              <rds-chip-list size="s">
                @for (value of triggerValues; track value; let i = $index) {
                  <rds-chip
                    closable
                    (closeChip)="removeValue(i)"
                    #valueContainer
                    [disabled]="disabled"
                    [tabindex]="
                      !focusedChip || focusedChip === valueContainer ? 0 : -1
                    "
                  >
                    {{ value }}
                  </rds-chip>
                }
              </rds-chip-list>
            }
          } @else {
            <span
              class="rds-form-field__select-trigger-value"
              rds-trigger-values
              multiple
              #rdsTriggerValues="rdsTriggerValues"
              [triggerValues]="triggerValues"
              [wrapperElement]="wrapperElement"
            ></span>
            @if (((rdsTriggerValues.hiddenValues | async) || 0) > 0) {
              <span
                class="rds-form-field__select-hidden-trigger-values-placeholder"
              >
                <rds-badge
                  type="minimal"
                  size="s"
                  [color]="disabled ? 'black' : 'blue'"
                  [label]="'+' + (rdsTriggerValues.hiddenValues | async)"
                />
              </span>
            }
          }
          @if (withChips && triggerValues.length && triggerValues[0]) {
            <div class="rds-form-field__select-x-wrapper">
              <button
                rds-icon-button
                size="m"
                [disabled]="disabled"
                (click)="clearSelectedOptions($event)"
              >
                <rds-icon namespace="filled" icon="close" />
              </button>
            </div>
          }
        </div>
      } @else {
        @if (shouldShowPlaceholder()) {
          <div class="rds-form-field__select-value">
            <div
              class="rds-form-field__select-trigger-value rds-form-field__select-trigger-value-placeholder"
            >
              {{ placeholder }}
            </div>
          </div>
        }
      }
      <div class="rds-form-field__select-arrow-wrapper">
        <div
          class="rds-form-field__select-arrow"
          [class.rds-form-field__select-arrow--rotate]="panelOpen"
        ></div>
      </div>
    </ng-container>

    <ng-template
      cdk-connected-overlay
      cdkConnectedOverlayLockPosition
      cdkConnectedOverlayHasBackdrop
      cdkConnectedOverlayPanelClass="test"
      cdkConnectedOverlayBackdropClass="cdk-overlay-transparent-backdrop"
      [cdkConnectedOverlayScrollStrategy]="scrollStrategy"
      [cdkConnectedOverlayOrigin]="origin"
      [cdkConnectedOverlayOpen]="panelOpen"
      [cdkConnectedOverlayPositions]="positions"
      [cdkConnectedOverlayMinWidth]="minWidth"
      (backdropClick)="close(); onBlur()"
      (attach)="onAttached()"
      (detach)="close()"
    >
      <div class="rds-select__panel-wrapper" [@transformPanelWrap]="null">
        <div
          #panel
          class="rds-select__panel"
          aria-multiselectable="true"
          [attr.id]="id + '-panel'"
          [@transformPanel]="'showing'"
          (@transformPanel.done)="panelDoneAnimatingStream.next($event.toState)"
          [style.transformOrigin]="transformOrigin"
          (keydown)="handleKeydown($event)"
        >
          <ng-content select="rds-select-search-input" />
          <ng-content select="rds-select-all-option" />
          <ng-content select="rds-multi-select-option" />
        </div>
      </div>
    </ng-template>
  `,
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    animations: SelectAnimations,
                    providers: [
                        {
                            provide: RdsAbstractFormFieldControl,
                            useExisting: RdsMultiSelectComponent,
                        },
                    ],
                }]
        }], propDecorators: { withChips: [{
                type: Input
            }], selectAllElements: [{
                type: ContentChildren,
                args: [RdsSelectAllOptionComponent]
            }], chips: [{
                type: ViewChildren,
                args: [RdsChipComponent]
            }] } });
