// @rds/angular-components/esm2022/lib/form-field/select/select-search-input.component.mjs
import { A, hasModifierKey, SPACE } from '@angular/cdk/keycodes';
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, DestroyRef, ElementRef, EventEmitter, HostBinding, HostListener, inject, Input, Output, Renderer2, ViewChild, ViewEncapsulation, } from '@angular/core';
import { FormControl } from '@angular/forms';
import { noop } from 'rxjs';
import { tap } from 'rxjs/operators';
import { RdsInputDirective } from '../input.directive';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import * as i0 from "@angular/core";
import * as i1 from "../../icons/icon.component";
import * as i2 from "@angular/forms";
import * as i3 from "../../button/icon-button.component";
import * as i4 from "../input.directive";
let uniqueId = 0;
export class RdsSelectSearchInputComponent {
    constructor() {
        this.renderer = inject(Renderer2);
        this.cdr = inject(ChangeDetectorRef);
        this.elementRef = inject(ElementRef);
        this.destroyRef = inject(DestroyRef);
        this.control = new FormControl('');
        this.size = 'm';
        this.id = `rds-select__search-input-${uniqueId++}`;
        this.placeholder = 'Search...';
        this.changed = new EventEmitter();
    }
    get cssClasses() {
        return [
            'rds-select__option',
            'rds-select__option--search-input',
            `rds-select__option--search-input--${this.size}`,
        ];
    }
    handleClick() {
        this.search.nativeElement.focus();
        this.renderer.addClass(this.elementRef.nativeElement, 'cdk-focused');
        this.renderer.addClass(this.elementRef.nativeElement, 'cdk-keyboard-focused');
        this.cdr.markForCheck();
    }
    stopPropagation($event) {
        if ($event.keyCode === SPACE ||
            ($event.keyCode === A && hasModifierKey($event, 'ctrlKey'))) {
            $event.stopPropagation();
        }
    }
    rect() {
        return this.elementRef.nativeElement.getBoundingClientRect();
    }
    isVisible() {
        return true;
    }
    ngAfterViewInit() {
        if (this.searchInput && this.searchInput.ngControl) {
            this.searchInput.ngControl.valueChanges
                ?.pipe(tap((v) => this.changed.emit(v)), takeUntilDestroyed(this.destroyRef))
                .subscribe({ next: noop, error: noop });
        }
        this.focus();
    }
    reset() {
        this.searchInput?.ngControl?.reset();
    }
    focus(origin, options) {
        const element = this.elementRef.nativeElement;
        if (typeof element.focus === 'function') {
            element.focus(options);
            this.handleClick();
        }
    }
    isDirtyValue() {
        return this.control.value !== '';
    }
    resetInput() {
        if (this.isDirtyValue()) {
            this.control.setValue('');
            this.search.nativeElement.focus();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSelectSearchInputComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsSelectSearchInputComponent, selector: "rds-select-search-input", inputs: { size: "size", placeholder: "placeholder" }, outputs: { changed: "changed" }, host: { listeners: { "click": "handleClick()", "focus": "handleClick()", "keydown": "stopPropagation($event)" }, properties: { "class": "this.cssClasses" } }, viewQueries: [{ propertyName: "searchInput", first: true, predicate: RdsInputDirective, descendants: true }, { propertyName: "search", first: true, predicate: ["search"], descendants: true, read: ElementRef, static: true }], ngImport: i0, template: `
    <rds-icon
      class="rds-search-icon rds-select__option--search-input-prefix"
      namespace="filled"
      icon="search"
      aria-hidden="true"
    />
    <input
      [formControl]="control"
      #search
      rdsInput
      (search)="$event.preventDefault(); $event.stopPropagation()"
      [placeholder]="placeholder"
    />
    <button
      rds-icon-button
      [size]="'s'"
      [class.hidden]="!isDirtyValue()"
      (click)="resetInput(); $event.stopPropagation()"
      class="rds-search-reset-button rds-select__option--search-input-suffix"
    >
      <rds-icon namespace="filled" icon="close" aria-hidden="true" />
    </button>
  `, isInline: true, styles: [".rds-select__option--search-input{display:flex;flex-direction:row;border-radius:2px;border:1px solid var(--rds-color-stroke-enabled);align-items:center;margin:8px 16px}.rds-select__option--search-input--l{height:48px;padding:4px 15px}.rds-select__option--search-input--l rds-icon{height:24px;width:24px;color:var(--rds-color-low-contrast)}.rds-select__option--search-input--l input{font-size:18px;font-weight:200;line-height:24px;text-transform:none;letter-spacing:.2px;font-family:roche-sans;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;width:calc(100% - 6px);border:none;box-shadow:none;background:transparent;color:var(--rds-color-high-contrast);caret-color:var(--rds-color-high-contrast)}.rds-select__option--search-input--l input:hover,.rds-select__option--search-input--l input:focus{outline:none}.rds-select__option--search-input--l input ::placeholder{color:var(--rds-color-low-contrast)}.rds-select__option--search-input .rds-select__option--search-input-suffix{padding-right:8px}.rds-select__option--search-input .rds-select__option--search-input-suffix rds-icon{margin-top:0}.rds-select__option--search-input--m{height:36px;padding:4px 11px}.rds-select__option--search-input--m rds-icon{height:18px;width:18px;color:var(--rds-color-low-contrast)}.rds-select__option--search-input--m input{font-size:16px;font-weight:200;line-height:24px;text-transform:none;letter-spacing:.2px;font-family:roche-sans;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;width:calc(100% + -0px);border:none;box-shadow:none;background:transparent;color:var(--rds-color-high-contrast);caret-color:var(--rds-color-high-contrast)}.rds-select__option--search-input--m input:hover,.rds-select__option--search-input--m input:focus{outline:none}.rds-select__option--search-input--m input ::placeholder{color:var(--rds-color-low-contrast)}.rds-select__option--search-input .rds-select__option--search-input-suffix{padding-right:4px}.rds-select__option--search-input .rds-select__option--search-input-suffix rds-icon{margin-top:2px}.rds-select__option--search-input--s{height:28px;padding:4px 7px}.rds-select__option--search-input--s rds-icon{height:18px;width:18px;color:var(--rds-color-low-contrast)}.rds-select__option--search-input--s input{font-size:14px;font-weight:200;line-height:20px;text-transform:none;letter-spacing:.3px;font-family:roche-sans;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;width:calc(100% + -0px);border:none;box-shadow:none;background:transparent;color:var(--rds-color-high-contrast);caret-color:var(--rds-color-high-contrast)}.rds-select__option--search-input--s input:hover,.rds-select__option--search-input--s input:focus{outline:none}.rds-select__option--search-input--s input ::placeholder{color:var(--rds-color-low-contrast)}.rds-select__option--search-input .rds-select__option--search-input-prefix{margin-right:8px}.rds-select__option--search-input .rds-select__option--search-input-suffix{padding-right:0}.rds-select__option--search-input .rds-select__option--search-input-suffix rds-icon{margin-top:-1px}.rds-select__option--search-input .rds-select__option--search-input-suffix.hidden{visibility:hidden}.rds-select__option--search-input:hover{border-color:var(--rds-color-stroke-hover)}\n"], dependencies: [{ kind: "component", type: i1.RdsIconComponent, selector: "rds-icon", inputs: ["namespace", "icon"] }, { kind: "directive", type: i2.DefaultValueAccessor, selector: "input:not([type=checkbox])[formControlName],textarea[formControlName],input:not([type=checkbox])[formControl],textarea[formControl],input:not([type=checkbox])[ngModel],textarea[ngModel],[ngDefaultControl]" }, { kind: "directive", type: i2.NgControlStatus, selector: "[formControlName],[ngModel],[formControl]" }, { kind: "directive", type: i2.FormControlDirective, selector: "[formControl]", inputs: ["formControl", "disabled", "ngModel"], outputs: ["ngModelChange"], exportAs: ["ngForm"] }, { kind: "component", type: i3.RdsIconButtonComponent, selector: "\n    button[rds-icon-button],\n    button[rdsIconButton],\n    button[rds-icon-warning-button],\n    button[rdsIconWarningButton],\n  ", inputs: ["size"] }, { kind: "directive", type: i4.RdsInputDirective, selector: " input[rds-input], input[rdsInput], textarea[rds-input], textarea[rdsInput]", inputs: ["id", "disabled", "required", "readonly", "type", "placeholder", "value"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSelectSearchInputComponent, decorators: [{
            type: Component,
            args: [{ selector: 'rds-select-search-input', template: `
    <rds-icon
      class="rds-search-icon rds-select__option--search-input-prefix"
      namespace="filled"
      icon="search"
      aria-hidden="true"
    />
    <input
      [formControl]="control"
      #search
      rdsInput
      (search)="$event.preventDefault(); $event.stopPropagation()"
      [placeholder]="placeholder"
    />
    <button
      rds-icon-button
      [size]="'s'"
      [class.hidden]="!isDirtyValue()"
      (click)="resetInput(); $event.stopPropagation()"
      class="rds-search-reset-button rds-select__option--search-input-suffix"
    >
      <rds-icon namespace="filled" icon="close" aria-hidden="true" />
    </button>
  `, changeDetection: ChangeDetectionStrategy.OnPush, encapsulation: ViewEncapsulation.None, styles: [".rds-select__option--search-input{display:flex;flex-direction:row;border-radius:2px;border:1px solid var(--rds-color-stroke-enabled);align-items:center;margin:8px 16px}.rds-select__option--search-input--l{height:48px;padding:4px 15px}.rds-select__option--search-input--l rds-icon{height:24px;width:24px;color:var(--rds-color-low-contrast)}.rds-select__option--search-input--l input{font-size:18px;font-weight:200;line-height:24px;text-transform:none;letter-spacing:.2px;font-family:roche-sans;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;width:calc(100% - 6px);border:none;box-shadow:none;background:transparent;color:var(--rds-color-high-contrast);caret-color:var(--rds-color-high-contrast)}.rds-select__option--search-input--l input:hover,.rds-select__option--search-input--l input:focus{outline:none}.rds-select__option--search-input--l input ::placeholder{color:var(--rds-color-low-contrast)}.rds-select__option--search-input .rds-select__option--search-input-suffix{padding-right:8px}.rds-select__option--search-input .rds-select__option--search-input-suffix rds-icon{margin-top:0}.rds-select__option--search-input--m{height:36px;padding:4px 11px}.rds-select__option--search-input--m rds-icon{height:18px;width:18px;color:var(--rds-color-low-contrast)}.rds-select__option--search-input--m input{font-size:16px;font-weight:200;line-height:24px;text-transform:none;letter-spacing:.2px;font-family:roche-sans;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;width:calc(100% + -0px);border:none;box-shadow:none;background:transparent;color:var(--rds-color-high-contrast);caret-color:var(--rds-color-high-contrast)}.rds-select__option--search-input--m input:hover,.rds-select__option--search-input--m input:focus{outline:none}.rds-select__option--search-input--m input ::placeholder{color:var(--rds-color-low-contrast)}.rds-select__option--search-input .rds-select__option--search-input-suffix{padding-right:4px}.rds-select__option--search-input .rds-select__option--search-input-suffix rds-icon{margin-top:2px}.rds-select__option--search-input--s{height:28px;padding:4px 7px}.rds-select__option--search-input--s rds-icon{height:18px;width:18px;color:var(--rds-color-low-contrast)}.rds-select__option--search-input--s input{font-size:14px;font-weight:200;line-height:20px;text-transform:none;letter-spacing:.3px;font-family:roche-sans;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;width:calc(100% + -0px);border:none;box-shadow:none;background:transparent;color:var(--rds-color-high-contrast);caret-color:var(--rds-color-high-contrast)}.rds-select__option--search-input--s input:hover,.rds-select__option--search-input--s input:focus{outline:none}.rds-select__option--search-input--s input ::placeholder{color:var(--rds-color-low-contrast)}.rds-select__option--search-input .rds-select__option--search-input-prefix{margin-right:8px}.rds-select__option--search-input .rds-select__option--search-input-suffix{padding-right:0}.rds-select__option--search-input .rds-select__option--search-input-suffix rds-icon{margin-top:-1px}.rds-select__option--search-input .rds-select__option--search-input-suffix.hidden{visibility:hidden}.rds-select__option--search-input:hover{border-color:var(--rds-color-stroke-hover)}\n"] }]
        }], propDecorators: { size: [{
                type: Input
            }], searchInput: [{
                type: ViewChild,
                args: [RdsInputDirective]
            }], search: [{
                type: ViewChild,
                args: ['search', { static: true, read: ElementRef }]
            }], cssClasses: [{
                type: HostBinding,
                args: ['class']
            }], placeholder: [{
                type: Input
            }], changed: [{
                type: Output
            }], handleClick: [{
                type: HostListener,
                args: ['click']
            }, {
                type: HostListener,
                args: ['focus']
            }], stopPropagation: [{
                type: HostListener,
                args: ['keydown', ['$event']]
            }] } });
