// @rds/angular-components/esm2022/lib/form-field/select/select.component.mjs
import { SelectionModel } from '@angular/cdk/collections';
import { DOWN_ARROW, END, ENTER, hasModifierKey, HOME, LEFT_ARROW, RIGHT_ARROW, SPACE, UP_ARROW, } from '@angular/cdk/keycodes';
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation, } from '@angular/core';
import { merge } from 'rxjs';
import { debounceTime, takeUntil } from 'rxjs/operators';
import { RdsAbstractFormFieldControl } from '../abstract-form-field-control';
import { SelectAnimations } from './select.animations';
import { RdsAbstractSelectComponent } from './abstract-select.component';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/overlay";
import * as i2 from "./trigger-values.component";
export class RdsSelectComponent extends RdsAbstractSelectComponent {
    constructor() {
        super(...arguments);
        this._name = 'rds-select';
    }
    set multiple(_multiple) {
        throw new Error(`Input 'multiple' is not valid anymore and will be removed in the next release. Use 'rds-multi-select' component instead.`);
    }
    get selected() {
        return this.selectionModel.selected[0];
    }
    get triggerValues() {
        if (this.empty) {
            return [''];
        }
        return [this.selectionModel.selected[0].viewValue];
    }
    ngOnInit() {
        super.ngOnInit();
        this.selectionModel = new SelectionModel(false);
    }
    handleClosedKeydown(event) {
        const keyCode = event.keyCode;
        const isArrowKey = keyCode === DOWN_ARROW ||
            keyCode === UP_ARROW ||
            keyCode === LEFT_ARROW ||
            keyCode === RIGHT_ARROW;
        const isOpenKey = keyCode === ENTER || keyCode === SPACE;
        const manager = this.keyManager;
        // Open the select on ALT + arrow key to match the native <select>
        if ((!manager.isTyping() && isOpenKey && !hasModifierKey(event)) ||
            (event.altKey && isArrowKey)) {
            event.preventDefault(); // prevents the page from scrolling down when pressing space
            this.open();
        }
        else {
            const previouslySelectedOption = this.selected;
            if (keyCode === HOME || keyCode === END) {
                keyCode === HOME
                    ? manager.setFirstItemActive()
                    : manager.setLastItemActive();
                event.preventDefault();
            }
            else {
                manager.onKeydown(event);
            }
            const selectedOption = this.selected;
            if (selectedOption && previouslySelectedOption !== selectedOption) {
                this.liveAnnouncer.announce(selectedOption.viewValue, 10000);
            }
        }
    }
    handleOpenKeydown(event) {
        const manager = this.keyManager;
        const keyCode = event.keyCode;
        const isArrowKey = keyCode === DOWN_ARROW || keyCode === UP_ARROW;
        const isTyping = manager.isTyping();
        if (isArrowKey && event.altKey) {
            // Close the select on ALT + arrow key to match the native <select>
            event.preventDefault();
            this.close();
            // Don't do anything in this case if the user is typing,
            // because the typing sequence can include the space key.
        }
        else if (!isTyping &&
            (keyCode === ENTER || keyCode === SPACE) &&
            manager.activeItem &&
            !hasModifierKey(event)) {
            event.preventDefault();
            manager.activeItem.handleClick();
        }
        else {
            manager.onKeydown(event);
        }
    }
    setSelectionByValue(value) {
        this.selectionModel.clear();
        const correspondingOption = this.selectValue(value);
        // Shift focus to the active item. Note that we shouldn't do this in multiple
        // mode, because we don't know what option the user interacted with last.
        if (correspondingOption) {
            this.keyManager.updateActiveItem(correspondingOption);
        }
        else if (!this.panelOpen) {
            // Otherwise reset the highlighted option. Note that we only want to do this while
            // closed, because doing it while open can shift the user's focus unnecessarily.
            this.keyManager.updateActiveItem(-1);
        }
        this.cdr.markForCheck();
    }
    initKeyManager() {
        super.initKeyManager();
        this.keyManager.tabOut
            .pipe(takeUntil(this.destroyKeyManager$), takeUntilDestroyed(this.destroyRef))
            .subscribe(() => {
            if (this.panelOpen) {
                // Select the active item when tabbing away. This is consistent with how the native
                // select behaves. Note that we only want to do this in single selection mode.
                if (this.keyManager.activeItem) {
                    this.keyManager.activeItem.toggle();
                }
            }
        });
        this.keyManager.change
            .pipe(takeUntil(this.destroyKeyManager$), takeUntilDestroyed(this.destroyRef))
            .subscribe(() => {
            if (!this._panelOpen && this.keyManager.activeItem) {
                this.keyManager.activeItem.toggle();
            }
        });
    }
    resetOptions() {
        super.resetOptions();
        this.optionSelectionChanges
            .pipe(takeUntil(this.options.changes), takeUntilDestroyed(this.destroyRef))
            .subscribe((event) => {
            this.onSelect(event.source, event.isUserInput);
            if (event.isUserInput && this._panelOpen) {
                this.close();
                this.focus();
            }
        });
        // Listen to changes in the internal state of the options and react accordingly.
        // Handles cases like the labels of the selected options changing.
        merge(...this.options.map((option) => option.checkedChanged))
            .pipe(
        // should emit one time if states changes for many options
        debounceTime(1), takeUntil(this.options.changes), takeUntilDestroyed(this.destroyRef))
            .subscribe(() => {
            this.stateChanges.next();
        });
    }
    onSelect(option, isUserInput = false) {
        super.onSelect(option);
        if (isUserInput) {
            this.keyManager.setActiveItem(option);
        }
        this.stateChanges.next();
    }
    getValueToEmit() {
        return this.selected ? this.selected.value : null;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSelectComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "17.3.1", type: RdsSelectComponent, selector: "rds-select", inputs: { multiple: "multiple" }, providers: [
            { provide: RdsAbstractFormFieldControl, useExisting: RdsSelectComponent },
        ], usesInheritance: true, ngImport: i0, template: `
    <ng-container aria-hidden="true" (click)="toggle()">
      @if (shouldShowValue()) {
        <div class="rds-form-field__select-value" #wrapperElement>
          <span
            class="rds-form-field__select-trigger-value"
            rds-trigger-values
            #rdsTriggerValues="rdsTriggerValues"
            [triggerValues]="triggerValues"
            [wrapperElement]="wrapperElement"
          ></span>
        </div>
      } @else {
        @if (shouldShowPlaceholder()) {
          <div class="rds-form-field__select-value">
            <div
              class="rds-form-field__select-trigger-value rds-form-field__select-trigger-value-placeholder"
            >
              {{ placeholder }}
            </div>
          </div>
        }
      }

      <div class="rds-form-field__select-arrow-wrapper">
        <div
          class="rds-form-field__select-arrow"
          [class.rds-form-field__select-arrow--rotate]="panelOpen"
        ></div>
      </div>

      <ng-template
        cdk-connected-overlay
        cdkConnectedOverlayLockPosition
        cdkConnectedOverlayHasBackdrop
        cdkConnectedOverlayBackdropClass="cdk-overlay-transparent-backdrop"
        [cdkConnectedOverlayScrollStrategy]="scrollStrategy"
        [cdkConnectedOverlayOrigin]="origin"
        [cdkConnectedOverlayOpen]="panelOpen"
        [cdkConnectedOverlayPositions]="positions"
        [cdkConnectedOverlayMinWidth]="minWidth"
        (backdropClick)="close(); onBlur()"
        (attach)="onAttached()"
        (detach)="close()"
      >
        <div class="rds-select__panel-wrapper" [@transformPanelWrap]="null">
          <div
            #panel
            class="rds-select__panel"
            [attr.id]="id + '-panel'"
            [@transformPanel]="'showing'"
            (@transformPanel.done)="
              panelDoneAnimatingStream.next($event.toState)
            "
            [style.transformOrigin]="transformOrigin"
            (keydown)="handleKeydown($event)"
          >
            <ng-content select="rds-select-search-input" />
            <ng-content select="rds-select-option" />
          </div>
        </div>
      </ng-template>
    </ng-container>
  `, isInline: true, dependencies: [{ kind: "directive", type: i1.CdkConnectedOverlay, selector: "[cdk-connected-overlay], [connected-overlay], [cdkConnectedOverlay]", inputs: ["cdkConnectedOverlayOrigin", "cdkConnectedOverlayPositions", "cdkConnectedOverlayPositionStrategy", "cdkConnectedOverlayOffsetX", "cdkConnectedOverlayOffsetY", "cdkConnectedOverlayWidth", "cdkConnectedOverlayHeight", "cdkConnectedOverlayMinWidth", "cdkConnectedOverlayMinHeight", "cdkConnectedOverlayBackdropClass", "cdkConnectedOverlayPanelClass", "cdkConnectedOverlayViewportMargin", "cdkConnectedOverlayScrollStrategy", "cdkConnectedOverlayOpen", "cdkConnectedOverlayDisableClose", "cdkConnectedOverlayTransformOriginOn", "cdkConnectedOverlayHasBackdrop", "cdkConnectedOverlayLockPosition", "cdkConnectedOverlayFlexibleDimensions", "cdkConnectedOverlayGrowAfterOpen", "cdkConnectedOverlayPush", "cdkConnectedOverlayDisposeOnNavigation"], outputs: ["backdropClick", "positionChange", "attach", "detach", "overlayKeydown", "overlayOutsideClick"], exportAs: ["cdkConnectedOverlay"] }, { kind: "component", type: i2.RdsTriggerValuesComponent, selector: "[rds-trigger-values]", inputs: ["triggerValues", "wrapperElement", "multiple"], outputs: ["hiddenValues"], exportAs: ["rdsTriggerValues"] }], animations: SelectAnimations, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSelectComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'rds-select',
                    template: `
    <ng-container aria-hidden="true" (click)="toggle()">
      @if (shouldShowValue()) {
        <div class="rds-form-field__select-value" #wrapperElement>
          <span
            class="rds-form-field__select-trigger-value"
            rds-trigger-values
            #rdsTriggerValues="rdsTriggerValues"
            [triggerValues]="triggerValues"
            [wrapperElement]="wrapperElement"
          ></span>
        </div>
      } @else {
        @if (shouldShowPlaceholder()) {
          <div class="rds-form-field__select-value">
            <div
              class="rds-form-field__select-trigger-value rds-form-field__select-trigger-value-placeholder"
            >
              {{ placeholder }}
            </div>
          </div>
        }
      }

      <div class="rds-form-field__select-arrow-wrapper">
        <div
          class="rds-form-field__select-arrow"
          [class.rds-form-field__select-arrow--rotate]="panelOpen"
        ></div>
      </div>

      <ng-template
        cdk-connected-overlay
        cdkConnectedOverlayLockPosition
        cdkConnectedOverlayHasBackdrop
        cdkConnectedOverlayBackdropClass="cdk-overlay-transparent-backdrop"
        [cdkConnectedOverlayScrollStrategy]="scrollStrategy"
        [cdkConnectedOverlayOrigin]="origin"
        [cdkConnectedOverlayOpen]="panelOpen"
        [cdkConnectedOverlayPositions]="positions"
        [cdkConnectedOverlayMinWidth]="minWidth"
        (backdropClick)="close(); onBlur()"
        (attach)="onAttached()"
        (detach)="close()"
      >
        <div class="rds-select__panel-wrapper" [@transformPanelWrap]="null">
          <div
            #panel
            class="rds-select__panel"
            [attr.id]="id + '-panel'"
            [@transformPanel]="'showing'"
            (@transformPanel.done)="
              panelDoneAnimatingStream.next($event.toState)
            "
            [style.transformOrigin]="transformOrigin"
            (keydown)="handleKeydown($event)"
          >
            <ng-content select="rds-select-search-input" />
            <ng-content select="rds-select-option" />
          </div>
        </div>
      </ng-template>
    </ng-container>
  `,
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    animations: SelectAnimations,
                    providers: [
                        { provide: RdsAbstractFormFieldControl, useExisting: RdsSelectComponent },
                    ],
                }]
        }], propDecorators: { multiple: [{
                type: Input
            }] } });
