// @rds/angular-components/esm2022/lib/form-field/select/select-all-option.component.mjs
import { ChangeDetectionStrategy, Component, EventEmitter, Output, ViewEncapsulation, } from '@angular/core';
import { RdsAbstractSelectableListItemComponent } from '../../list-item';
import { AbstractFocusableOption } from './abstract-focusable-option';
import { RdsSelectAllChangedEvent } from './select-all-changed.event';
import * as i0 from "@angular/core";
import * as i1 from "../../checkbox/pseudo-checkbox.component";
let uniqueId = 0;
export class RdsSelectAllOptionComponent extends RdsAbstractSelectableListItemComponent {
    constructor() {
        super(...arguments);
        this.id = `rds-select__select-all-${uniqueId++}`;
        this.selectAllChanged = new EventEmitter();
    }
    rect() {
        return this.elementRef.nativeElement.getBoundingClientRect();
    }
    isVisible() {
        return true;
    }
    get hostElement() {
        return this.elementRef.nativeElement;
    }
    setActiveStyles() {
        this.focus('keyboard');
    }
    setInactiveStyles() {
        this.blur();
    }
    getSelectTypeClass() {
        return 'rds-select__option--select-all';
    }
    toggle() {
        super.toggle();
        this.selectAllChanged.emit(new RdsSelectAllChangedEvent(this, this.checked));
        this.cdr.markForCheck();
    }
    handleClick() {
        this.toggle();
    }
    markForCheck() {
        this.cdr.markForCheck();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSelectAllOptionComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsSelectAllOptionComponent, selector: "rds-select-all-option", outputs: { selectAllChanged: "selectAllChanged" }, providers: [
            {
                provide: AbstractFocusableOption,
                useExisting: RdsSelectAllOptionComponent,
            },
        ], usesInheritance: true, ngImport: i0, template: `
    <rds-pseudo-checkbox noSpacing [checked]="checked" [disabled]="disabled" />
    <span>Select all</span>
  `, isInline: true, dependencies: [{ kind: "component", type: i1.RdsPseudoCheckboxComponent, selector: "rds-pseudo-checkbox", inputs: ["checked", "disabled", "indeterminate", "noSpacing"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSelectAllOptionComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'rds-select-all-option',
                    template: `
    <rds-pseudo-checkbox noSpacing [checked]="checked" [disabled]="disabled" />
    <span>Select all</span>
  `,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    providers: [
                        {
                            provide: AbstractFocusableOption,
                            useExisting: RdsSelectAllOptionComponent,
                        },
                    ],
                }]
        }], propDecorators: { selectAllChanged: [{
                type: Output
            }] } });
