// @rds/angular-components/esm2022/lib/form-field/select/select-option-separator.component.mjs
import { ChangeDetectionStrategy, Component, HostBinding, ViewEncapsulation, } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsSelectOptionSeparatorComponent {
    constructor() {
        this.class = 'rds-select__option--separator';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSelectOptionSeparatorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsSelectOptionSeparatorComponent, selector: "rds-select-option-separator", host: { properties: { "class": "this.class" } }, ngImport: i0, template: '<div>&nbsp;</div>', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSelectOptionSeparatorComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'rds-select-option-separator',
                    template: '<div>&nbsp;</div>',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }] } });
