// @rds/angular-components/esm2022/lib/form-field/select/single-select-option.component.mjs
import { Component, HostBinding, Input } from '@angular/core';
import { Subject } from 'rxjs';
import { RdsListItemComponent } from '../../list-item';
import { RDS_SINGLE_SELECT_LIST_ITEM_TEMPLATE, RdsSingleSelectListItemComponent, } from '../../list-item';
import { AbstractFocusableOption } from './abstract-focusable-option';
import { RdsSelectOptionSelectionChangedEvent } from './select-option-selection-changed.event';
import * as i0 from "@angular/core";
import * as i1 from "../../icons/icon.component";
let nextId = 0;
export class RdsSingleSelectOptionComponent extends RdsSingleSelectListItemComponent {
    constructor() {
        super(...arguments);
        this._currentViewValue = '';
        this.hostClass = 'rds-select__option';
        this.id = `rds-option-${nextId++}`;
        this.selectionChange$ = new Subject();
    }
    get viewValue() {
        return (this.elementRef.nativeElement.textContent || '').trim();
    }
    ngAfterViewChecked() {
        if (this.checked) {
            const viewValue = this.viewValue;
            if (viewValue !== this._currentViewValue) {
                this._currentViewValue = viewValue;
            }
        }
    }
    rect() {
        return this.elementRef.nativeElement.getBoundingClientRect();
    }
    isVisible() {
        return !this.elementRef.nativeElement.classList.contains('hide');
    }
    toggleVisibility(value) {
        if (value === true) {
            this.show();
        }
        else if (value === false) {
            this.hide();
        }
        else {
            if (this.isVisible()) {
                this.hide();
            }
            else if (!this.isVisible()) {
                this.show();
            }
        }
    }
    hide() {
        this.renderer.addClass(this.elementRef.nativeElement, 'hide');
    }
    show() {
        this.renderer.removeClass(this.elementRef.nativeElement, 'hide');
    }
    toggle() {
        this.handleClick();
    }
    handleClick() {
        if (!this.disabled && !this.checked) {
            this.checked = true;
            this.selectionChange$.next(new RdsSelectOptionSelectionChangedEvent(this, true));
        }
    }
    setActiveStyles() {
        this.focus('keyboard');
    }
    setInactiveStyles() {
        if (this._focused) {
            this.blur();
            this.cdr.markForCheck();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSingleSelectOptionComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsSingleSelectOptionComponent, selector: "rds-select-option", inputs: { id: "id", value: "value" }, host: { properties: { "class": "this.hostClass", "attr.id": "this.id" } }, providers: [
            {
                provide: RdsListItemComponent,
                useExisting: RdsSingleSelectOptionComponent,
            },
            {
                provide: AbstractFocusableOption,
                useExisting: RdsSingleSelectOptionComponent,
            },
        ], usesInheritance: true, ngImport: i0, template: "<ng-content select=\"rds-avatar\"/>\n<ng-content select=\"rds-icon\"></ng-content>\n<span class=\"rds-list-item-text\">\n  <ng-content />\n</span>\n\n<rds-icon icon=\"checkmark\" class=\"rds-list-item-checkmark\"/>", isInline: true, dependencies: [{ kind: "component", type: i1.RdsIconComponent, selector: "rds-icon", inputs: ["namespace", "icon"] }] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSingleSelectOptionComponent, decorators: [{
            type: Component,
            args: [{
                    template: RDS_SINGLE_SELECT_LIST_ITEM_TEMPLATE,
                    selector: 'rds-select-option',
                    providers: [
                        {
                            provide: RdsListItemComponent,
                            useExisting: RdsSingleSelectOptionComponent,
                        },
                        {
                            provide: AbstractFocusableOption,
                            useExisting: RdsSingleSelectOptionComponent,
                        },
                    ],
                }]
        }], propDecorators: { hostClass: [{
                type: HostBinding,
                args: ['class']
            }], id: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['attr.id']
            }], value: [{
                type: Input
            }] } });
