// @rds/angular-components/esm2022/lib/sidenav/sidenav-menuitem/sidenav-menuitem.component.mjs
import { ChangeDetectionStrategy, Component, DestroyRef, ElementRef, HostBinding, HostListener, inject, ViewEncapsulation, } from '@angular/core';
import { Subject } from 'rxjs';
import { FocusMonitor } from '@angular/cdk/a11y';
import { RdsSidenavService } from '../sidenav.service';
import { filter } from 'rxjs/operators';
import { rdsExpansionAnimations } from '../../accordion';
import { RDS_SIDENAV_MENU_PANEL } from '../sidenav.types';
import { SidenavItem } from '../sidenav-item';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import * as i0 from "@angular/core";
export class RdsSidenavMenuitemComponent {
    constructor() {
        this._parentMenu = inject(RDS_SIDENAV_MENU_PANEL, {
            optional: true,
        });
        this.sidenavService = inject(RdsSidenavService);
        this.destroyRef = inject(DestroyRef);
        this.elementRef = inject(ElementRef);
        this.focusMonitor = inject(FocusMonitor);
        this.role = 'menuitem';
        this._hovered = new Subject();
        this._focused = new Subject();
        this.nativeElement = this.elementRef.nativeElement;
        this.rdsMenuItemClass = true;
        this.rdsFocusIndicatorClass = true;
    }
    focus(origin, options) {
        if (origin) {
            this.focusMonitor.focusVia(this.nativeElement, origin, options);
        }
        else {
            this.nativeElement.focus(options);
        }
    }
    ngOnInit() {
        this.sidenavService.isMinimal$
            .pipe(takeUntilDestroyed(this.destroyRef))
            .subscribe((v) => (this.isMinimal = v));
    }
    ngAfterViewInit() {
        this.focusMonitor
            .monitor(this.elementRef, false)
            .pipe(filter((v) => !!v), takeUntilDestroyed(this.destroyRef))
            .subscribe(() => this._focused.next(this));
    }
    ngOnDestroy() {
        this.focusMonitor.stopMonitoring(this.elementRef);
        this._hovered.complete();
        this._focused.complete();
    }
    _handleMouseEnter() {
        this._hovered.next(this);
    }
    getLabel() {
        const clone = this.elementRef.nativeElement.cloneNode(true);
        const icons = clone.querySelectorAll('rds-icon');
        for (let i = 0; i < icons.length; i++) {
            icons[i].remove();
        }
        return clone.textContent?.trim() || '';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSidenavMenuitemComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsSidenavMenuitemComponent, selector: "rds-sidenav-menuitem, [rds-sidenav-menuitem], [rdsSidenavMenuitem]", host: { listeners: { "mouseenter": "_handleMouseEnter()" }, properties: { "attr.role": "this.role", "class.rds-sidenav-minimal": "this.isMinimal", "class.rds-sidenav-menuitem": "this.rdsMenuItemClass", "class.rds-focus-indicator": "this.rdsFocusIndicatorClass" } }, providers: [
            { provide: SidenavItem, useExisting: RdsSidenavMenuitemComponent },
        ], exportAs: ["rdsSidenavMenuitem"], ngImport: i0, template: "<ng-content select=\"rds-icon\" />\n<div class=\"sr-only-when-minimal-and-first-level\">\n  <ng-content />\n</div>\n<div class=\"rds-sidenav-badge-container\">\n  <ng-content select=\"rds-badge\" />\n</div>\n", animations: [rdsExpansionAnimations.bodyExpansion], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSidenavMenuitemComponent, decorators: [{
            type: Component,
            args: [{ selector: 'rds-sidenav-menuitem, [rds-sidenav-menuitem], [rdsSidenavMenuitem]', exportAs: 'rdsSidenavMenuitem', changeDetection: ChangeDetectionStrategy.OnPush, encapsulation: ViewEncapsulation.None, animations: [rdsExpansionAnimations.bodyExpansion], providers: [
                        { provide: SidenavItem, useExisting: RdsSidenavMenuitemComponent },
                    ], template: "<ng-content select=\"rds-icon\" />\n<div class=\"sr-only-when-minimal-and-first-level\">\n  <ng-content />\n</div>\n<div class=\"rds-sidenav-badge-container\">\n  <ng-content select=\"rds-badge\" />\n</div>\n" }]
        }], propDecorators: { role: [{
                type: HostBinding,
                args: ['attr.role']
            }], isMinimal: [{
                type: HostBinding,
                args: ['class.rds-sidenav-minimal']
            }], rdsMenuItemClass: [{
                type: HostBinding,
                args: ['class.rds-sidenav-menuitem']
            }], rdsFocusIndicatorClass: [{
                type: HostBinding,
                args: ['class.rds-focus-indicator']
            }], _handleMouseEnter: [{
                type: HostListener,
                args: ['mouseenter']
            }] } });
