// @rds/angular-components/esm2022/lib/search/search.module.mjs
import { NgModule } from '@angular/core';
import { RdsSearchComponent } from './search.component';
import { RdsFormFieldModule } from '../form-field';
import { RdsIconComponent } from '../icons';
import { FormsModule } from '@angular/forms';
import { RdsButtonModule } from '../button';
import { RdsGlobalSearchComponent } from './global-search.component';
import * as i0 from "@angular/core";
const COMPONENTS = [RdsSearchComponent, RdsGlobalSearchComponent];
export class RdsSearchModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSearchModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsSearchModule, declarations: [RdsSearchComponent, RdsGlobalSearchComponent], imports: [RdsFormFieldModule, RdsIconComponent, FormsModule, RdsButtonModule], exports: [RdsSearchComponent, RdsGlobalSearchComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSearchModule, imports: [RdsFormFieldModule, FormsModule, RdsButtonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSearchModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [RdsFormFieldModule, RdsIconComponent, FormsModule, RdsButtonModule],
                    declarations: [...COMPONENTS],
                    exports: [...COMPONENTS],
                }]
        }] });
