// @rds/angular-components/esm2022/lib/search/search.intl.mjs
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
export class RdsSearchIntl {
    constructor() {
        this._changes = new Subject();
        this._searchInputPlaceholder = 'Search...';
        this._searchButtonText = 'Search';
        this._emptyValue = '';
        this._ariaInputLabel = 'Global';
        this._ariaClearLabel = 'Clear search';
    }
    get changes() {
        return this._changes.asObservable();
    }
    get searchInputPlaceholder() {
        return this._searchInputPlaceholder;
    }
    set searchInputPlaceholder(placeholderText) {
        this._searchInputPlaceholder = placeholderText;
        this._changes.next();
    }
    get searchButtonText() {
        return this._searchButtonText;
    }
    set searchButtonText(buttonText) {
        this._searchButtonText = buttonText;
        this._changes.next();
    }
    get emptyValue() {
        return this._emptyValue;
    }
    set emptyValue(emptyValue) {
        this._emptyValue = emptyValue;
        this._changes.next();
    }
    get ariaInputLabel() {
        return this._ariaInputLabel;
    }
    set ariaInputLabel(ariaInputLabel) {
        this._ariaInputLabel = ariaInputLabel;
        this._changes.next();
    }
    get ariaClearLabel() {
        return this._ariaClearLabel;
    }
    set ariaClearLabel(ariaClearLabel) {
        this._ariaClearLabel = ariaClearLabel;
        this._changes.next();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSearchIntl, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSearchIntl, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSearchIntl, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
