// @rds/angular-components/esm2022/lib/paginator/paginator-basic/paginator-basic.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RDS_PAGINATOR_INTL_PROVIDER } from '../paginator-intl';
import { A11yModule } from '@angular/cdk/a11y';
import { RdsIconComponent } from '../../icons';
import { RdsTooltipModule } from '../../tooltip';
import { RdsButtonModule } from '../../button';
import { RdsPaginatorBasicComponent } from './paginator-basic.component';
import * as i0 from "@angular/core";
export class RdsPaginatorBasicModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsPaginatorBasicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsPaginatorBasicModule, declarations: [RdsPaginatorBasicComponent], imports: [CommonModule,
            RdsIconComponent,
            RdsTooltipModule,
            RdsButtonModule,
            A11yModule], exports: [RdsPaginatorBasicComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsPaginatorBasicModule, providers: [RDS_PAGINATOR_INTL_PROVIDER], imports: [CommonModule,
            RdsTooltipModule,
            RdsButtonModule,
            A11yModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsPaginatorBasicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [RdsPaginatorBasicComponent],
                    exports: [RdsPaginatorBasicComponent],
                    imports: [
                        CommonModule,
                        RdsIconComponent,
                        RdsTooltipModule,
                        RdsButtonModule,
                        A11yModule,
                    ],
                    providers: [RDS_PAGINATOR_INTL_PROVIDER],
                }]
        }] });
