// @rds/angular-components/esm2022/lib/autocomplete/autocomplete-intl.service.mjs
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
export class RdsAutocompleteIntl {
    constructor() {
        this._change$ = new BehaviorSubject(RdsAutocompleteIntl.translations);
        this.change$ = this._change$.asObservable();
    }
    static { this.translations = {
        placeholder: 'Select an option',
        loadingText: 'Loading...',
        noDataText: 'No results found',
        newItemText: 'Create new',
    }; }
    setTranslation(key, value) {
        RdsAutocompleteIntl.translations[key] = value;
        this._change$.next({ ...RdsAutocompleteIntl.translations });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAutocompleteIntl, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAutocompleteIntl, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAutocompleteIntl, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
