// @rds/angular-components/esm2022/lib/autocomplete/autocomplete-option.directive.mjs
import { Directive, ElementRef, HostBinding, inject, Input, Renderer2, } from '@angular/core';
import { RdsAbstractAutocompleteComponent } from './abstract-autocomplete.component';
import * as i0 from "@angular/core";
export class RdsAutocompleteOptionDirective {
    constructor() {
        this.nativeElement = inject(ElementRef).nativeElement;
        this.renderer = inject(Renderer2);
        this.class = 'rds-autocomplete-option';
    }
    set keyboardFocusedItem(keyboardFocusedItem) {
        this._keyboardFocusedItem = keyboardFocusedItem;
        if (this === keyboardFocusedItem) {
            this.renderer.addClass(this.nativeElement, 'cdk-focused');
            this.renderer.addClass(this.nativeElement, 'cdk-keyboard-focused');
        }
        else {
            this.renderer.removeClass(this.nativeElement, 'cdk-focused');
            this.renderer.removeClass(this.nativeElement, 'cdk-keyboard-focused');
        }
    }
    get keyboardFocusedItem() {
        return this._keyboardFocusedItem;
    }
    ngOnInit() {
        this.autocomplete._addOption(this.nativeElement, this);
    }
    ngOnDestroy() {
        this.autocomplete._deleteOption(this.nativeElement);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAutocompleteOptionDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsAutocompleteOptionDirective, isStandalone: true, selector: "rds-autocomplete-option, [rds-autocomplete-option], [rdsAutocompleteOption]", inputs: { autocomplete: ["rdsAutocompleteOption", "autocomplete"], keyboardFocusedItem: "keyboardFocusedItem" }, host: { properties: { "class": "this.class" } }, exportAs: ["rdsAutocompleteOption"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAutocompleteOptionDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'rds-autocomplete-option, [rds-autocomplete-option], [rdsAutocompleteOption]',
                    exportAs: 'rdsAutocompleteOption',
                    standalone: true,
                }]
        }], propDecorators: { autocomplete: [{
                type: Input,
                args: [{ required: true, alias: 'rdsAutocompleteOption' }]
            }], keyboardFocusedItem: [{
                type: Input,
                args: [{ required: true }]
            }], class: [{
                type: HostBinding,
                args: ['class']
            }] } });
