// @rds/angular-components/esm2022/lib/autocomplete/autocomplete-highlight.directive.mjs
import { booleanAttribute, Directive, ElementRef, inject, Input, Renderer2, SecurityContext, } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import * as i0 from "@angular/core";
const INVERSE_CLASSNAME = 'rds-autocomplate-highlight-inverse';
export class RdsAutocompleteHighlightDirective {
    constructor() {
        this.nativeElement = inject(ElementRef).nativeElement;
        this.sanitizer = inject(DomSanitizer);
        this.renderer = inject(Renderer2);
        this.caseSensitive = false;
        this.customClasses = '';
        this.inverse = false;
        this.disabled = false;
    }
    ngOnChanges(changes) {
        if (this.nativeElement && changes['inverse']) {
            this.applyInversion();
        }
        if (this.nativeElement &&
            (changes['searchTerm'] || changes['caseSensitive'] || changes['disabled'])) {
            const text = this.nativeElement.textContent || '';
            if (this.searchTerm === '' || this.disabled) {
                this.nativeElement.innerHTML = text;
            }
            else {
                const regex = new RegExp(this.searchTerm, this.caseSensitive ? 'g' : 'gi');
                const newText = text.replace(regex, (match) => `<mark class="rds-autocomplete-highlight${' ' + this.customClasses}">${match}</mark>`);
                this.nativeElement.innerHTML =
                    this.sanitizer.sanitize(SecurityContext.HTML, newText) || '';
            }
        }
    }
    applyInversion() {
        if (this.inverse &&
            !this.nativeElement.classList.contains(INVERSE_CLASSNAME)) {
            this.renderer.addClass(this.nativeElement, INVERSE_CLASSNAME);
        }
        else if (!this.inverse &&
            this.nativeElement.classList.contains(INVERSE_CLASSNAME)) {
            this.renderer.removeClass(this.nativeElement, INVERSE_CLASSNAME);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAutocompleteHighlightDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "16.1.0", version: "17.3.1", type: RdsAutocompleteHighlightDirective, isStandalone: true, selector: "[rdsAutocomplateHighlight], [rds-autocomplate-highlight]", inputs: { searchTerm: ["rdsAutocomplateHighlight", "searchTerm"], caseSensitive: ["rdsAutocomplateHighlightCaseSensitive", "caseSensitive", booleanAttribute], customClasses: ["rdsAutocomplateHighlightCustomClasses", "customClasses"], inverse: ["rdsAutocomplateHighlightInverse", "inverse", booleanAttribute], disabled: ["rdsAutocomplateHighlightDisabled", "disabled", booleanAttribute] }, usesOnChanges: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAutocompleteHighlightDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rdsAutocomplateHighlight], [rds-autocomplate-highlight]',
                    standalone: true,
                }]
        }], propDecorators: { searchTerm: [{
                type: Input,
                args: ['rdsAutocomplateHighlight']
            }], caseSensitive: [{
                type: Input,
                args: [{
                        alias: 'rdsAutocomplateHighlightCaseSensitive',
                        transform: booleanAttribute,
                    }]
            }], customClasses: [{
                type: Input,
                args: ['rdsAutocomplateHighlightCustomClasses']
            }], inverse: [{
                type: Input,
                args: [{
                        alias: 'rdsAutocomplateHighlightInverse',
                        transform: booleanAttribute,
                    }]
            }], disabled: [{
                type: Input,
                args: [{
                        alias: 'rdsAutocomplateHighlightDisabled',
                        transform: booleanAttribute,
                    }]
            }] } });
