// @rds/angular-components/esm2022/lib/button/button.component.mjs
import { ChangeDetectionStrategy, Component, ViewEncapsulation, } from '@angular/core';
import { RdsButtonBaseComponent } from './button-base';
import * as i0 from "@angular/core";
export class RdsButtonComponent extends RdsButtonBaseComponent {
    constructor() {
        super(...arguments);
        this.buttonHostAttributesToClassMapper = {
            'rds-primary-button': 'rds-button--primary',
            'rds-warning-button': 'rds-button--warning',
            'rds-secondary-button': 'rds-button--secondary',
            'rds-secondary-warning-button': 'rds-button--secondary-warning',
            rdsPrimaryButton: 'rds-button--primary',
            rdsWarningButton: 'rds-button--warning',
            rdsSecondaryButton: 'rds-button--secondary',
            rdsSecondaryWarningButton: 'rds-button--secondary-warning',
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsButtonComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsButtonComponent, selector: "\n    button[rds-primary-button],\n    button[rds-warning-button],\n    button[rds-secondary-button],\n    button[rds-secondary-warning-button],\n    button[rdsPrimaryButton],\n    button[rdsWarningButton],\n    button[rdsSecondaryButton],\n    button[rdsSecondaryWarningButton],\n  ", usesInheritance: true, ngImport: i0, template: '<ng-content />', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsButtonComponent, decorators: [{
            type: Component,
            args: [{
                    selector: `
    button[rds-primary-button],
    button[rds-warning-button],
    button[rds-secondary-button],
    button[rds-secondary-warning-button],
    button[rdsPrimaryButton],
    button[rdsWarningButton],
    button[rdsSecondaryButton],
    button[rdsSecondaryWarningButton],
  `,
                    template: '<ng-content />',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                }]
        }] });
