// @rds/angular-components/esm2022/lib/button/button-module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RdsButtonComponent } from './button.component';
import { RdsIconComponent } from '../icons';
import { RdsIconButtonComponent } from './icon-button.component';
import { RdsButtonBaseComponent } from './button-base';
import { RdsFabButtonComponent } from './fab-button.component';
import { RdsTextButtonComponent } from './text-button.component';
import { ObserversModule } from '@angular/cdk/observers';
import * as i0 from "@angular/core";
const ELEMENTS = [
    RdsButtonComponent,
    RdsIconButtonComponent,
    RdsFabButtonComponent,
    RdsTextButtonComponent,
];
export class RdsButtonModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsButtonModule, declarations: [RdsButtonBaseComponent, RdsButtonComponent,
            RdsIconButtonComponent,
            RdsFabButtonComponent,
            RdsTextButtonComponent], imports: [CommonModule, ObserversModule, RdsIconComponent], exports: [RdsButtonComponent,
            RdsIconButtonComponent,
            RdsFabButtonComponent,
            RdsTextButtonComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsButtonModule, imports: [CommonModule, ObserversModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, ObserversModule, RdsIconComponent],
                    declarations: [RdsButtonBaseComponent, ...ELEMENTS],
                    exports: [...ELEMENTS],
                }]
        }] });
