// @rds/angular-components/esm2022/lib/button/fab-button.component.mjs
import { ChangeDetectionStrategy, Component, ViewEncapsulation, } from '@angular/core';
import { RdsButtonBaseComponent } from './button-base';
import * as i0 from "@angular/core";
export class RdsFabButtonComponent extends RdsButtonBaseComponent {
    constructor() {
        super(...arguments);
        this.buttonHostAttributesToClassMapper = {
            'rds-fab-button': 'rds-button--fab',
            rdsFabButton: 'rds-button--fab',
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsFabButtonComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsFabButtonComponent, selector: "\n    button[rds-fab-button],\n    button[rdsFabButton],\n  ", usesInheritance: true, ngImport: i0, template: '<ng-content />', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsFabButtonComponent, decorators: [{
            type: Component,
            args: [{
                    selector: `
    button[rds-fab-button],
    button[rdsFabButton],
  `,
                    template: '<ng-content />',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                }]
        }] });
