// @rds/angular-components/esm2022/lib/button/text-button.component.mjs
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation, } from '@angular/core';
import { RdsButtonBaseComponent } from './button-base';
import { throwRdsButtonInvalidIconOnlyError } from './button-error';
import { throwInvalidSizeError } from '../core/errors';
import * as i0 from "@angular/core";
export class RdsTextButtonComponent extends RdsButtonBaseComponent {
    constructor() {
        super(...arguments);
        this.buttonHostAttributesToClassMapper = {
            'rds-text-button': 'rds-button--text',
            rdsTextButton: 'rds-button--text',
            'rds-text-warning-button': 'rds-button--text-warning',
            rdsTextWarningButton: 'rds-button--text-warning',
        };
    }
    get size() {
        return this._size;
    }
    set size(size) {
        if (size !== 's' && size !== 'm' && size !== 'l') {
            throwInvalidSizeError(this, size);
        }
        this._size = size;
    }
    checkIcons() {
        super.checkIcons();
        if (this._hasOnlyIcon) {
            throwRdsButtonInvalidIconOnlyError(this._getButtonClassName());
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTextButtonComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsTextButtonComponent, selector: "\n    button[rds-text-button],\n    button[rdsTextButton],\n    button[rds-text-warning-button],\n    button[rdsTextWarningButton],\n  ", inputs: { size: "size" }, usesInheritance: true, ngImport: i0, template: '<ng-content />', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTextButtonComponent, decorators: [{
            type: Component,
            args: [{
                    selector: `
    button[rds-text-button],
    button[rdsTextButton],
    button[rds-text-warning-button],
    button[rdsTextWarningButton],
  `,
                    template: '<ng-content />',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                }]
        }], propDecorators: { size: [{
                type: Input
            }] } });
