// @rds/angular-components/esm2022/lib/toast/rds-toast-container.directive.mjs
import { Directive, ElementRef, inject } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsToastContainerDirective {
    constructor() {
        this.elementRef = inject(ElementRef);
    }
    getContainerElement() {
        return this.elementRef.nativeElement;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsToastContainerDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsToastContainerDirective, selector: "[rdsToastContainer]", exportAs: ["rdsToastContainer"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsToastContainerDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rdsToastContainer]',
                    exportAs: 'rdsToastContainer',
                }]
        }] });
