// @rds/angular-components/esm2022/lib/toast/rds-toast.module.mjs
import { CommonModule } from '@angular/common';
import { inject, NgModule } from '@angular/core';
import { RdsToastComponent } from './rds-toast.component';
import { RdsIconComponent } from '../icons';
import { DEFAULT_GLOBAL_CONFIG, RDS_TOAST_GLOBAL_CONFIG_TOKEN, } from './rds-toast.config';
import { RdsToastContainerDirective } from './rds-toast-container.directive';
import { RdsButtonModule } from '../button';
import * as i0 from "@angular/core";
export class RdsToastModule {
    static forRoot(config) {
        return {
            ngModule: RdsToastModule,
            providers: [
                {
                    provide: RDS_TOAST_GLOBAL_CONFIG_TOKEN,
                    useValue: {
                        ...DEFAULT_GLOBAL_CONFIG,
                        ...config,
                    },
                },
            ],
        };
    }
    constructor() {
        this.parentModule = inject(RdsToastModule, {
            optional: true,
            skipSelf: true,
        });
        if (this.parentModule) {
            throw new Error('RdsToastModule is already loaded. Import it in the AppModule only.');
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsToastModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsToastModule, declarations: [RdsToastComponent, RdsToastContainerDirective], imports: [CommonModule, RdsIconComponent, RdsButtonModule], exports: [RdsToastComponent, RdsToastContainerDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsToastModule, imports: [CommonModule, RdsButtonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsToastModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, RdsIconComponent, RdsButtonModule],
                    declarations: [RdsToastComponent, RdsToastContainerDirective],
                    exports: [RdsToastComponent, RdsToastContainerDirective],
                }]
        }], ctorParameters: () => [] });
