// @rds/angular-components/esm2022/lib/toast/rds-toast.ref.mjs
import { Subject } from 'rxjs';
export class RdsToastRef {
    constructor(_rdsToastOverlayRef) {
        this._rdsToastOverlayRef = _rdsToastOverlayRef;
        this._afterClosed = new Subject();
        this._activate = new Subject();
        this._manualClose = new Subject();
        this._isClosed = false;
    }
    manualClose() {
        this._manualClose.next();
        this._manualClose.complete();
    }
    manualClosed() {
        return this._manualClose.asObservable();
    }
    close() {
        this._isClosed = true;
        this._rdsToastOverlayRef.detach();
        this._afterClosed.next();
        this._manualClose.next();
        this._afterClosed.complete();
        this._manualClose.complete();
        this._activate.complete();
    }
    isClosed() {
        return this._isClosed;
    }
    afterClosed() {
        return this._afterClosed.asObservable();
    }
    isInactive() {
        return this._activate.isStopped;
    }
    activate() {
        this._activate.next();
        this._activate.complete();
    }
    afterActivate() {
        return this._activate.asObservable();
    }
}
