// @rds/angular-components/esm2022/lib/carousel/carousel.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RdsCarouselComponent } from './carousel.component';
import { RdsCarouselItemDirective } from './carousel-item.directive';
import { RdsIconComponent } from '../icons';
import { RdsButtonModule } from '../button';
import * as i0 from "@angular/core";
const ELEMENTS = [RdsCarouselComponent, RdsCarouselItemDirective];
export class RdsCarouselModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCarouselModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsCarouselModule, declarations: [RdsCarouselComponent, RdsCarouselItemDirective], imports: [CommonModule, RdsIconComponent, RdsButtonModule], exports: [RdsCarouselComponent, RdsCarouselItemDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCarouselModule, imports: [CommonModule, RdsButtonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCarouselModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, RdsIconComponent, RdsButtonModule],
                    declarations: [...ELEMENTS],
                    exports: [...ELEMENTS],
                }]
        }] });
