// @rds/angular-components/esm2022/lib/carousel/carousel-item.directive.mjs
import { ChangeDetectorRef, Directive, HostBinding, inject, } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsCarouselItemDirective {
    constructor() {
        this.cdr = inject(ChangeDetectorRef);
        this.class = true;
        this._shown = false;
    }
    _hide() {
        if (this._shown) {
            this._shown = false;
            this.cdr.markForCheck();
        }
    }
    _show() {
        if (!this._shown) {
            this._shown = true;
            this.cdr.markForCheck();
        }
    }
    isShown() {
        return this._shown;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCarouselItemDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsCarouselItemDirective, selector: "rds-carousel-item", host: { properties: { "class.rds-carousel__item": "this.class", "class.rds-carousel__item--visible": "this._shown" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCarouselItemDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'rds-carousel-item',
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class.rds-carousel__item']
            }], _shown: [{
                type: HostBinding,
                args: ['class.rds-carousel__item--visible']
            }] } });
