// @rds/angular-components/esm2022/lib/toggle-button/toggle-button.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RdsToggleButtonComponent } from './toggle-button.component';
import { RdsToggleDirective } from './toggle.directive';
import * as i0 from "@angular/core";
export class RdsToggleButtonModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsToggleButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsToggleButtonModule, declarations: [RdsToggleButtonComponent, RdsToggleDirective], imports: [CommonModule], exports: [RdsToggleButtonComponent, RdsToggleDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsToggleButtonModule, imports: [CommonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsToggleButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [RdsToggleButtonComponent, RdsToggleDirective],
                    exports: [RdsToggleButtonComponent, RdsToggleDirective],
                }]
        }] });
