// @rds/angular-components/esm2022/lib/toggle-button/toggle.directive.mjs
import { booleanAttribute, ContentChildren, DestroyRef, Directive, ErrorHandler, EventEmitter, HostBinding, inject, Input, Output, QueryList, } from '@angular/core';
import { asyncScheduler, merge, Subject } from 'rxjs';
import { delay, filter, takeUntil } from 'rxjs/operators';
import { RdsToggleButtonComponent } from './toggle-button.component';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import * as i0 from "@angular/core";
export class RdsToggleDirective {
    constructor() {
        this.destroyRef = inject(DestroyRef);
        this.errorHandler = inject(ErrorHandler);
        this._changedDestroy$ = new Subject();
        this.type = 'basic';
        this.hostRole = 'group';
        this._disabled = false;
        this.valueChange = new EventEmitter();
    }
    get hostClass() {
        return ['rds-toggle', `rds-toggle--${this.type}`];
    }
    set disabled(value) {
        this._disabled = booleanAttribute(value);
        if (this.buttonToggles) {
            this.buttonToggles.forEach((item) => (item.disabled = this._disabled));
        }
    }
    get disabled() {
        return this._disabled;
    }
    ngAfterContentInit() {
        Promise.resolve().then(() => {
            const valueItem = this.buttonToggles.find((b) => b.value === this.value);
            if (valueItem) {
                this.selected = valueItem;
                valueItem.checked = true;
                const checkedButtons = this.buttonToggles.filter((b) => b.checked && b.value !== this.value);
                if (checkedButtons.length > 0) {
                    checkedButtons.forEach((b) => (b.checked = false));
                }
            }
            else {
                const checkedToggles = this.buttonToggles.filter((b) => b.checked);
                if (checkedToggles.length > 1) {
                    this.errorHandler.handleError(new Error('Invalid status, more than one toggle is checked'));
                }
                const firstCheckedToggle = checkedToggles[0];
                if (firstCheckedToggle) {
                    firstCheckedToggle.checked = true;
                    this.rawValue = firstCheckedToggle.value;
                    this.selected = firstCheckedToggle;
                }
                if (this.disabled) {
                    this.buttonToggles.forEach((item) => (item.disabled = this.disabled));
                }
            }
            this.refreshSeparators();
            this.refreshChangeListener();
        });
        this.buttonToggles.changes
            .pipe(delay(0, asyncScheduler), takeUntilDestroyed(this.destroyRef))
            .subscribe(() => {
            this.refreshChangeListener();
            this.refreshSeparators();
        });
    }
    ngOnChanges(changes) {
        if (changes['value'] &&
            changes['value'].previousValue !== changes['value'].currentValue &&
            this.buttonToggles &&
            this.buttonToggles.some((bc) => bc.value === this.value)) {
            this.buttonToggles.find((bc) => bc.value === this.value)?.onButtonClick();
        }
        this.refreshSeparators();
    }
    ngOnDestroy() {
        this._changedDestroy$.next();
        this._changedDestroy$.complete();
    }
    handleToggleButtonChange(toggleChangedEvent) {
        const { buttonToggle, checked } = toggleChangedEvent;
        if (this.selected === buttonToggle) {
            return; // cannot deselect
        }
        if (this.selected) {
            this.selected.checked = false;
        }
        buttonToggle.checked = checked;
        this.rawValue = buttonToggle.value;
        this.selected = buttonToggle;
    }
    refreshSeparators() {
        if (this.buttonToggles && this.type === 'outlined') {
            this.buttonToggles.forEach((item, index, array) => {
                if (item.disabled) {
                    item.showSeparator = true;
                }
                else if (index === array.length - 1) {
                    item.showSeparator = false;
                }
                else if (index !== array.length - 1) {
                    item.showSeparator = !item.checked;
                }
                else if (item.showSeparator) {
                    item.showSeparator = false;
                }
                if (item.checked && index > 0) {
                    array[index - 1].showSeparator = false;
                }
            });
        }
        else if (this.buttonToggles && this.type === 'basic') {
            this.buttonToggles.forEach((item, index, array) => {
                item.showSeparator = index !== array.length - 1;
            });
        }
    }
    refreshChangeListener() {
        this._changedDestroy$.next();
        merge(...this.buttonToggles.map((toggle) => toggle.changed))
            .pipe(filter((event) => event.checked), takeUntil(this._changedDestroy$))
            .subscribe((toggleChangedEvent) => {
            this.handleToggleButtonChange(toggleChangedEvent);
            this.valueChange.emit(this.rawValue);
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsToggleDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsToggleDirective, selector: "rds-toggle", inputs: { type: "type", disabled: "disabled", value: "value" }, outputs: { valueChange: "valueChange" }, host: { properties: { "attr.role": "this.hostRole", "class": "this.hostClass", "attr.aria-disabled": "this.disabled", "class.rds-toggle--disabled": "this.disabled" } }, queries: [{ propertyName: "buttonToggles", predicate: RdsToggleButtonComponent }], usesOnChanges: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsToggleDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'rds-toggle',
                }]
        }], propDecorators: { type: [{
                type: Input
            }], hostRole: [{
                type: HostBinding,
                args: ['attr.role']
            }], hostClass: [{
                type: HostBinding,
                args: ['class']
            }], disabled: [{
                type: HostBinding,
                args: ['attr.aria-disabled']
            }, {
                type: HostBinding,
                args: ['class.rds-toggle--disabled']
            }, {
                type: Input
            }], value: [{
                type: Input
            }], valueChange: [{
                type: Output
            }], buttonToggles: [{
                type: ContentChildren,
                args: [RdsToggleButtonComponent]
            }] } });
