// @rds/angular-components/esm2022/lib/empty-state/empty-state.module.mjs
import { ObserversModule } from '@angular/cdk/observers';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RdsDividerModule } from '../divider';
import { RdsEmptyStateButtonDirective } from './empty-state-button.directive';
import { RdsEmptyStateIconDirective } from './empty-state-icon.directive';
import { RdsEmptyStateImageDirective } from './empty-state-image.directive';
import { RdsEmptyStateSubtitleDirective } from './empty-state-subtitle.directive';
import { RdsEmptyStateTitleDirective } from './empty-state-title.directive';
import { RdsEmptyStateComponent } from './empty-state.component';
import * as i0 from "@angular/core";
const ELEMENTS = [
    RdsEmptyStateComponent,
    RdsEmptyStateButtonDirective,
    RdsEmptyStateImageDirective,
    RdsEmptyStateSubtitleDirective,
    RdsEmptyStateTitleDirective,
    RdsEmptyStateIconDirective,
];
export class RdsEmptyStateModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsEmptyStateModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsEmptyStateModule, declarations: [RdsEmptyStateComponent,
            RdsEmptyStateButtonDirective,
            RdsEmptyStateImageDirective,
            RdsEmptyStateSubtitleDirective,
            RdsEmptyStateTitleDirective,
            RdsEmptyStateIconDirective], imports: [CommonModule, ObserversModule, RdsDividerModule], exports: [RdsEmptyStateComponent,
            RdsEmptyStateButtonDirective,
            RdsEmptyStateImageDirective,
            RdsEmptyStateSubtitleDirective,
            RdsEmptyStateTitleDirective,
            RdsEmptyStateIconDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsEmptyStateModule, imports: [CommonModule, ObserversModule, RdsDividerModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsEmptyStateModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [...ELEMENTS],
                    exports: [...ELEMENTS],
                    imports: [CommonModule, ObserversModule, RdsDividerModule],
                }]
        }] });
