// @rds/angular-components/esm2022/lib/sidenav/sidenav.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RdsSidenavComponent } from './sidenav.component';
import { RdsIconComponent } from '../icons';
import { RdsButtonModule } from '../button';
import { RdsSidenavFooterComponent } from './sidenav-footer/sidenav-footer.component';
import { RdsSidenavMenuitemComponent } from './sidenav-menuitem/sidenav-menuitem.component';
import { RdsSidenavMenuComponent } from './sidenav-menu/sidenav-menu.component';
import { RdsSidenavMenuTriggerComponent } from './sidenav-menuitem-trigger/sidenav-menu-trigger.component';
import { RdsSidenavGroupComponent } from './sidenav-group/sidenav-group.component';
import { RdsSidenavGroupTitleDirective } from './sidenav-group/sidenav-group-title.directive';
import { OverlayModule } from '@angular/cdk/overlay';
import { RdsDividerModule } from '../divider';
import { RdsObserveResizeDirective } from '../utils';
import { RdsSidenavMenuHeaderComponent } from './sidenav-menu-header/sidenav-menu-header.component';
import * as i0 from "@angular/core";
const PUBLIC_SIDENAV_ELEMENTS = [
    RdsSidenavComponent,
    RdsSidenavMenuComponent,
    RdsSidenavMenuTriggerComponent,
    RdsSidenavGroupComponent,
    RdsSidenavGroupTitleDirective,
    RdsSidenavMenuitemComponent,
    RdsSidenavMenuHeaderComponent,
];
export class RdsSidenavModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSidenavModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsSidenavModule, declarations: [RdsSidenavComponent,
            RdsSidenavMenuComponent,
            RdsSidenavMenuTriggerComponent,
            RdsSidenavGroupComponent,
            RdsSidenavGroupTitleDirective,
            RdsSidenavMenuitemComponent,
            RdsSidenavMenuHeaderComponent, RdsSidenavFooterComponent], imports: [CommonModule,
            RdsIconComponent,
            RdsButtonModule,
            OverlayModule,
            RdsDividerModule,
            RdsObserveResizeDirective], exports: [RdsSidenavComponent,
            RdsSidenavMenuComponent,
            RdsSidenavMenuTriggerComponent,
            RdsSidenavGroupComponent,
            RdsSidenavGroupTitleDirective,
            RdsSidenavMenuitemComponent,
            RdsSidenavMenuHeaderComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSidenavModule, imports: [CommonModule,
            RdsButtonModule,
            OverlayModule,
            RdsDividerModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSidenavModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...PUBLIC_SIDENAV_ELEMENTS],
                    declarations: [...PUBLIC_SIDENAV_ELEMENTS, RdsSidenavFooterComponent],
                    imports: [
                        CommonModule,
                        RdsIconComponent,
                        RdsButtonModule,
                        OverlayModule,
                        RdsDividerModule,
                        RdsObserveResizeDirective,
                    ],
                }]
        }] });
