// @rds/angular-components/esm2022/lib/popover/popover.module.mjs
import { RdsTooltipModule } from '../tooltip';
import { CommonModule } from '@angular/common';
import { RdsPopoverDirective } from './popover.directive';
import { NgModule } from '@angular/core';
import { RDS_POPOVER_SCROLL_STRATEGY_FACTORY_PROVIDER } from './popover.tokens';
import { RdsPopoverHeaderDirective } from './popover-text-content/popover-header.directive';
import { RdsPopoverParagraphDirective } from './popover-text-content/popover-paragraph.directive';
import { RdsPopoverTextContentComponent } from './popover-text-content/popover-text-content.component';
import * as i0 from "@angular/core";
export class RdsPopoverModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsPopoverModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsPopoverModule, declarations: [RdsPopoverDirective,
            RdsPopoverHeaderDirective,
            RdsPopoverParagraphDirective,
            RdsPopoverTextContentComponent], imports: [CommonModule, RdsTooltipModule], exports: [RdsPopoverDirective,
            RdsTooltipModule,
            RdsPopoverHeaderDirective,
            RdsPopoverParagraphDirective,
            RdsPopoverTextContentComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsPopoverModule, providers: [RDS_POPOVER_SCROLL_STRATEGY_FACTORY_PROVIDER], imports: [CommonModule, RdsTooltipModule, RdsTooltipModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsPopoverModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, RdsTooltipModule],
                    declarations: [
                        RdsPopoverDirective,
                        RdsPopoverHeaderDirective,
                        RdsPopoverParagraphDirective,
                        RdsPopoverTextContentComponent,
                    ],
                    exports: [
                        RdsPopoverDirective,
                        RdsTooltipModule,
                        RdsPopoverHeaderDirective,
                        RdsPopoverParagraphDirective,
                        RdsPopoverTextContentComponent,
                    ],
                    providers: [RDS_POPOVER_SCROLL_STRATEGY_FACTORY_PROVIDER],
                }]
        }] });
