// @rds/angular-components/esm2022/lib/stepper/step.component.mjs
import { CdkStep, STEPPER_GLOBAL_OPTIONS } from '@angular/cdk/stepper';
import { ChangeDetectionStrategy, Component, ContentChild, forwardRef, inject, ViewContainerRef, ViewEncapsulation, } from '@angular/core';
import { TemplatePortal } from '@angular/cdk/portal';
import { map, startWith, switchMap } from 'rxjs/operators';
import { Subscription } from 'rxjs';
import { RdsStepLabelDirective } from './step-label.directive';
import { RdsStepContentDirective } from './step-content.directive';
import { RdsStepperComponent } from './stepper.component';
import { ErrorStateMatcher } from '../core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/portal";
export class RdsStepComponent extends CdkStep {
    constructor() {
        super(inject(forwardRef(() => RdsStepperComponent)), inject(STEPPER_GLOBAL_OPTIONS, { optional: true }) || undefined);
        this.errorStateMatcher = inject(ErrorStateMatcher, {
            skipSelf: true,
        });
        this.viewContainerRef = inject(ViewContainerRef, {
            skipSelf: true,
        });
        this._isSelected = Subscription.EMPTY;
    }
    ngAfterContentInit() {
        this._isSelected = this._stepper.steps.changes
            .pipe(switchMap(() => {
            return this._stepper.selectionChange.pipe(map((event) => event.selectedStep === this), startWith(this._stepper.selected === this));
        }))
            .subscribe((isSelected) => {
            if (isSelected && this._lazyContent && !this._portal) {
                this._portal = new TemplatePortal(this._lazyContent._template, this.viewContainerRef);
            }
        });
    }
    ngOnDestroy() {
        this._isSelected.unsubscribe();
    }
    /** Custom error state matcher that additionally checks for validity of interacted form. */
    isErrorState(control, form) {
        const originalErrorState = this.errorStateMatcher.isErrorState(control, form);
        // Custom error state checks for the validity of form that is not submitted or touched
        // since user can trigger a form change by calling for another step without directly
        // interacting with the current form.
        const customErrorState = !!(control && control.invalid && this.interacted);
        return originalErrorState || customErrorState;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsStepComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsStepComponent, selector: "rds-step", providers: [
            { provide: ErrorStateMatcher, useExisting: RdsStepComponent },
            { provide: CdkStep, useExisting: RdsStepComponent },
        ], queries: [{ propertyName: "stepLabel", first: true, predicate: RdsStepLabelDirective, descendants: true }, { propertyName: "_lazyContent", first: true, predicate: RdsStepContentDirective, descendants: true }], exportAs: ["rdsStep"], usesInheritance: true, ngImport: i0, template: ` <ng-template>
    <ng-content />
    <ng-template [cdkPortalOutlet]="_portal" />
  </ng-template>`, isInline: true, dependencies: [{ kind: "directive", type: i1.CdkPortalOutlet, selector: "[cdkPortalOutlet]", inputs: ["cdkPortalOutlet"], outputs: ["attached"], exportAs: ["cdkPortalOutlet"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsStepComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'rds-step',
                    template: ` <ng-template>
    <ng-content />
    <ng-template [cdkPortalOutlet]="_portal" />
  </ng-template>`,
                    providers: [
                        { provide: ErrorStateMatcher, useExisting: RdsStepComponent },
                        { provide: CdkStep, useExisting: RdsStepComponent },
                    ],
                    encapsulation: ViewEncapsulation.None,
                    exportAs: 'rdsStep',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], ctorParameters: () => [], propDecorators: { stepLabel: [{
                type: ContentChild,
                args: [RdsStepLabelDirective]
            }], _lazyContent: [{
                type: ContentChild,
                args: [RdsStepContentDirective, { static: false }]
            }] } });
