// @rds/angular-components/esm2022/lib/stepper/stepper.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CdkStepperModule } from '@angular/cdk/stepper';
import { RdsStepperNextDirective } from './stepper-next.directive';
import { RdsStepperPreviousDirective } from './stepper-previous.directive';
import { RdsStepContentDirective } from './step-content.directive';
import { RdsStepHeaderComponent } from './step-header.component';
import { RdsStepLabelDirective } from './step-label.directive';
import { RdsStepComponent } from './step.component';
import { RdsStepperComponent } from './stepper.component';
import { PortalModule } from '@angular/cdk/portal';
import * as i0 from "@angular/core";
export class RdsStepperModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsStepperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsStepperModule, declarations: [RdsStepperNextDirective,
            RdsStepperPreviousDirective,
            RdsStepContentDirective,
            RdsStepHeaderComponent,
            RdsStepLabelDirective,
            RdsStepComponent,
            RdsStepperComponent], imports: [CommonModule, PortalModule, CdkStepperModule], exports: [RdsStepperNextDirective,
            RdsStepperPreviousDirective,
            RdsStepContentDirective,
            RdsStepHeaderComponent,
            RdsStepLabelDirective,
            RdsStepComponent,
            RdsStepperComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsStepperModule, imports: [CommonModule, PortalModule, CdkStepperModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsStepperModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, PortalModule, CdkStepperModule],
                    declarations: [
                        RdsStepperNextDirective,
                        RdsStepperPreviousDirective,
                        RdsStepContentDirective,
                        RdsStepHeaderComponent,
                        RdsStepLabelDirective,
                        RdsStepComponent,
                        RdsStepperComponent,
                    ],
                    exports: [
                        RdsStepperNextDirective,
                        RdsStepperPreviousDirective,
                        RdsStepContentDirective,
                        RdsStepHeaderComponent,
                        RdsStepLabelDirective,
                        RdsStepComponent,
                        RdsStepperComponent,
                    ],
                }]
        }] });
