// @rds/angular-components/esm2022/lib/chips/chip-edit-input.mjs
import { Directive, ElementRef, HostBinding, inject } from '@angular/core';
import { DOCUMENT } from '@angular/common';
import { TAB_INDEX_NOT_FOCUSABLE } from '../utils';
import * as i0 from "@angular/core";
export class RdsChipEditInputDirective {
    constructor() {
        this.elementRef = inject(ElementRef);
        this.document = inject(DOCUMENT);
        this.classes = 'rds-chip-edit-input';
        this.role = 'textbox';
        this.tabindex = TAB_INDEX_NOT_FOCUSABLE;
        this.contenteditable = true;
    }
    initialize(initialValue) {
        this.getNativeElement().focus();
        this.setValue(initialValue);
    }
    getNativeElement() {
        return this.elementRef.nativeElement;
    }
    setValue(value) {
        this.getNativeElement().textContent = value;
        this.moveCursorToEndOfInput();
    }
    getValue() {
        return this.getNativeElement().textContent || '';
    }
    moveCursorToEndOfInput() {
        const range = this.document.createRange();
        range.selectNodeContents(this.getNativeElement());
        range.collapse(false);
        const sel = window.getSelection();
        if (sel) {
            sel.removeAllRanges();
            sel.addRange(range);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsChipEditInputDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsChipEditInputDirective, selector: "span[rdsChipEditInput], span[rds-chip-edit-input]", host: { properties: { "class": "this.classes", "attr.role": "this.role", "tabIndex": "this.tabindex", "attr.contenteditable": "this.contenteditable" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsChipEditInputDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'span[rdsChipEditInput], span[rds-chip-edit-input]',
                }]
        }], propDecorators: { classes: [{
                type: HostBinding,
                args: ['class']
            }], role: [{
                type: HostBinding,
                args: ['attr.role']
            }], tabindex: [{
                type: HostBinding,
                args: ['tabIndex']
            }], contenteditable: [{
                type: HostBinding,
                args: ['attr.contenteditable']
            }] } });
