// @rds/angular-components/esm2022/lib/chips/chip-avatar.directive.mjs
import { Directive, HostBinding, inject } from '@angular/core';
import { RdsAvatarComponent } from '../avatar';
import * as i0 from "@angular/core";
export class RdsChipAvatarDirective {
    constructor() {
        this.avatar = inject(RdsAvatarComponent, {
            host: true,
            self: true,
            optional: true,
        });
        this.class = 'rds-chip__avatar';
        if (this.avatar) {
            this.avatar.disableInteractions = true;
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsChipAvatarDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsChipAvatarDirective, selector: "\n  img[rds-chip-avatar], img[rdsChipAvatar],\n  svg[rds-chip-avatar], svg[rdsChipAvatar],\n  rds-avatar[rds-chip-avatar], rds-avatar[rdsChipAvatar],\n  ", host: { properties: { "class": "this.class" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsChipAvatarDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: `
  img[rds-chip-avatar], img[rdsChipAvatar],
  svg[rds-chip-avatar], svg[rdsChipAvatar],
  rds-avatar[rds-chip-avatar], rds-avatar[rdsChipAvatar],
  `,
                }]
        }], ctorParameters: () => [], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }] } });
