// @rds/angular-components/esm2022/lib/chips/chip-action.mjs
import { booleanAttribute, Directive, ElementRef, HostBinding, inject, Input, numberAttribute, } from '@angular/core';
import { TAB_INDEX_NOT_FOCUSABLE } from '../utils';
import * as i0 from "@angular/core";
export class RdsChipActionDirective {
    constructor() {
        this.nativeElement = inject(ElementRef).nativeElement;
        this._tabIndex = TAB_INDEX_NOT_FOCUSABLE;
        this.classes = 'rds-chip-action';
        this.disabled = false;
    }
    set tabIndex(value) {
        this._tabIndex = numberAttribute(value, TAB_INDEX_NOT_FOCUSABLE);
    }
    get tabIndex() {
        return this.disabled ? TAB_INDEX_NOT_FOCUSABLE : this._tabIndex;
    }
    get disabledAttribute() {
        // When this chip action is disabled and focusing disabled chips is not permitted, return empty
        // string to indicate that disabled attribute should be included.
        return this.disabled ? '' : null;
    }
    get tabindexAttribute() {
        return this.disabled ? null : this.tabIndex.toString();
    }
    focus() {
        this.nativeElement.focus();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsChipActionDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "16.1.0", version: "17.3.1", type: RdsChipActionDirective, selector: "[rds-chip-action], [rdsChipAction]", inputs: { tabIndex: "tabIndex", disabled: ["disabled", "disabled", booleanAttribute] }, host: { properties: { "class": "this.classes", "attr.aria-disabled": "this.disabled", "attr.disabled": "this.disabledAttribute", "tabIndex": "this.tabindexAttribute" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsChipActionDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rds-chip-action], [rdsChipAction]',
                }]
        }], propDecorators: { classes: [{
                type: HostBinding,
                args: ['class']
            }], tabIndex: [{
                type: Input
            }], disabled: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }, {
                type: HostBinding,
                args: ['attr.aria-disabled']
            }], disabledAttribute: [{
                type: HostBinding,
                args: ['attr.disabled']
            }], tabindexAttribute: [{
                type: HostBinding,
                args: ['tabIndex']
            }] } });
