// @rds/angular-components/esm2022/lib/chips/chips.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RdsIconComponent } from '../icons';
import { RdsChipAvatarDirective } from './chip-avatar.directive';
import { RdsChipLeadingIconDirective } from './chip-leading-icon.directive';
import { RdsChipListComponent } from './chip-list/chip-list.component';
import { RdsChipComponent } from './chip/chip.component';
import { RDS_AVATAR_DIRECTIVES } from '../avatar';
import { RdsButtonModule } from '../button';
import { RdsChipSelectableComponent } from './chip-selectable/chip-selectable.component';
import { RdsChipSetComponent } from './chip-set';
import { RdsChipActionDirective } from './chip-action';
import { RdsChipEditInputDirective } from './chip-edit-input';
import { RDS_CHIPS_DEFAULT_OPTIONS } from './chip.tokens';
import * as i0 from "@angular/core";
import * as i1 from "../avatar/avatar.component";
import * as i2 from "../avatar/avatar-description.directive";
import * as i3 from "../avatar/avatar-group.component";
import * as i4 from "../avatar/avatar-image.directive";
import * as i5 from "../avatar/avatar-icon.directive";
import * as i6 from "../avatar/avatar-initials.directive";
import * as i7 from "../avatar/avatar-label.directive";
const CHIP_DECLARATIONS = [
    RdsChipListComponent,
    RdsChipComponent,
    RdsChipLeadingIconDirective,
    RdsChipAvatarDirective,
    RdsChipSelectableComponent,
    RdsChipSelectableComponent,
    RdsChipActionDirective,
];
const PRIVATE_DECLARATIONS = [RdsChipEditInputDirective, RdsChipSetComponent];
export class RdsChipsModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsChipsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsChipsModule, declarations: [RdsChipListComponent,
            RdsChipComponent,
            RdsChipLeadingIconDirective,
            RdsChipAvatarDirective,
            RdsChipSelectableComponent,
            RdsChipSelectableComponent,
            RdsChipActionDirective, RdsChipEditInputDirective, RdsChipSetComponent], imports: [CommonModule,
            RdsIconComponent,
            RdsButtonModule, i1.RdsAvatarComponent, i2.RdsAvatarDescriptionDirective, i3.RdsAvatarGroupComponent, i4.RdsAvatarImageDirective, i5.RdsAvatarIconDirective, i6.RdsAvatarInitialsDirective, i7.RdsAvatarLabelDirective], exports: [RdsChipListComponent,
            RdsChipComponent,
            RdsChipLeadingIconDirective,
            RdsChipAvatarDirective,
            RdsChipSelectableComponent,
            RdsChipSelectableComponent,
            RdsChipActionDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsChipsModule, providers: [
            {
                provide: RDS_CHIPS_DEFAULT_OPTIONS,
                useValue: {
                    separatorKeys: ['Enter', ','],
                },
            },
        ], imports: [CommonModule,
            RdsButtonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsChipsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        RdsIconComponent,
                        RdsButtonModule,
                        ...RDS_AVATAR_DIRECTIVES,
                    ],
                    declarations: [...CHIP_DECLARATIONS, ...PRIVATE_DECLARATIONS],
                    exports: [...CHIP_DECLARATIONS],
                    providers: [
                        {
                            provide: RDS_CHIPS_DEFAULT_OPTIONS,
                            useValue: {
                                separatorKeys: ['Enter', ','],
                            },
                        },
                    ],
                }]
        }] });
