// @rds/angular-components/esm2022/lib/anchor-menu/anchor-menu-item/anchor-menu-item.component.mjs
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, DestroyRef, HostBinding, inject, Input, ViewEncapsulation, } from '@angular/core';
import { filter, takeUntil } from 'rxjs/operators';
import { Subject } from 'rxjs';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import * as i0 from "@angular/core";
export class RdsAnchorMenuItemComponent {
    constructor() {
        this.cdr = inject(ChangeDetectorRef);
        this.destroyRef = inject(DestroyRef);
        this.classes = 'rds-anchor-menu-item';
        this.isActive = false;
        this.killItemSubscription$ = new Subject();
    }
    ngOnInit() {
        this.subscribeItem();
    }
    ngOnChanges(changes) {
        if (changes['item'] && !changes['item'].isFirstChange()) {
            this.killItemSubscription$.next();
            this.subscribeItem();
        }
    }
    ngOnDestroy() {
        this.killItemSubscription$.complete();
    }
    subscribeItem() {
        this.item?.isActive$
            .pipe(filter((v) => v !== this.isActive), takeUntil(this.killItemSubscription$), takeUntilDestroyed(this.destroyRef))
            .subscribe((v) => {
            this.isActive = v;
            this.cdr.markForCheck();
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAnchorMenuItemComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsAnchorMenuItemComponent, selector: "a[rdsAnchorMenuItem], a[rds-anchor-menu-item]", inputs: { item: ["rdsAnchorMenuItem", "item"] }, host: { properties: { "class": "this.classes", "class.rds-anchor-menu-item-active": "this.isActive" } }, usesOnChanges: true, ngImport: i0, template: "<ng-content />\n", styles: [".rds-anchor-menu-item{--rds-anchor-menu-item-shadow-color: var(--rds-color-grey-1);font-size:14px;font-weight:400;line-height:20px;text-transform:none;letter-spacing:.3px;font-family:roche-sans;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;padding:8px 16px;box-shadow:inset 1px 0 0 0 var(--rds-anchor-menu-item-shadow-color);text-decoration:none;color:var(--rds-color-low-contrast)}.rds-anchor-menu-item:hover{--rds-anchor-menu-item-shadow-color: var(--rds-color-informative);color:var(--rds-color-high-contrast)}.rds-anchor-menu-item:focus-visible{border-radius:2px;outline:2px solid var(--rds-color-stroke-active-focus);outline-offset:-2px}.rds-anchor-menu-item-active,.rds-anchor-menu-item-active:hover{--rds-anchor-menu-item-shadow-color: var(--rds-color-informative);color:var(--rds-color-informative)}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAnchorMenuItemComponent, decorators: [{
            type: Component,
            args: [{ selector: 'a[rdsAnchorMenuItem], a[rds-anchor-menu-item]', changeDetection: ChangeDetectionStrategy.OnPush, encapsulation: ViewEncapsulation.None, template: "<ng-content />\n", styles: [".rds-anchor-menu-item{--rds-anchor-menu-item-shadow-color: var(--rds-color-grey-1);font-size:14px;font-weight:400;line-height:20px;text-transform:none;letter-spacing:.3px;font-family:roche-sans;-webkit-font-smoothing:antialiased;-moz-osx-font-smoothing:grayscale;padding:8px 16px;box-shadow:inset 1px 0 0 0 var(--rds-anchor-menu-item-shadow-color);text-decoration:none;color:var(--rds-color-low-contrast)}.rds-anchor-menu-item:hover{--rds-anchor-menu-item-shadow-color: var(--rds-color-informative);color:var(--rds-color-high-contrast)}.rds-anchor-menu-item:focus-visible{border-radius:2px;outline:2px solid var(--rds-color-stroke-active-focus);outline-offset:-2px}.rds-anchor-menu-item-active,.rds-anchor-menu-item-active:hover{--rds-anchor-menu-item-shadow-color: var(--rds-color-informative);color:var(--rds-color-informative)}\n"] }]
        }], propDecorators: { item: [{
                type: Input,
                args: ['rdsAnchorMenuItem']
            }], classes: [{
                type: HostBinding,
                args: ['class']
            }], isActive: [{
                type: HostBinding,
                args: ['class.rds-anchor-menu-item-active']
            }] } });
