// @rds/angular-components/esm2022/lib/radio-button/radio-button.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RdsRadioButtonComponent } from './radio-button.component';
import { RdsRadioGroupDirective } from './radio-button-group.directive';
import { RdsHelperTextModule } from '../form-field/helper-text.module';
import * as i0 from "@angular/core";
export class RdsRadioButtonModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsRadioButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsRadioButtonModule, declarations: [RdsRadioButtonComponent, RdsRadioGroupDirective], imports: [CommonModule, RdsHelperTextModule], exports: [RdsRadioButtonComponent, RdsRadioGroupDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsRadioButtonModule, imports: [CommonModule, RdsHelperTextModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsRadioButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, RdsHelperTextModule],
                    declarations: [RdsRadioButtonComponent, RdsRadioGroupDirective],
                    exports: [RdsRadioButtonComponent, RdsRadioGroupDirective],
                }]
        }] });
