// @rds/angular-components/esm2022/lib/avatar/avatar-label.directive.mjs
import { Directive, HostBinding } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsAvatarLabelDirective {
    constructor() {
        this.class = 'rds-avatar-label';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAvatarLabelDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsAvatarLabelDirective, isStandalone: true, selector: "rds-avatar-label, [rds-avatar-label], [rdsAvatarLabel]", host: { properties: { "class": "this.class" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAvatarLabelDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'rds-avatar-label, [rds-avatar-label], [rdsAvatarLabel]',
                    standalone: true,
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }] } });
