// @rds/angular-components/esm2022/lib/avatar/avatar-image.directive.mjs
import { Directive, HostBinding } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsAvatarImageDirective {
    constructor() {
        this.class = 'rds-avatar-image';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAvatarImageDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsAvatarImageDirective, isStandalone: true, selector: "img[rds-avatar-image], img[rdsAvatarImage]", host: { properties: { "class": "this.class" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAvatarImageDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'img[rds-avatar-image], img[rdsAvatarImage]',
                    standalone: true,
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }] } });
