// @rds/angular-components/esm2022/lib/avatar/avatar.module.mjs
import { NgModule } from '@angular/core';
import { RdsAvatarDescriptionDirective } from './avatar-description.directive';
import { RdsAvatarGroupComponent } from './avatar-group.component';
import { RdsAvatarIconDirective } from './avatar-icon.directive';
import { RdsAvatarImageDirective } from './avatar-image.directive';
import { RdsAvatarInitialsDirective } from './avatar-initials.directive';
import { RdsAvatarLabelDirective } from './avatar-label.directive';
import { RdsAvatarComponent } from './avatar.component';
import * as i0 from "@angular/core";
export const RDS_AVATAR_DIRECTIVES = [
    RdsAvatarComponent,
    RdsAvatarDescriptionDirective,
    RdsAvatarGroupComponent,
    RdsAvatarImageDirective,
    RdsAvatarIconDirective,
    RdsAvatarInitialsDirective,
    RdsAvatarLabelDirective,
];
export class RdsAvatarModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAvatarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsAvatarModule, imports: [RdsAvatarComponent,
            RdsAvatarDescriptionDirective,
            RdsAvatarGroupComponent,
            RdsAvatarImageDirective,
            RdsAvatarIconDirective,
            RdsAvatarInitialsDirective,
            RdsAvatarLabelDirective], exports: [RdsAvatarComponent,
            RdsAvatarDescriptionDirective,
            RdsAvatarGroupComponent,
            RdsAvatarImageDirective,
            RdsAvatarIconDirective,
            RdsAvatarInitialsDirective,
            RdsAvatarLabelDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAvatarModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAvatarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...RDS_AVATAR_DIRECTIVES],
                    exports: [...RDS_AVATAR_DIRECTIVES],
                }]
        }] });
