// @rds/angular-components/esm2022/lib/avatar/avatar-initials.directive.mjs
import { Directive, HostBinding } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsAvatarInitialsDirective {
    constructor() {
        this.class = 'rds-avatar-initials';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAvatarInitialsDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsAvatarInitialsDirective, isStandalone: true, selector: "[rds-avatar-initials], [rdsAvatarInitials]", host: { properties: { "class": "this.class" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAvatarInitialsDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rds-avatar-initials], [rdsAvatarInitials]',
                    standalone: true,
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }] } });
