// @rds/angular-components/esm2022/lib/avatar/avatar-description.directive.mjs
import { Directive, HostBinding } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsAvatarDescriptionDirective {
    constructor() {
        this.class = 'rds-avatar-description';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAvatarDescriptionDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsAvatarDescriptionDirective, isStandalone: true, selector: "rds-avatar-description, [rds-avatar-description], [rdsAvatarDescription]", host: { properties: { "class": "this.class" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAvatarDescriptionDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'rds-avatar-description, [rds-avatar-description], [rdsAvatarDescription]',
                    standalone: true,
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }] } });
