// @rds/angular-components/esm2022/lib/theme/theme.directive.mjs
import { Directive, ElementRef, inject, Input } from '@angular/core';
import { RdsThemeService } from './theme.service';
import * as i0 from "@angular/core";
export class RdsThemeDirective {
    constructor() {
        this.themeService = inject(RdsThemeService);
        this.nativeElement = inject(ElementRef).nativeElement;
    }
    set theme(theme) {
        if (!!theme && this.themeService.get(this.nativeElement) !== theme) {
            this._theme = theme;
            this.themeService.set(this.theme, this.nativeElement);
        }
    }
    get theme() {
        return this._theme;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsThemeDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsThemeDirective, isStandalone: true, selector: "[rds-theme], [rdsTheme]", inputs: { theme: ["rds-theme", "theme"] }, providers: [RdsThemeService], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsThemeDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rds-theme], [rdsTheme]',
                    standalone: true,
                    providers: [RdsThemeService],
                }]
        }], propDecorators: { theme: [{
                type: Input,
                args: ['rds-theme']
            }] } });
