// @rds/angular-components/esm2022/lib/progress-bar/progress-bar.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RdsProgressBarComponent } from './progress-bar.component';
import { LottieComponent } from 'ngx-lottie';
import * as i0 from "@angular/core";
export class RdsProgressBarModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsProgressBarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsProgressBarModule, declarations: [RdsProgressBarComponent], imports: [CommonModule, LottieComponent], exports: [RdsProgressBarComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsProgressBarModule, imports: [CommonModule, LottieComponent] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsProgressBarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, LottieComponent],
                    exports: [RdsProgressBarComponent],
                    declarations: [RdsProgressBarComponent],
                }]
        }] });
