// @rds/angular-components/esm2022/lib/dropdown/dropdown.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RdsDropdownContainerComponent } from './dropdown-container.component';
import { OverlayModule } from '@angular/cdk/overlay';
import { RdsDropdownDirective } from './dropdown.directive';
import { RdsDropdownToggleDirective } from './dropdown-toggle.directive';
import { RdsDropdownMenuDirective } from './dropdown-menu.directive';
import { RdsDropdownToggleButtonComponent } from './dropdown-toggle-button.component';
import { RdsIconComponent } from '../icons';
import * as i0 from "@angular/core";
const COMPONENTS = [
    RdsDropdownContainerComponent,
    RdsDropdownToggleButtonComponent,
];
const DIRECTIVES = [
    RdsDropdownDirective,
    RdsDropdownMenuDirective,
    RdsDropdownToggleDirective,
];
export class RdsDropdownModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDropdownModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsDropdownModule, declarations: [RdsDropdownContainerComponent,
            RdsDropdownToggleButtonComponent, RdsDropdownDirective,
            RdsDropdownMenuDirective,
            RdsDropdownToggleDirective], imports: [CommonModule, OverlayModule, RdsIconComponent], exports: [RdsDropdownContainerComponent,
            RdsDropdownToggleButtonComponent, RdsDropdownDirective,
            RdsDropdownMenuDirective,
            RdsDropdownToggleDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDropdownModule, imports: [CommonModule, OverlayModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDropdownModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [...COMPONENTS, ...DIRECTIVES],
                    exports: [...COMPONENTS, ...DIRECTIVES],
                    imports: [CommonModule, OverlayModule, RdsIconComponent],
                }]
        }] });
