// @rds/angular-components/esm2022/lib/dropdown/dropdown-toggle.directive.mjs
import { DestroyRef, Directive, ElementRef, HostBinding, HostListener, inject, Input, } from '@angular/core';
import { fromEvent, Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import * as i0 from "@angular/core";
export class RdsDropdownToggleDirective {
    constructor() {
        this.elementRef = inject(ElementRef);
        this.destroyRef = inject(DestroyRef);
        this.toggle = new Subject();
        this._isOpen = false;
        this._eventsDestroyer$ = new Subject();
        this.trigger = 'click';
        this.class = 'rds-dropdown-toggle';
        this.disabled = false;
    }
    static { this.TRIGGER_SPLITTER = ', '; }
    get isOpen() {
        return this._isOpen;
    }
    set isOpen(isOpen) {
        this._isOpen = isOpen;
    }
    onEnter($event) {
        this.toggle.next($event);
    }
    ngOnChanges(changes) {
        if (changes['trigger'] && !changes['trigger'].isFirstChange()) {
            this._refreshEvents();
        }
    }
    ngAfterContentInit() {
        this._refreshEvents();
    }
    ngOnDestroy() {
        this._eventsDestroyer$.next();
        this._eventsDestroyer$.complete();
    }
    _refreshEvents() {
        this._eventsDestroyer$.next();
        if (this.trigger === null || this.trigger === undefined) {
            this.trigger = 'click';
        }
        this.trigger
            .split(RdsDropdownToggleDirective.TRIGGER_SPLITTER)
            .forEach((trigger) => fromEvent(this.elementRef.nativeElement, trigger)
            .pipe(takeUntil(this._eventsDestroyer$), takeUntilDestroyed(this.destroyRef))
            .subscribe((event) => this.toggle.next(event)));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDropdownToggleDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsDropdownToggleDirective, selector: "[rdsDropdownToggle], [rds-dropdown-toggle]", inputs: { trigger: "trigger", isOpen: "isOpen" }, host: { listeners: { "keydown.enter": "onEnter($event)" }, properties: { "class": "this.class", "disabled": "this.disabled", "attr.aria-expanded": "this.isOpen" } }, exportAs: ["rdsDropdownToggle"], usesOnChanges: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDropdownToggleDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rdsDropdownToggle], [rds-dropdown-toggle]',
                    exportAs: 'rdsDropdownToggle',
                }]
        }], propDecorators: { trigger: [{
                type: Input
            }], class: [{
                type: HostBinding,
                args: ['class']
            }], disabled: [{
                type: HostBinding,
                args: ['disabled']
            }], isOpen: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['attr.aria-expanded']
            }], onEnter: [{
                type: HostListener,
                args: ['keydown.enter', ['$event']]
            }] } });
