// @rds/angular-components/esm2022/lib/datepicker/datepicker.intl.mjs
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
export class RdsDatepickerIntl {
    constructor() {
        this.changes = new Subject();
        this.calendarLabel = 'Calendar';
        this.openCalendarLabel = 'Open calendar';
        this.closeCalendarLabel = 'Close calendar';
        this.prevMonthLabel = 'Previous month';
        this.nextMonthLabel = 'Next month';
        this.prevYearLabel = 'Previous year';
        this.nextYearLabel = 'Next year';
        this.prevMultiYearLabel = 'Previous 20 years';
        this.nextMultiYearLabel = 'Next 20 years';
        this.switchToMonthViewLabel = 'Choose date';
        this.switchToMultiYearViewLabel = 'Choose month and year';
        this.monthLabel = 'Weekly';
        this.yearLabel = 'Monthly';
        this.multiYearLabel = 'Yearly';
    }
    formatYearRange(start, end) {
        return `${start} \u2013 ${end}`;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDatepickerIntl, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDatepickerIntl, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDatepickerIntl, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
