// @rds/angular-components/esm2022/lib/datepicker/date-range-picker.component.mjs
import { ChangeDetectionStrategy, Component, ViewEncapsulation, } from '@angular/core';
import { RdsDatepickerBaseDirective, } from './datepicker-base';
import { RDS_RANGE_DATE_SELECTION_MODEL_PROVIDER, } from './date-selection.model';
import { RDS_CALENDAR_RANGE_STRATEGY_PROVIDER } from './date-range-selection.strategy';
import * as i0 from "@angular/core";
export class RdsDateRangePickerComponent extends RdsDatepickerBaseDirective {
    _forwardContentValues(instance) {
        super._forwardContentValues(instance);
        const input = this.datepickerInput;
        if (input) {
            instance.comparisonStart = input.comparisonStart;
            instance.comparisonEnd = input.comparisonEnd;
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDateRangePickerComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsDateRangePickerComponent, selector: "rds-date-range-picker", providers: [
            RDS_RANGE_DATE_SELECTION_MODEL_PROVIDER,
            RDS_CALENDAR_RANGE_STRATEGY_PROVIDER,
            {
                provide: RdsDatepickerBaseDirective,
                useExisting: RdsDateRangePickerComponent,
            },
        ], exportAs: ["rdsDateRangePicker"], usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDateRangePickerComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'rds-date-range-picker',
                    template: '',
                    exportAs: 'rdsDateRangePicker',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    providers: [
                        RDS_RANGE_DATE_SELECTION_MODEL_PROVIDER,
                        RDS_CALENDAR_RANGE_STRATEGY_PROVIDER,
                        {
                            provide: RdsDatepickerBaseDirective,
                            useExisting: RdsDateRangePickerComponent,
                        },
                    ],
                }]
        }] });
