// @rds/angular-components/esm2022/lib/datepicker/datepicker-actions.component.mjs
import { ChangeDetectionStrategy, Component, Directive, HostListener, inject, TemplateRef, ViewChild, ViewContainerRef, ViewEncapsulation, } from '@angular/core';
import { TemplatePortal } from '@angular/cdk/portal';
import { RdsDatepickerBaseDirective, } from './datepicker-base';
import * as i0 from "@angular/core";
export class RdsDatepickerApplyDirective {
    constructor() {
        this.datepicker = inject(RdsDatepickerBaseDirective);
    }
    _applySelection() {
        this.datepicker._applyPendingSelection();
        this.datepicker.close();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDatepickerApplyDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsDatepickerApplyDirective, selector: "[rdsDatepickerApply], [rdsDateRangePickerApply], [rds-datepicker-apply], [rds-date-range-picker-apply]", host: { listeners: { "click": "_applySelection()" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDatepickerApplyDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rdsDatepickerApply], [rdsDateRangePickerApply], [rds-datepicker-apply], [rds-date-range-picker-apply]',
                }]
        }], propDecorators: { _applySelection: [{
                type: HostListener,
                args: ['click']
            }] } });
export class RdsDatepickerCancelDirective {
    constructor() {
        this.datepicker = inject(RdsDatepickerBaseDirective);
    }
    _applySelection() {
        this.datepicker.close();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDatepickerCancelDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsDatepickerCancelDirective, selector: "[rdsDatepickerCancel], [rdsDateRangePickerCancel], [rds-datepicker-cancel], [rds-date-range-picker-cancel]", host: { listeners: { "click": "_applySelection()" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDatepickerCancelDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rdsDatepickerCancel], [rdsDateRangePickerCancel], [rds-datepicker-cancel], [rds-date-range-picker-cancel]',
                }]
        }], propDecorators: { _applySelection: [{
                type: HostListener,
                args: ['click']
            }] } });
export class RdsDatepickerActionsComponent {
    constructor() {
        this.datepicker = inject(RdsDatepickerBaseDirective);
        this.viewContainerRef = inject(ViewContainerRef);
        this._template = null;
        this._portal = null;
    }
    ngAfterViewInit() {
        this._portal = new TemplatePortal(this._template, this.viewContainerRef);
        this.datepicker.registerActions(this._portal);
    }
    ngOnDestroy() {
        this.datepicker.removeActions(this._portal);
        if (this._portal && this._portal.isAttached) {
            this._portal?.detach();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDatepickerActionsComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsDatepickerActionsComponent, selector: "rds-datepicker-actions, rds-date-range-picker-actions", viewQueries: [{ propertyName: "_template", first: true, predicate: TemplateRef, descendants: true }], ngImport: i0, template: `
    <ng-template>
      <div class="rds-datepicker-actions">
        <ng-content />
      </div>
    </ng-template>
  `, isInline: true, styles: [".rds-datepicker-actions{background:var(--rds-color-bg-base-box);display:flex;justify-content:flex-end;align-items:center;padding:16px;border-top:1px solid var(--rds-color-grey-2)}.rds-datepicker-actions .rds-button+.rds-button{margin-left:16px}[dir=rtl] .rds-datepicker-actions .rds-button+.rds-button{margin-left:0;margin-right:16px}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDatepickerActionsComponent, decorators: [{
            type: Component,
            args: [{ selector: 'rds-datepicker-actions, rds-date-range-picker-actions', template: `
    <ng-template>
      <div class="rds-datepicker-actions">
        <ng-content />
      </div>
    </ng-template>
  `, changeDetection: ChangeDetectionStrategy.OnPush, encapsulation: ViewEncapsulation.None, styles: [".rds-datepicker-actions{background:var(--rds-color-bg-base-box);display:flex;justify-content:flex-end;align-items:center;padding:16px;border-top:1px solid var(--rds-color-grey-2)}.rds-datepicker-actions .rds-button+.rds-button{margin-left:16px}[dir=rtl] .rds-datepicker-actions .rds-button+.rds-button{margin-left:0;margin-right:16px}\n"] }]
        }], propDecorators: { _template: [{
                type: ViewChild,
                args: [TemplateRef]
            }] } });
