// @rds/angular-components/esm2022/lib/datepicker/date-range-selection.strategy.mjs
import { inject, Injectable, InjectionToken, Optional, SkipSelf, } from '@angular/core';
import { DateAdapter } from './adapter';
import { DateRange } from './date-selection.model';
import * as i0 from "@angular/core";
export const RDS_DATE_RANGE_SELECTION_STRATEGY = new InjectionToken('RDS_DATE_RANGE_SELECTION_STRATEGY');
export class DefaultRdsCalendarRangeStrategy {
    constructor() {
        this.dateAdapter = inject(DateAdapter);
    }
    selectionFinished(date, currentRange) {
        let { start, end } = currentRange;
        if (start == null) {
            start = date;
        }
        else if (end == null &&
            date &&
            this.dateAdapter.compareDate(date, start) >= 0) {
            end = date;
        }
        else {
            start = date;
            end = null;
        }
        return new DateRange(start, end);
    }
    createPreview(activeDate, currentRange) {
        let start = null;
        let end = null;
        if (currentRange.start && !currentRange.end && activeDate) {
            start = currentRange.start;
            end = activeDate;
        }
        return new DateRange(start, end);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: DefaultRdsCalendarRangeStrategy, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: DefaultRdsCalendarRangeStrategy }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: DefaultRdsCalendarRangeStrategy, decorators: [{
            type: Injectable
        }] });
export function RDS_CALENDAR_RANGE_STRATEGY_PROVIDER_FACTORY(parent) {
    return parent || new DefaultRdsCalendarRangeStrategy();
}
export const RDS_CALENDAR_RANGE_STRATEGY_PROVIDER = {
    provide: RDS_DATE_RANGE_SELECTION_STRATEGY,
    deps: [
        [new Optional(), new SkipSelf(), RDS_DATE_RANGE_SELECTION_STRATEGY],
        DateAdapter,
    ],
    useFactory: RDS_CALENDAR_RANGE_STRATEGY_PROVIDER_FACTORY,
};
