// @rds/angular-components/esm2022/lib/datepicker/datepicker.component.mjs
import { ChangeDetectionStrategy, Component, ViewEncapsulation, } from '@angular/core';
import { RdsDatepickerBaseDirective, } from './datepicker-base';
import { RDS_SINGLE_DATE_SELECTION_MODEL_PROVIDER } from './date-selection.model';
import * as i0 from "@angular/core";
export class RdsDatepickerComponent extends RdsDatepickerBaseDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDatepickerComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsDatepickerComponent, selector: "rds-datepicker", providers: [
            RDS_SINGLE_DATE_SELECTION_MODEL_PROVIDER,
            {
                provide: RdsDatepickerBaseDirective,
                useExisting: RdsDatepickerComponent,
            },
        ], exportAs: ["rdsDatepicker"], usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDatepickerComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'rds-datepicker',
                    template: '',
                    exportAs: 'rdsDatepicker',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    providers: [
                        RDS_SINGLE_DATE_SELECTION_MODEL_PROVIDER,
                        {
                            provide: RdsDatepickerBaseDirective,
                            useExisting: RdsDatepickerComponent,
                        },
                    ],
                }]
        }] });
