// @rds/angular-components/esm2022/lib/datepicker/datepicker-toggle.component.mjs
import { Attribute, booleanAttribute, ChangeDetectionStrategy, ChangeDetectorRef, Component, ContentChild, DestroyRef, Directive, forwardRef, HostBinding, HostListener, inject, Input, numberAttribute, ViewEncapsulation, } from '@angular/core';
import { merge, of, Subscription } from 'rxjs';
import { RdsDatepickerIntl } from './datepicker.intl';
import { RdsFormFieldComponent } from '../form-field';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { TAB_INDEX_DEFAULT } from '../utils';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "../button/icon-button.component";
import * as i3 from "../icons/icon.component";
export class RdsDatepickerToggleIconDirective {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDatepickerToggleIconDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsDatepickerToggleIconDirective, selector: "[rdsDatepickerToggleIcon], [rds-datepicker-toggle-icon]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDatepickerToggleIconDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rdsDatepickerToggleIcon], [rds-datepicker-toggle-icon]',
                }]
        }] });
export class RdsDatepickerToggleComponent {
    get classes() {
        const classes = ['rds-datepicker-toggle'];
        if (this.forDatepicker && this.forDatepicker.opened) {
            classes.push('rds-datepicker-toggle-active');
        }
        return classes;
    }
    get dataRdsCalendar() {
        return this.forDatepicker ? this.forDatepicker.id : null;
    }
    set disabled(value) {
        const newValue = value === undefined ? undefined : booleanAttribute(value);
        if (newValue !== this._disabled) {
            this._disabled = newValue;
            this.cdr.markForCheck();
        }
    }
    get disabled() {
        if (this._disabled === undefined && this.forDatepicker) {
            return this.forDatepicker.disabled;
        }
        return this._disabled || false;
    }
    get _buttonSize() {
        if (this.formField) {
            return this.formField?.size !== 'l' ? this.formField?.size : 'm';
        }
        else {
            return 's';
        }
    }
    _onClick($event) {
        this._open($event);
    }
    constructor(tabIndex) {
        this.intl = inject(RdsDatepickerIntl);
        this.destroyRef = inject(DestroyRef);
        this.cdr = inject(ChangeDetectorRef);
        this.formField = inject(forwardRef(() => RdsFormFieldComponent), { optional: true });
        this._stateChanges = Subscription.EMPTY;
        this._tabIndex = null;
        this.ariaLabel = '';
        this._customIcon = null;
        const parsedTabIndex = numberAttribute(tabIndex, TAB_INDEX_DEFAULT);
        this.tabIndex =
            parsedTabIndex || parsedTabIndex === 0 ? parsedTabIndex : null;
    }
    ngOnChanges(changes) {
        if (changes['datepicker']) {
            this._watchStateChanges();
        }
    }
    ngOnDestroy() {
        this._stateChanges.unsubscribe();
    }
    ngAfterContentInit() {
        this._watchStateChanges();
        if (this.formField) {
            this.formField.control.stateChanges
                .pipe(takeUntilDestroyed(this.destroyRef))
                .subscribe(() => this.cdr.markForCheck());
        }
    }
    _open(event) {
        if (this.forDatepicker && !this.disabled) {
            this.forDatepicker.open();
            event.stopPropagation();
        }
    }
    _watchStateChanges() {
        const datepickerStateChanged = this.forDatepicker
            ? this.forDatepicker.stateChanges.asObservable()
            : of(null);
        const inputStateChanged = this.forDatepicker && this.forDatepicker.datepickerInput
            ? this.forDatepicker.datepickerInput.stateChanges
            : of(null);
        const datepickerToggled = this.forDatepicker
            ? merge(this.forDatepicker.openedStream, this.forDatepicker.closedStream)
            : of(null);
        this._stateChanges.unsubscribe();
        this._stateChanges = merge(this.intl.changes, datepickerStateChanged, inputStateChanged, datepickerToggled).subscribe(() => this.cdr.markForCheck());
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDatepickerToggleComponent, deps: [{ token: 'tabindex', attribute: true }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "17.3.1", type: RdsDatepickerToggleComponent, selector: "rds-datepicker-toggle", inputs: { forDatepicker: "forDatepicker", tabIndex: ["tabIndex", "tabIndex", numberAttribute], ariaLabel: ["aria-label", "ariaLabel"], disabled: "disabled" }, host: { listeners: { "click": "_onClick($event)" }, properties: { "class": "this.classes", "attr.data-rds-calendar": "this.dataRdsCalendar", "attr.tabindex": "this._tabIndex" } }, queries: [{ propertyName: "_customIcon", first: true, predicate: RdsDatepickerToggleIconDirective, descendants: true }], exportAs: ["rdsDatepickerToggle"], usesOnChanges: true, ngImport: i0, template: "<button\n  rds-icon-button\n  type=\"button\"\n  [size]=\"_buttonSize\"\n  [ngClass]=\"{ active: forDatepicker?.opened }\"\n  class=\"rds-button--with-only-icon\"\n  [attr.aria-haspopup]=\"forDatepicker ? 'dialog' : null\"\n  [attr.aria-label]=\"ariaLabel || intl.openCalendarLabel\"\n  [attr.tabindex]=\"disabled ? -1 : tabIndex\"\n  [disabled]=\"disabled\"\n>\n  @if (!_customIcon) {\n    <rds-icon\n      class=\"rds-datepicker-toggle-default-icon\"\n      icon=\"calendar\"\n      namespace=\"outlined\"\n    />\n  }\n  <ng-content\n    select=\"[rdsDatepickerToggleIcon], [rds-datepicker-toggle-icon]\"\n  />\n</button>\n", styles: [".rds-form-field .rds-form-field__content-prefix .rds-datepicker-toggle,.rds-form-field .rds-form-field__content-suffix .rds-datepicker-toggle{display:flex}.rds-form-field .rds-form-field__content-prefix .rds-datepicker-toggle-default-icon,.rds-form-field .rds-form-field__content-suffix .rds-datepicker-toggle-default-icon{display:block;width:1.5em;height:1.5em}.rds-form-field .rds-form-field__content-prefix .rds-icon-button .rds-datepicker-toggle-default-icon,.rds-form-field .rds-form-field__content-suffix .rds-icon-button .rds-datepicker-toggle-default-icon{margin:auto}\n"], dependencies: [{ kind: "directive", type: i1.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { kind: "component", type: i2.RdsIconButtonComponent, selector: "\n    button[rds-icon-button],\n    button[rdsIconButton],\n    button[rds-icon-warning-button],\n    button[rdsIconWarningButton],\n  ", inputs: ["size"] }, { kind: "component", type: i3.RdsIconComponent, selector: "rds-icon", inputs: ["namespace", "icon"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDatepickerToggleComponent, decorators: [{
            type: Component,
            args: [{ selector: 'rds-datepicker-toggle', exportAs: 'rdsDatepickerToggle', encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, template: "<button\n  rds-icon-button\n  type=\"button\"\n  [size]=\"_buttonSize\"\n  [ngClass]=\"{ active: forDatepicker?.opened }\"\n  class=\"rds-button--with-only-icon\"\n  [attr.aria-haspopup]=\"forDatepicker ? 'dialog' : null\"\n  [attr.aria-label]=\"ariaLabel || intl.openCalendarLabel\"\n  [attr.tabindex]=\"disabled ? -1 : tabIndex\"\n  [disabled]=\"disabled\"\n>\n  @if (!_customIcon) {\n    <rds-icon\n      class=\"rds-datepicker-toggle-default-icon\"\n      icon=\"calendar\"\n      namespace=\"outlined\"\n    />\n  }\n  <ng-content\n    select=\"[rdsDatepickerToggleIcon], [rds-datepicker-toggle-icon]\"\n  />\n</button>\n", styles: [".rds-form-field .rds-form-field__content-prefix .rds-datepicker-toggle,.rds-form-field .rds-form-field__content-suffix .rds-datepicker-toggle{display:flex}.rds-form-field .rds-form-field__content-prefix .rds-datepicker-toggle-default-icon,.rds-form-field .rds-form-field__content-suffix .rds-datepicker-toggle-default-icon{display:block;width:1.5em;height:1.5em}.rds-form-field .rds-form-field__content-prefix .rds-icon-button .rds-datepicker-toggle-default-icon,.rds-form-field .rds-form-field__content-suffix .rds-icon-button .rds-datepicker-toggle-default-icon{margin:auto}\n"] }]
        }], ctorParameters: () => [{ type: undefined, decorators: [{
                    type: Attribute,
                    args: ['tabindex']
                }] }], propDecorators: { classes: [{
                type: HostBinding,
                args: ['class']
            }], dataRdsCalendar: [{
                type: HostBinding,
                args: ['attr.data-rds-calendar']
            }], _tabIndex: [{
                type: HostBinding,
                args: ['attr.tabindex']
            }], forDatepicker: [{
                type: Input
            }], tabIndex: [{
                type: Input,
                args: [{ transform: numberAttribute }]
            }], ariaLabel: [{
                type: Input,
                args: ['aria-label']
            }], disabled: [{
                type: Input
            }], _customIcon: [{
                type: ContentChild,
                args: [RdsDatepickerToggleIconDirective]
            }], _onClick: [{
                type: HostListener,
                args: ['click', ['$event']]
            }] } });
