// @rds/angular-components/esm2022/lib/datepicker/datepicker.module.mjs
import { A11yModule } from '@angular/cdk/a11y';
import { OverlayModule } from '@angular/cdk/overlay';
import { PortalModule } from '@angular/cdk/portal';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { CdkScrollableModule } from '@angular/cdk/scrolling';
import { RdsCalendarComponent, RdsCalendarHeaderComponent, } from './calendar.component';
import { RdsCalendarBodyComponent } from './calendar-body.component';
import { RdsDatepickerComponent } from './datepicker.component';
import { RDS_DATEPICKER_SCROLL_STRATEGY_FACTORY_PROVIDER, RdsDatepickerContentComponent, } from './datepicker-base';
import { RdsDatepickerInputDirective } from './datepicker-input';
import { RdsDatepickerIntl } from './datepicker.intl';
import { RdsDatepickerToggleComponent, RdsDatepickerToggleIconDirective, } from './datepicker-toggle.component';
import { RdsMonthViewComponent } from './month-view.component';
import { RdsMultiYearViewComponent } from './multi-year-view.component';
import { RdsYearViewComponent } from './year-view.component';
import { RdsDateRangeInputComponent } from './date-range-input.component';
import { RdsEndDateDirective, RdsStartDateDirective, } from './date-range-input.parts';
import { RdsDateRangePickerComponent } from './date-range-picker.component';
import { RdsDatepickerActionsComponent, RdsDatepickerApplyDirective, RdsDatepickerCancelDirective, } from './datepicker-actions.component';
import { RdsButtonModule } from '../button';
import { RdsDialogModule } from '../dialog';
import { RdsIconComponent } from '../icons';
import { RdsToggleButtonModule } from '../toggle-button';
import * as i0 from "@angular/core";
const ELEMENTS = [
    RdsCalendarComponent,
    RdsCalendarHeaderComponent,
    RdsCalendarBodyComponent,
    RdsDatepickerComponent,
    RdsDatepickerContentComponent,
    RdsDatepickerInputDirective,
    RdsDatepickerToggleComponent,
    RdsDatepickerToggleIconDirective,
    RdsMonthViewComponent,
    RdsYearViewComponent,
    RdsMultiYearViewComponent,
    RdsDateRangeInputComponent,
    RdsStartDateDirective,
    RdsEndDateDirective,
    RdsDateRangePickerComponent,
    RdsDatepickerActionsComponent,
    RdsDatepickerCancelDirective,
    RdsDatepickerApplyDirective,
];
export class RdsDatepickerModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDatepickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsDatepickerModule, declarations: [RdsCalendarComponent,
            RdsCalendarHeaderComponent,
            RdsCalendarBodyComponent,
            RdsDatepickerComponent,
            RdsDatepickerContentComponent,
            RdsDatepickerInputDirective,
            RdsDatepickerToggleComponent,
            RdsDatepickerToggleIconDirective,
            RdsMonthViewComponent,
            RdsYearViewComponent,
            RdsMultiYearViewComponent,
            RdsDateRangeInputComponent,
            RdsStartDateDirective,
            RdsEndDateDirective,
            RdsDateRangePickerComponent,
            RdsDatepickerActionsComponent,
            RdsDatepickerCancelDirective,
            RdsDatepickerApplyDirective], imports: [CommonModule,
            RdsButtonModule,
            RdsDialogModule,
            OverlayModule,
            A11yModule,
            PortalModule,
            RdsIconComponent,
            RdsToggleButtonModule], exports: [CdkScrollableModule, OverlayModule, RdsCalendarComponent,
            RdsCalendarHeaderComponent,
            RdsCalendarBodyComponent,
            RdsDatepickerComponent,
            RdsDatepickerContentComponent,
            RdsDatepickerInputDirective,
            RdsDatepickerToggleComponent,
            RdsDatepickerToggleIconDirective,
            RdsMonthViewComponent,
            RdsYearViewComponent,
            RdsMultiYearViewComponent,
            RdsDateRangeInputComponent,
            RdsStartDateDirective,
            RdsEndDateDirective,
            RdsDateRangePickerComponent,
            RdsDatepickerActionsComponent,
            RdsDatepickerCancelDirective,
            RdsDatepickerApplyDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDatepickerModule, providers: [
            RdsDatepickerIntl,
            RDS_DATEPICKER_SCROLL_STRATEGY_FACTORY_PROVIDER,
        ], imports: [CommonModule,
            RdsButtonModule,
            RdsDialogModule,
            OverlayModule,
            A11yModule,
            PortalModule,
            RdsToggleButtonModule, CdkScrollableModule, OverlayModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDatepickerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        RdsButtonModule,
                        RdsDialogModule,
                        OverlayModule,
                        A11yModule,
                        PortalModule,
                        RdsIconComponent,
                        RdsToggleButtonModule,
                    ],
                    exports: [CdkScrollableModule, OverlayModule, ...ELEMENTS],
                    declarations: [...ELEMENTS],
                    providers: [
                        RdsDatepickerIntl,
                        RDS_DATEPICKER_SCROLL_STRATEGY_FACTORY_PROVIDER,
                    ],
                }]
        }] });
