// @rds/angular-components/esm2022/lib/drawer/drawer.module.mjs
import { NgModule } from '@angular/core';
import { RdsDrawerActionsDirective } from './drawer-actions.directive';
import { RdsDrawerContainerComponent } from './drawer-container.component';
import { RdsDrawerContentComponent } from './drawer-content.component';
import { RdsDrawerHeaderDirective } from './drawer-header.directive';
import { RdsDrawerComponent } from './drawer.component';
import * as i0 from "@angular/core";
export const RDS_DRAWER_DIRECTIVES = [
    RdsDrawerContainerComponent,
    RdsDrawerContentComponent,
    RdsDrawerComponent,
    RdsDrawerHeaderDirective,
    RdsDrawerActionsDirective,
];
export class RdsDrawerModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDrawerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsDrawerModule, imports: [RdsDrawerContainerComponent,
            RdsDrawerContentComponent,
            RdsDrawerComponent,
            RdsDrawerHeaderDirective,
            RdsDrawerActionsDirective], exports: [RdsDrawerContainerComponent,
            RdsDrawerContentComponent,
            RdsDrawerComponent,
            RdsDrawerHeaderDirective,
            RdsDrawerActionsDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDrawerModule, imports: [RdsDrawerComponent] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDrawerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...RDS_DRAWER_DIRECTIVES],
                    exports: [...RDS_DRAWER_DIRECTIVES],
                }]
        }] });
