// @rds/angular-components/esm2022/lib/drawer/drawer-content.component.mjs
import { CdkScrollable, ScrollDispatcher } from '@angular/cdk/scrolling';
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, DestroyRef, ElementRef, forwardRef, HostBinding, inject, NgZone, ViewEncapsulation, } from '@angular/core';
import { RdsDrawerContainerComponent } from './drawer-container.component';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import * as i0 from "@angular/core";
export class RdsDrawerContentComponent extends CdkScrollable {
    get marginRight() {
        return this.contentMargins.right ?? 0;
    }
    get marginLeft() {
        return this.contentMargins.left ?? 0;
    }
    constructor() {
        super(inject(ElementRef), inject(ScrollDispatcher), inject(NgZone));
        this.destroyRef = inject(DestroyRef);
        this.cdr = inject(ChangeDetectorRef);
        this.container = inject(forwardRef(() => RdsDrawerContainerComponent));
        this.class = 'rds-drawer-content';
    }
    ngAfterContentInit() {
        this.container.contentMargins$
            .pipe(takeUntilDestroyed(this.destroyRef))
            .subscribe((m) => {
            this.contentMargins = m;
            this.cdr.markForCheck();
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDrawerContentComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsDrawerContentComponent, isStandalone: true, selector: "rds-drawer-content", host: { properties: { "class": "this.class", "style.margin-right.px": "this.marginRight", "style.margin-left.px": "this.marginLeft" } }, usesInheritance: true, ngImport: i0, template: '<ng-content />', isInline: true, styles: [".rds-drawer-content{width:100%;height:100%}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDrawerContentComponent, decorators: [{
            type: Component,
            args: [{ selector: 'rds-drawer-content', template: '<ng-content />', encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, standalone: true, styles: [".rds-drawer-content{width:100%;height:100%}\n"] }]
        }], ctorParameters: () => [], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }], marginRight: [{
                type: HostBinding,
                args: ['style.margin-right.px']
            }], marginLeft: [{
                type: HostBinding,
                args: ['style.margin-left.px']
            }] } });
