// @rds/angular-components/esm2022/lib/checkbox/checkbox.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RdsCheckboxComponent } from './checkbox.component';
import { ObserversModule } from '@angular/cdk/observers';
import { RdsPseudoCheckboxComponent } from './pseudo-checkbox.component';
import { RdsHelperTextModule } from '../form-field/helper-text.module';
import { ReactiveFormsModule } from '@angular/forms';
import * as i0 from "@angular/core";
export class RdsCheckboxModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCheckboxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsCheckboxModule, declarations: [RdsCheckboxComponent, RdsPseudoCheckboxComponent], imports: [CommonModule,
            ObserversModule,
            RdsHelperTextModule,
            ReactiveFormsModule], exports: [RdsCheckboxComponent, RdsPseudoCheckboxComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCheckboxModule, imports: [CommonModule,
            ObserversModule,
            RdsHelperTextModule,
            ReactiveFormsModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsCheckboxModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        ObserversModule,
                        RdsHelperTextModule,
                        ReactiveFormsModule,
                    ],
                    declarations: [RdsCheckboxComponent, RdsPseudoCheckboxComponent],
                    exports: [RdsCheckboxComponent, RdsPseudoCheckboxComponent],
                }]
        }] });
