// @rds/angular-components/esm2022/lib/footer/footer-sitemap.directive.mjs
import { Directive, HostBinding } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsFooterSitemapSectionDirective {
    constructor() {
        this.class = 'rds-footer-sitemap-section';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsFooterSitemapSectionDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsFooterSitemapSectionDirective, selector: "rds-footer-sitemap-section", host: { properties: { "class": "this.class" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsFooterSitemapSectionDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'rds-footer-sitemap-section',
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }] } });
export class RdsFooterSitemapDirective {
    constructor() {
        this.class = 'rds-footer-sitemap';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsFooterSitemapDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsFooterSitemapDirective, selector: "rds-footer-sitemap", host: { properties: { "class": "this.class" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsFooterSitemapDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'rds-footer-sitemap',
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }] } });
