// @rds/angular-components/esm2022/lib/footer/footer-link.component.mjs
import { ChangeDetectionStrategy, Component, HostBinding, ViewEncapsulation, } from '@angular/core';
import { RdsLinkComponent } from '../link';
import * as i0 from "@angular/core";
export class RdsFooterLinkComponent extends RdsLinkComponent {
    constructor() {
        super(...arguments);
        this.class = 'rds-footer-link';
        this.type = 'secondary';
        this.size = 's';
        this._standalone = true;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsFooterLinkComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsFooterLinkComponent, selector: "rds-footer-link", host: { properties: { "class": "this.class" } }, usesInheritance: true, ngImport: i0, template: `<ng-content />`, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsFooterLinkComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'rds-footer-link',
                    template: `<ng-content />`,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }] } });
