// @rds/angular-components/esm2022/lib/footer/footer.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RdsButtonModule } from '../button';
import { RdsDividerModule } from '../divider';
import { RdsFormFieldModule } from '../form-field';
import { RdsFooterCaptionDirective } from './footer-caption.directive';
import { RdsFooterCustomContentDirective } from './footer-custom-content.directive';
import { RdsFooterDisclaimerDirective } from './footer-disclaimer.directive';
import { RdsFooterLinkComponent } from './footer-link.component';
import { RdsFooterNavigationDirective } from './footer-navigation.directive';
import { RdsFooterSitemapDirective, RdsFooterSitemapSectionDirective, } from './footer-sitemap.directive';
import { RdsFooterTitleDirective } from './footer-title.directive';
import { RdsFooterComponent } from './footer.component';
import { ObserversModule } from '@angular/cdk/observers';
import * as i0 from "@angular/core";
const ELEMENTS = [
    RdsFooterCaptionDirective,
    RdsFooterCustomContentDirective,
    RdsFooterComponent,
    RdsFooterNavigationDirective,
    RdsFooterLinkComponent,
    RdsFooterDisclaimerDirective,
    RdsFooterSitemapDirective,
    RdsFooterSitemapSectionDirective,
    RdsFooterTitleDirective,
];
export class RdsFooterModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsFooterModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsFooterModule, declarations: [RdsFooterCaptionDirective,
            RdsFooterCustomContentDirective,
            RdsFooterComponent,
            RdsFooterNavigationDirective,
            RdsFooterLinkComponent,
            RdsFooterDisclaimerDirective,
            RdsFooterSitemapDirective,
            RdsFooterSitemapSectionDirective,
            RdsFooterTitleDirective], imports: [CommonModule,
            FormsModule,
            ObserversModule,
            ReactiveFormsModule,
            RdsDividerModule,
            RdsFormFieldModule,
            RdsButtonModule], exports: [RdsFooterCaptionDirective,
            RdsFooterCustomContentDirective,
            RdsFooterComponent,
            RdsFooterNavigationDirective,
            RdsFooterLinkComponent,
            RdsFooterDisclaimerDirective,
            RdsFooterSitemapDirective,
            RdsFooterSitemapSectionDirective,
            RdsFooterTitleDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsFooterModule, imports: [CommonModule,
            FormsModule,
            ObserversModule,
            ReactiveFormsModule,
            RdsDividerModule,
            RdsFormFieldModule,
            RdsButtonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsFooterModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [...ELEMENTS],
                    exports: [...ELEMENTS],
                    imports: [
                        CommonModule,
                        FormsModule,
                        ObserversModule,
                        ReactiveFormsModule,
                        RdsDividerModule,
                        RdsFormFieldModule,
                        RdsButtonModule,
                    ],
                }]
        }] });
