// @rds/angular-components/esm2022/lib/paginator/paginator.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RDS_PAGINATOR_INTL_PROVIDER } from './paginator-intl';
import { RdsPaginatorComponent } from './paginator.component';
import { A11yModule } from '@angular/cdk/a11y';
import { RdsDropdownModule } from '../dropdown';
import { RdsIconComponent } from '../icons';
import { RdsTooltipModule } from '../tooltip';
import { RdsButtonModule } from '../button';
import { RdsMenuModule } from '../menu';
import { RdsPaginatorBasicModule } from './paginator-basic/paginator-basic.module';
import * as i0 from "@angular/core";
export class RdsPaginatorModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsPaginatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsPaginatorModule, declarations: [RdsPaginatorComponent], imports: [CommonModule,
            RdsDropdownModule,
            RdsMenuModule,
            RdsIconComponent,
            RdsTooltipModule,
            RdsButtonModule,
            RdsPaginatorBasicModule,
            A11yModule], exports: [RdsPaginatorComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsPaginatorModule, providers: [RDS_PAGINATOR_INTL_PROVIDER], imports: [CommonModule,
            RdsDropdownModule,
            RdsMenuModule,
            RdsTooltipModule,
            RdsButtonModule,
            RdsPaginatorBasicModule,
            A11yModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsPaginatorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [RdsPaginatorComponent],
                    exports: [RdsPaginatorComponent],
                    imports: [
                        CommonModule,
                        RdsDropdownModule,
                        RdsMenuModule,
                        RdsIconComponent,
                        RdsTooltipModule,
                        RdsButtonModule,
                        RdsPaginatorBasicModule,
                        A11yModule,
                    ],
                    providers: [RDS_PAGINATOR_INTL_PROVIDER],
                }]
        }] });
