// @rds/angular-components/esm2022/lib/tree/tree-node-outlet.directive.mjs
import { CDK_TREE_NODE_OUTLET_NODE, CdkTreeNodeOutlet, } from '@angular/cdk/tree';
import { Directive, inject, ViewContainerRef } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsTreeNodeOutletDirective {
    constructor() {
        this.viewContainer = inject(ViewContainerRef);
        this._node = inject(CDK_TREE_NODE_OUTLET_NODE, { optional: true });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTreeNodeOutletDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsTreeNodeOutletDirective, selector: "[rdsTreeNodeOutlet], [rds-tree-node-outlet]", providers: [
            {
                provide: CdkTreeNodeOutlet,
                useExisting: RdsTreeNodeOutletDirective,
            },
        ], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTreeNodeOutletDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rdsTreeNodeOutlet], [rds-tree-node-outlet]',
                    providers: [
                        {
                            provide: CdkTreeNodeOutlet,
                            useExisting: RdsTreeNodeOutletDirective,
                        },
                    ],
                }]
        }] });
