// @rds/angular-components/esm2022/lib/tree/tree-node-text.component.mjs
import { ChangeDetectionStrategy, Component, ElementRef, HostBinding, Input, ViewChild, ViewEncapsulation, } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsTreeNodeTextComponent {
    constructor() {
        this.truncateClass = true;
        this.containerClass = true;
        this.contentLeftClass = true;
        this.alignRightClass = false;
        this.alignCenterClass = false;
        this.textAlignment = 'left';
    }
    ngOnChanges(changes) {
        if (changes['textAlignment'] &&
            changes['textAlignment'].previousValue !==
                changes['textAlignment'].currentValue) {
            if (this.textAlignment === 'right') {
                this.alignRightClass = true;
                this.alignCenterClass = false;
            }
            else if (this.textAlignment === 'center') {
                this.alignRightClass = false;
                this.alignCenterClass = true;
            }
            else {
                this.alignRightClass = false;
                this.alignCenterClass = false;
            }
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTreeNodeTextComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsTreeNodeTextComponent, selector: "rds-tree-node-text", inputs: { textAlignment: "textAlignment" }, host: { properties: { "class.rds-tree-text-node": "this.truncateClass", "class.rds-tree-text-node__container": "this.containerClass", "class.rds-tree-node--content-left": "this.contentLeftClass", "class.rds-tree-text-node__container-align-right": "this.alignRightClass", "class.rds-tree-text-node__container-align-center": "this.alignCenterClass" } }, viewQueries: [{ propertyName: "rdsTreeNodeElement", first: true, predicate: ["rdsTreeNodeElement"], descendants: true, static: true }], exportAs: ["rdsTreeText"], usesOnChanges: true, ngImport: i0, template: `
    <span class="rds-tree-text-node" #rdsTreeNodeElement>
      <ng-content />
    </span>
  `, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTreeNodeTextComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'rds-tree-node-text',
                    exportAs: 'rdsTreeText',
                    template: `
    <span class="rds-tree-text-node" #rdsTreeNodeElement>
      <ng-content />
    </span>
  `,
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], propDecorators: { truncateClass: [{
                type: HostBinding,
                args: ['class.rds-tree-text-node']
            }], containerClass: [{
                type: HostBinding,
                args: ['class.rds-tree-text-node__container']
            }], contentLeftClass: [{
                type: HostBinding,
                args: ['class.rds-tree-node--content-left']
            }], alignRightClass: [{
                type: HostBinding,
                args: ['class.rds-tree-text-node__container-align-right']
            }], alignCenterClass: [{
                type: HostBinding,
                args: ['class.rds-tree-text-node__container-align-center']
            }], rdsTreeNodeElement: [{
                type: ViewChild,
                args: ['rdsTreeNodeElement', { static: true }]
            }], textAlignment: [{
                type: Input
            }] } });
