// @rds/angular-components/esm2022/lib/tree/tree-node-content-right.directive.mjs
import { Directive, HostBinding, HostListener } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsTreeNodeContentRightDirective {
    constructor() {
        this.contentRightClass = true;
    }
    onClick(event) {
        event.stopPropagation();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTreeNodeContentRightDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsTreeNodeContentRightDirective, selector: "[rdsTreeNodeContentRight], [rds-tree-node-content-right]", host: { listeners: { "keydown.space": "onClick($event)", "click": "onClick($event)" }, properties: { "class.rds-tree-node--content-right": "this.contentRightClass" } }, exportAs: ["rdsTreeNodeContentRight"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTreeNodeContentRightDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rdsTreeNodeContentRight], [rds-tree-node-content-right]',
                    exportAs: 'rdsTreeNodeContentRight',
                }]
        }], propDecorators: { contentRightClass: [{
                type: HostBinding,
                args: ['class.rds-tree-node--content-right']
            }], onClick: [{
                type: HostListener,
                args: ['keydown.space', ['$event']]
            }, {
                type: HostListener,
                args: ['click', ['$event']]
            }] } });
