// @rds/angular-components/esm2022/lib/tree/tree-node-content-left.directive.mjs
import { Directive, HostBinding, HostListener } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsTreeNodeContentLeftDirective {
    constructor() {
        this.contentLeftClass = true;
    }
    onClick(event) {
        event.stopPropagation();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTreeNodeContentLeftDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsTreeNodeContentLeftDirective, selector: "[rdsTreeNodeContentLeft], [rds-tree-node-content-left]", host: { listeners: { "keydown.space": "onClick($event)", "click": "onClick($event)" }, properties: { "class.rds-tree-node--content-left": "this.contentLeftClass" } }, exportAs: ["rdsTreeNodeContentLeft"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTreeNodeContentLeftDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rdsTreeNodeContentLeft], [rds-tree-node-content-left]',
                    exportAs: 'rdsTreeNodeContentLeft',
                }]
        }], propDecorators: { contentLeftClass: [{
                type: HostBinding,
                args: ['class.rds-tree-node--content-left']
            }], onClick: [{
                type: HostListener,
                args: ['keydown.space', ['$event']]
            }, {
                type: HostListener,
                args: ['click', ['$event']]
            }] } });
