// @rds/angular-components/esm2022/lib/tree/tree.module.mjs
import { NgModule } from '@angular/core';
import { RdsTreeNodeDirective } from './tree-node.directive';
import { RdsTreeComponent } from './tree.component';
import { RdsTreeNodeToggleDirective } from './tree-node-toggle.directive';
import { RdsTreeNodeOutletDirective } from './tree-node-outlet.directive';
import { RdsTreeNodeDefDirective } from './tree-node-def.directive';
import { CdkTreeModule } from '@angular/cdk/tree';
import { RdsTreeNodeContentLeftDirective } from './tree-node-content-left.directive';
import { RdsTreeNodeContentRightDirective } from './tree-node-content-right.directive';
import { RdsTreeNodeTextComponent } from './tree-node-text.component';
import * as i0 from "@angular/core";
const RDS_TREE_ELEMENTS = [
    RdsTreeNodeDefDirective,
    RdsTreeNodeToggleDirective,
    RdsTreeComponent,
    RdsTreeNodeDirective,
    RdsTreeNodeOutletDirective,
    RdsTreeNodeContentLeftDirective,
    RdsTreeNodeContentRightDirective,
    RdsTreeNodeTextComponent,
];
export class RdsTreeModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTreeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsTreeModule, declarations: [RdsTreeNodeDefDirective,
            RdsTreeNodeToggleDirective,
            RdsTreeComponent,
            RdsTreeNodeDirective,
            RdsTreeNodeOutletDirective,
            RdsTreeNodeContentLeftDirective,
            RdsTreeNodeContentRightDirective,
            RdsTreeNodeTextComponent], imports: [CdkTreeModule], exports: [RdsTreeNodeDefDirective,
            RdsTreeNodeToggleDirective,
            RdsTreeComponent,
            RdsTreeNodeDirective,
            RdsTreeNodeOutletDirective,
            RdsTreeNodeContentLeftDirective,
            RdsTreeNodeContentRightDirective,
            RdsTreeNodeTextComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTreeModule, imports: [CdkTreeModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTreeModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CdkTreeModule],
                    exports: [...RDS_TREE_ELEMENTS],
                    declarations: [...RDS_TREE_ELEMENTS],
                }]
        }] });
