// @rds/angular-components/esm2022/lib/tree/tree-node-toggle.directive.mjs
import { CdkTreeNodeToggle } from '@angular/cdk/tree';
import { booleanAttribute, Directive, ElementRef, HostListener, inject, Input, } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsTreeNodeToggleDirective extends CdkTreeNodeToggle {
    constructor() {
        super(...arguments);
        this.hostTag = inject(ElementRef).nativeElement.tagName;
        this.recursive = false;
        this.disabled = false;
    }
    onInteraction(event) {
        // CdkTreeNodeToggle listens to (click) and button's enter clicks are handled
        // using that event, that's why we omit it here - interaction would be doubled
        if (this.hostTag !== 'BUTTON') {
            this._toggle(event);
        }
    }
    _toggle(event) {
        if (!this.disabled) {
            super._toggle(event);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTreeNodeToggleDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "16.1.0", version: "17.3.1", type: RdsTreeNodeToggleDirective, selector: "[rdsTreeNodeToggle], [rds-tree-node-toggle]", inputs: { recursive: ["rdsTreeNodeToggleRecursive", "recursive", booleanAttribute], disabled: ["disabled", "disabled", booleanAttribute] }, host: { listeners: { "keydown.enter": "onInteraction($event)" } }, providers: [
            { provide: CdkTreeNodeToggle, useExisting: RdsTreeNodeToggleDirective },
        ], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTreeNodeToggleDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rdsTreeNodeToggle], [rds-tree-node-toggle]',
                    providers: [
                        { provide: CdkTreeNodeToggle, useExisting: RdsTreeNodeToggleDirective },
                    ],
                }]
        }], propDecorators: { recursive: [{
                type: Input,
                args: [{ alias: 'rdsTreeNodeToggleRecursive', transform: booleanAttribute }]
            }], disabled: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }], onInteraction: [{
                type: HostListener,
                args: ['keydown.enter', ['$event']]
            }] } });
