// @rds/angular-components/esm2022/lib/timeline/timeline.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RdsTimelineComponent } from './timeline.component';
import { RdsTimelineEventComponent } from './timeline-event/timeline-event.component';
import { RdsTimelineEventIconMarkerDirective } from './timeline-event/timeline-event-icon-marker.directive';
import { RdsTimelineOppositeDirective } from './timeline-event/timeline-event-opposite.directive';
import { RdsIconComponent } from '../icons';
import { RdsButtonModule } from '../button';
import * as i0 from "@angular/core";
const ELEMENTS = [
    RdsTimelineComponent,
    RdsTimelineEventComponent,
    RdsTimelineEventIconMarkerDirective,
    RdsTimelineOppositeDirective,
];
export class RdsTimelineModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTimelineModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsTimelineModule, declarations: [RdsTimelineComponent,
            RdsTimelineEventComponent,
            RdsTimelineEventIconMarkerDirective,
            RdsTimelineOppositeDirective], imports: [CommonModule, RdsIconComponent, RdsButtonModule], exports: [RdsTimelineComponent,
            RdsTimelineEventComponent,
            RdsTimelineEventIconMarkerDirective,
            RdsTimelineOppositeDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTimelineModule, imports: [CommonModule, RdsButtonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTimelineModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, RdsIconComponent, RdsButtonModule],
                    declarations: [...ELEMENTS],
                    exports: [...ELEMENTS],
                }]
        }] });
