// @rds/angular-components/esm2022/lib/accordion/accordion.module.mjs
import { CdkAccordionModule } from '@angular/cdk/accordion';
import { PortalModule } from '@angular/cdk/portal';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RdsExpansionPanelComponent } from './expansion-panel.component';
import { RdsExpansionPanelContentDirective } from './expansion-panel-content.directive';
import { RdsExpansionPanelHeaderComponent, RdsExpansionPanelTitleDirective, } from './expansion-panel-header.component';
import { RdsIconComponent } from '../icons';
import { RdsAccordionDirective } from './accordion.directive';
import * as i0 from "@angular/core";
const ELEMENTS = [
    RdsAccordionDirective,
    RdsExpansionPanelComponent,
    RdsExpansionPanelHeaderComponent,
    RdsExpansionPanelTitleDirective,
    RdsExpansionPanelContentDirective,
];
export class RdsAccordionModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAccordionModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsAccordionModule, declarations: [RdsAccordionDirective,
            RdsExpansionPanelComponent,
            RdsExpansionPanelHeaderComponent,
            RdsExpansionPanelTitleDirective,
            RdsExpansionPanelContentDirective], imports: [CommonModule, CdkAccordionModule, PortalModule, RdsIconComponent], exports: [RdsAccordionDirective,
            RdsExpansionPanelComponent,
            RdsExpansionPanelHeaderComponent,
            RdsExpansionPanelTitleDirective,
            RdsExpansionPanelContentDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAccordionModule, imports: [CommonModule, CdkAccordionModule, PortalModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAccordionModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, CdkAccordionModule, PortalModule, RdsIconComponent],
                    exports: [...ELEMENTS],
                    declarations: [...ELEMENTS],
                }]
        }] });
