// @rds/angular-components/esm2022/lib/accordion/accordion.directive.mjs
import { booleanAttribute, ContentChildren, Directive, HostBinding, Input, QueryList, } from '@angular/core';
import { FocusKeyManager } from '@angular/cdk/a11y';
import { startWith } from 'rxjs/operators';
import { RDS_ACCORDION, } from './accordion-base';
import { RdsExpansionPanelHeaderComponent } from './expansion-panel-header.component';
import { Subject } from 'rxjs';
import { CdkAccordion } from '@angular/cdk/accordion';
import * as i0 from "@angular/core";
let nextId = 0;
export class RdsAccordionDirective extends CdkAccordion {
    constructor() {
        super(...arguments);
        this._stateChanges = new Subject();
        this._openCloseAllActions = new Subject();
        this.id = `rds-accordion-${nextId++}`;
        this._ownHeaders = new QueryList();
        this.size = 'm';
        this.togglePosition = 'after';
    }
    get class() {
        return ['rds-accordion', `rds-accordion-size-${this.size}`];
    }
    ngAfterContentInit() {
        this._headers.changes
            .pipe(startWith(this._headers))
            .subscribe((headers) => {
            this._ownHeaders.reset(headers.filter((header) => header.panel.accordion === this));
            this._ownHeaders.notifyOnChanges();
        });
        this._keyManager = new FocusKeyManager(this._ownHeaders)
            .withWrap()
            .withHomeAndEnd();
    }
    ngOnDestroy() {
        super.ngOnDestroy();
        this._ownHeaders.destroy();
    }
    _handleHeaderKeydown(event) {
        this._keyManager.onKeydown(event);
    }
    _handleHeaderFocus(header) {
        this._keyManager.updateActiveItem(header);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAccordionDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "16.1.0", version: "17.3.1", type: RdsAccordionDirective, selector: "rds-accordion", inputs: { size: "size", multi: ["multi", "multi", booleanAttribute], togglePosition: "togglePosition" }, host: { properties: { "class": "this.class", "class.rds-accordion-multi": "this.multi" } }, providers: [
            {
                provide: RDS_ACCORDION,
                useExisting: RdsAccordionDirective,
            },
        ], queries: [{ propertyName: "_headers", predicate: RdsExpansionPanelHeaderComponent, descendants: true }], exportAs: ["rdsAccordion"], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAccordionDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'rds-accordion',
                    exportAs: 'rdsAccordion',
                    providers: [
                        {
                            provide: RDS_ACCORDION,
                            useExisting: RdsAccordionDirective,
                        },
                    ],
                }]
        }], propDecorators: { _headers: [{
                type: ContentChildren,
                args: [RdsExpansionPanelHeaderComponent, { descendants: true }]
            }], size: [{
                type: Input
            }], class: [{
                type: HostBinding,
                args: ['class']
            }], multi: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }, {
                type: HostBinding,
                args: ['class.rds-accordion-multi']
            }], togglePosition: [{
                type: Input
            }] } });
