// @rds/angular-components/esm2022/lib/tooltip/tooltip.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RdsTooltipComponent } from './tooltip.component';
import { OverlayModule } from '@angular/cdk/overlay';
import { RDS_TOOLTIP_SCROLL_STRATEGY_FACTORY_PROVIDER } from './tokens';
import { RdsButtonModule } from '../button';
import { RdsIconComponent } from '../icons';
import { RdsTooltipDirective } from './tooltip.directive';
import * as i0 from "@angular/core";
export class RdsTooltipModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTooltipModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsTooltipModule, declarations: [RdsTooltipDirective, RdsTooltipComponent], imports: [CommonModule, OverlayModule, RdsButtonModule, RdsIconComponent], exports: [RdsTooltipDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTooltipModule, providers: [RDS_TOOLTIP_SCROLL_STRATEGY_FACTORY_PROVIDER], imports: [CommonModule, OverlayModule, RdsButtonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsTooltipModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, OverlayModule, RdsButtonModule, RdsIconComponent],
                    declarations: [RdsTooltipDirective, RdsTooltipComponent],
                    exports: [RdsTooltipDirective],
                    providers: [RDS_TOOLTIP_SCROLL_STRATEGY_FACTORY_PROVIDER],
                }]
        }] });
