// @rds/angular-components/esm2022/lib/header/header-searchbar.component.mjs
import { ENTER, hasModifierKey } from '@angular/cdk/keycodes';
import { ChangeDetectionStrategy, Component, ElementRef, EventEmitter, HostBinding, Input, Output, ViewChild, ViewEncapsulation, } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../icons/icon.component";
import * as i2 from "../button/icon-button.component";
import * as i3 from "@angular/forms";
export class RdsHeaderSearchbarComponent {
    constructor() {
        this._defaultPlaceholder = 'Search for the products, page, document…';
        this.class = 'rds-header-search';
        this.searchValue = '';
        this.searchChange = new EventEmitter();
        this.closeSearch = new EventEmitter();
        this.search = new EventEmitter();
    }
    onInputChange(newValue) {
        this.searchValue = newValue;
        this.searchChange.next(this.searchValue);
    }
    onCloseSearch() {
        this.closeSearch.emit();
    }
    onKeyPress(event) {
        if (event.keyCode === ENTER && !hasModifierKey(event)) {
            this.emitValue();
        }
    }
    emitValue() {
        this.search.emit(this.searchValue);
    }
    focusSearch() {
        this.searchInput.nativeElement.focus();
    }
    getPlaceholder() {
        return this.placeholder || this._defaultPlaceholder;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderSearchbarComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsHeaderSearchbarComponent, selector: "rds-header-searchbar", inputs: { placeholder: "placeholder", searchValue: "searchValue" }, outputs: { searchChange: "searchChange", closeSearch: "closeSearch", search: "search" }, host: { properties: { "class": "this.class" } }, viewQueries: [{ propertyName: "searchInput", first: true, predicate: ["input"], descendants: true, static: true }], ngImport: i0, template: `
    <button
      rds-icon-button
      size="m"
      class="rds-header-search-search-icon"
      (click)="emitValue()"
    >
      <rds-icon namespace="outlined" icon="search" />
    </button>
    <input
      #input
      [ngModel]="searchValue"
      (ngModelChange)="onInputChange($event)"
      (keydown)="onKeyPress($event)"
      class="rds-header-search-input"
      [placeholder]="getPlaceholder()"
    />
    <button
      rds-icon-button
      (click)="onCloseSearch()"
      class="rds-header-search-close-icon"
    >
      <rds-icon namespace="outlined" icon="close" />
    </button>
  `, isInline: true, dependencies: [{ kind: "component", type: i1.RdsIconComponent, selector: "rds-icon", inputs: ["namespace", "icon"] }, { kind: "component", type: i2.RdsIconButtonComponent, selector: "\n    button[rds-icon-button],\n    button[rdsIconButton],\n    button[rds-icon-warning-button],\n    button[rdsIconWarningButton],\n  ", inputs: ["size"] }, { kind: "directive", type: i3.DefaultValueAccessor, selector: "input:not([type=checkbox])[formControlName],textarea[formControlName],input:not([type=checkbox])[formControl],textarea[formControl],input:not([type=checkbox])[ngModel],textarea[ngModel],[ngDefaultControl]" }, { kind: "directive", type: i3.NgControlStatus, selector: "[formControlName],[ngModel],[formControl]" }, { kind: "directive", type: i3.NgModel, selector: "[ngModel]:not([formControlName]):not([formControl])", inputs: ["name", "disabled", "ngModel", "ngModelOptions"], outputs: ["ngModelChange"], exportAs: ["ngModel"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderSearchbarComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'rds-header-searchbar',
                    template: `
    <button
      rds-icon-button
      size="m"
      class="rds-header-search-search-icon"
      (click)="emitValue()"
    >
      <rds-icon namespace="outlined" icon="search" />
    </button>
    <input
      #input
      [ngModel]="searchValue"
      (ngModelChange)="onInputChange($event)"
      (keydown)="onKeyPress($event)"
      class="rds-header-search-input"
      [placeholder]="getPlaceholder()"
    />
    <button
      rds-icon-button
      (click)="onCloseSearch()"
      class="rds-header-search-close-icon"
    >
      <rds-icon namespace="outlined" icon="close" />
    </button>
  `,
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }], placeholder: [{
                type: Input
            }], searchValue: [{
                type: Input
            }], searchChange: [{
                type: Output
            }], closeSearch: [{
                type: Output
            }], search: [{
                type: Output
            }], searchInput: [{
                type: ViewChild,
                args: ['input', { static: true }]
            }] } });
