// @rds/angular-components/esm2022/lib/header/header-app-name.directive.mjs
import { Directive, ElementRef, EventEmitter, HostBinding, HostListener, inject, Output, } from '@angular/core';
import { ENTER, hasModifierKey, SPACE } from '@angular/cdk/keycodes';
import { FocusMonitor } from '@angular/cdk/a11y';
import { TAB_INDEX_DEFAULT, TAB_INDEX_NOT_FOCUSABLE } from '../utils';
import * as i0 from "@angular/core";
export class RdsHeaderAppNameDirective {
    constructor() {
        this.elementRef = inject(ElementRef);
        this.focusMonitor = inject(FocusMonitor);
        this.class = 'rds-header-app-name';
        this.triggered = new EventEmitter();
    }
    get isTriggerable() {
        return this.triggered.observed;
    }
    get tabIndex() {
        return this.isTriggerable ? TAB_INDEX_DEFAULT : TAB_INDEX_NOT_FOCUSABLE;
    }
    ngOnInit() {
        this.focusMonitor.monitor(this.elementRef.nativeElement);
    }
    ngOnDestroy() {
        this.focusMonitor.stopMonitoring(this.elementRef.nativeElement);
    }
    onLogoClick(event) {
        if ((event instanceof KeyboardEvent &&
            !hasModifierKey(event, 'shiftKey', 'altKey', 'ctrlKey') &&
            (event.keyCode === SPACE || event.keyCode === ENTER)) ||
            event instanceof MouseEvent) {
            if (event instanceof KeyboardEvent && event.keyCode === SPACE) {
                event.stopPropagation();
                event.preventDefault();
            }
            this.triggered.next(event);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderAppNameDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsHeaderAppNameDirective, selector: "rds-header-app-name", outputs: { triggered: "triggered" }, host: { listeners: { "click": "onLogoClick($event)", "keydown": "onLogoClick($event)" }, properties: { "class": "this.class", "class.rds-header-app-name-triggerable": "this.isTriggerable", "tabIndex": "this.tabIndex" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderAppNameDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'rds-header-app-name',
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }], isTriggerable: [{
                type: HostBinding,
                args: ['class.rds-header-app-name-triggerable']
            }], tabIndex: [{
                type: HostBinding,
                args: ['tabIndex']
            }], triggered: [{
                type: Output
            }], onLogoClick: [{
                type: HostListener,
                args: ['click', ['$event']]
            }, {
                type: HostListener,
                args: ['keydown', ['$event']]
            }] } });
