// @rds/angular-components/esm2022/lib/header/header-start-divider.directive.mjs
import { Directive, forwardRef, HostBinding, inject } from '@angular/core';
import { RdsDividerComponent } from '../divider';
import * as i0 from "@angular/core";
export class RdsHeaderStartDividerDirective {
    constructor() {
        this.divider = inject(forwardRef(() => RdsDividerComponent));
        this.class = 'rds-header-start-divider';
        this.divider.vertical = true;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderStartDividerDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsHeaderStartDividerDirective, selector: "rds-divider[rds-header-start-divider]", host: { properties: { "class": "this.class" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderStartDividerDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'rds-divider[rds-header-start-divider]',
                }]
        }], ctorParameters: () => [], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }] } });
