// @rds/angular-components/esm2022/lib/header/header.module.mjs
import { ObserversModule } from '@angular/cdk/observers';
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { RdsButtonModule } from '../button';
import { RdsFormFieldModule } from '../form-field';
import { RdsLinkComponent } from '../link';
import { RdsIconComponent } from '../icons';
import { RdsHeaderActionsDirective } from './header-actions.directive';
import { RdsHeaderAppNameDirective } from './header-app-name.directive';
import { RdsHeaderLinkComponent } from './header-link.component';
import { RdsHeaderNavigationPrimaryDirective } from './header-navigation-primary.directive';
import { RdsHeaderNavigationSecondaryDirective } from './header-navigation-secondary.directive';
import { RdsHeaderSearchbarComponent } from './header-searchbar.component';
import { RdsHeaderComponent } from './header.component';
import { RdsSearchModule } from '../search';
import { RdsHeaderStartDividerDirective } from './header-start-divider.directive';
import { RdsDividerModule } from '../divider';
import { RdsHeaderEndDividerDirective } from './header-end-divider.directive';
import * as i0 from "@angular/core";
const ELEMENTS = [
    RdsHeaderComponent,
    RdsHeaderActionsDirective,
    RdsHeaderAppNameDirective,
    RdsHeaderLinkComponent,
    RdsHeaderNavigationPrimaryDirective,
    RdsHeaderNavigationSecondaryDirective,
    RdsHeaderSearchbarComponent,
    RdsHeaderStartDividerDirective,
    RdsHeaderEndDividerDirective,
];
export class RdsHeaderModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderModule, declarations: [RdsHeaderComponent,
            RdsHeaderActionsDirective,
            RdsHeaderAppNameDirective,
            RdsHeaderLinkComponent,
            RdsHeaderNavigationPrimaryDirective,
            RdsHeaderNavigationSecondaryDirective,
            RdsHeaderSearchbarComponent,
            RdsHeaderStartDividerDirective,
            RdsHeaderEndDividerDirective], imports: [CommonModule,
            RdsIconComponent,
            OverlayModule,
            RdsFormFieldModule,
            RdsButtonModule,
            FormsModule,
            ObserversModule,
            RdsLinkComponent,
            RdsSearchModule,
            RdsDividerModule], exports: [RdsHeaderComponent,
            RdsHeaderActionsDirective,
            RdsHeaderAppNameDirective,
            RdsHeaderLinkComponent,
            RdsHeaderNavigationPrimaryDirective,
            RdsHeaderNavigationSecondaryDirective,
            RdsHeaderSearchbarComponent,
            RdsHeaderStartDividerDirective,
            RdsHeaderEndDividerDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderModule, imports: [CommonModule,
            OverlayModule,
            RdsFormFieldModule,
            RdsButtonModule,
            FormsModule,
            ObserversModule,
            RdsSearchModule,
            RdsDividerModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        RdsIconComponent,
                        OverlayModule,
                        RdsFormFieldModule,
                        RdsButtonModule,
                        FormsModule,
                        ObserversModule,
                        RdsLinkComponent,
                        RdsSearchModule,
                        RdsDividerModule,
                    ],
                    declarations: [...ELEMENTS],
                    exports: [...ELEMENTS],
                }]
        }] });
