// @rds/angular-components/esm2022/lib/header/header-link.component.mjs
import { ENTER, hasModifierKey, SPACE } from '@angular/cdk/keycodes';
import { ChangeDetectionStrategy, Component, EventEmitter, HostBinding, HostListener, Output, ViewEncapsulation, } from '@angular/core';
import { RdsLinkComponent } from '../link';
import * as i0 from "@angular/core";
export class RdsHeaderLinkComponent extends RdsLinkComponent {
    constructor() {
        super(...arguments);
        this.class = 'rds-header-link';
        this.toggle = new EventEmitter();
        this._standalone = true;
    }
    ngOnInit() {
        this.type = 'secondary';
        this.size = 's';
    }
    onKeydown(event) {
        const code = event.keyCode;
        if ((code === ENTER || code === SPACE) && !hasModifierKey(event)) {
            this.toggle.next(!this.active);
            event.preventDefault();
        }
    }
    onClick() {
        if (!this.disabled) {
            this.toggle.next(!this.active);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderLinkComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsHeaderLinkComponent, selector: "rds-header-link", outputs: { toggle: "toggle" }, host: { listeners: { "keydown": "onKeydown($event)", "click": "onClick()" }, properties: { "class": "this.class" } }, usesInheritance: true, ngImport: i0, template: `<ng-content />`, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderLinkComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'rds-header-link',
                    template: `<ng-content />`,
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], propDecorators: { class: [{
                type: HostBinding,
                args: ['class']
            }], toggle: [{
                type: Output
            }], onKeydown: [{
                type: HostListener,
                args: ['keydown', ['$event']]
            }], onClick: [{
                type: HostListener,
                args: ['click']
            }] } });
