// @rds/angular-components/esm2022/lib/sidenav/sidenav-menu-header/sidenav-menu-header.component.mjs
import { ChangeDetectionStrategy, Component, ElementRef, HostBinding, inject, } from '@angular/core';
import { FocusMonitor } from '@angular/cdk/a11y';
import { filter } from 'rxjs/operators';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import * as i0 from "@angular/core";
import * as i1 from "../../icons/icon.component";
export class RdsSidenavMenuHeaderComponent {
    constructor() {
        this.focusMonitor = inject(FocusMonitor);
        this.nativeElement = inject(ElementRef).nativeElement;
        this.disabled = false;
        this._focused = this.focusMonitor.monitor(this.nativeElement, false).pipe(filter((v) => !!v), takeUntilDestroyed());
        this.classes = 'rds-sidenav-menu-header';
    }
    ngOnDestroy() {
        this.focusMonitor.stopMonitoring(this.nativeElement);
    }
    focus(origin, options) {
        if (origin) {
            this.focusMonitor.focusVia(this.nativeElement, origin, options);
        }
        else {
            this.nativeElement.focus(options);
        }
    }
    getLabel() {
        const clone = this.nativeElement.cloneNode(true);
        const icons = clone.querySelectorAll('rds-icon');
        for (let i = 0; i < icons.length; i++) {
            icons[i].remove();
        }
        return clone.textContent?.trim() || '';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSidenavMenuHeaderComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsSidenavMenuHeaderComponent, selector: "a[rds-sidenav-menu-header], a[rdsSidenavMenuHeader]", host: { properties: { "class": "this.classes" } }, ngImport: i0, template: "<ng-content /> <rds-icon icon=\"arrow_right\" />\n", dependencies: [{ kind: "component", type: i1.RdsIconComponent, selector: "rds-icon", inputs: ["namespace", "icon"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSidenavMenuHeaderComponent, decorators: [{
            type: Component,
            args: [{ selector: 'a[rds-sidenav-menu-header], a[rdsSidenavMenuHeader]', changeDetection: ChangeDetectionStrategy.OnPush, template: "<ng-content /> <rds-icon icon=\"arrow_right\" />\n" }]
        }], propDecorators: { classes: [{
                type: HostBinding,
                args: ['class']
            }] } });
