// @rds/angular-components/esm2022/lib/toast/overlay/rds-toast-overlay-container.service.mjs
import { DOCUMENT } from '@angular/common';
import { inject, Injectable } from '@angular/core';
import { RdsToastModule } from '../rds-toast.module';
import * as i0 from "@angular/core";
export class RdsToastOverlayContainerService {
    constructor() {
        this.document = inject(DOCUMENT);
    }
    ngOnDestroy() {
        if (this.containerElement && this.containerElement.parentNode) {
            this.containerElement.parentNode.removeChild(this.containerElement);
        }
    }
    getContainerElement() {
        if (!this.containerElement) {
            this._createContainer();
        }
        return this.containerElement;
    }
    _createContainer() {
        const container = this.document.createElement('div');
        container.classList.add('rds-toast-overlay-container');
        this.document.body.appendChild(container);
        this.containerElement = container;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsToastOverlayContainerService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsToastOverlayContainerService, providedIn: RdsToastModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsToastOverlayContainerService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: RdsToastModule,
                }]
        }] });
