// @rds/angular-components/esm2022/lib/toast/overlay/rds-toast-overlay.service.mjs
import { DOCUMENT } from '@angular/common';
import { ApplicationRef, ComponentFactoryResolver, inject, Injectable, } from '@angular/core';
import { RdsDomPortalHost } from '../portal/rds-dom-portal-host';
import { RdsToastOverlayContainerService } from './rds-toast-overlay-container.service';
import { RdsToastOverlayRef } from './rds-toast-overlay-ref';
import { RdsToastModule } from '../rds-toast.module';
import * as i0 from "@angular/core";
export class RdsToastOverlayService {
    constructor() {
        this.toastOverlayContainerService = inject(RdsToastOverlayContainerService);
        this.componentFactoryResolver = inject(ComponentFactoryResolver);
        this.applicationRef = inject(ApplicationRef);
        this.document = inject(DOCUMENT);
        this._paneElements = new Map();
    }
    create(position, overlayContainer) {
        return this._createOverlayRef(this._getPaneElement(position, overlayContainer));
    }
    _getPaneElement(positionClass, overlayContainer) {
        const posClass = `rds-toast-${positionClass}`;
        if (!this._paneElements.get(overlayContainer)) {
            this._paneElements.set(overlayContainer, {});
        }
        const paneElement = this._paneElements.get(overlayContainer);
        if (!paneElement[posClass]) {
            paneElement[posClass] = this._createPaneElement(posClass, overlayContainer);
        }
        return paneElement[posClass];
    }
    _createPaneElement(positionClass, overlayContainer) {
        const pane = this.document.createElement('div');
        pane.id = 'toast-container';
        pane.classList.add(positionClass);
        pane.classList.add('rds-toast-container');
        if (!overlayContainer) {
            this.toastOverlayContainerService.getContainerElement().appendChild(pane);
        }
        else {
            overlayContainer.getContainerElement().appendChild(pane);
        }
        return pane;
    }
    _createPortalHost(pane) {
        return new RdsDomPortalHost(pane, this.componentFactoryResolver, this.applicationRef);
    }
    _createOverlayRef(pane) {
        return new RdsToastOverlayRef(this._createPortalHost(pane));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsToastOverlayService, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsToastOverlayService, providedIn: RdsToastModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsToastOverlayService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: RdsToastModule,
                }]
        }] });
