// @rds/angular-components/esm2022/lib/divider/divider.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RdsDividerComponent } from './divider.component';
import * as i0 from "@angular/core";
export class RdsDividerModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDividerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsDividerModule, declarations: [RdsDividerComponent], imports: [CommonModule], exports: [RdsDividerComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDividerModule, imports: [CommonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDividerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [RdsDividerComponent],
                    exports: [RdsDividerComponent],
                    imports: [CommonModule],
                }]
        }] });
