// @rds/angular-components/esm2022/lib/divider/divider.component.mjs
import { booleanAttribute, ChangeDetectionStrategy, Component, HostBinding, Input, ViewEncapsulation, } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsDividerComponent {
    constructor() {
        this.hostRole = 'separator';
        this.class = 'rds-divider';
        this.vertical = false;
    }
    get orientation() {
        return this.vertical ? 'vertical' : 'horizontal';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDividerComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "16.1.0", version: "17.3.1", type: RdsDividerComponent, selector: "rds-divider", inputs: { vertical: ["vertical", "vertical", booleanAttribute] }, host: { properties: { "attr.role": "this.hostRole", "class": "this.class", "attr.aria-orientation": "this.orientation", "class.rds-divider-vertical": "this.vertical" } }, ngImport: i0, template: '', isInline: true, styles: [".rds-divider{display:block;width:100%;border-bottom:1px solid var(--rds-color-grey-2);margin:0}.rds-divider.rds-divider-vertical{border-bottom:0;border-right:1px solid var(--rds-color-grey-2);height:100%}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsDividerComponent, decorators: [{
            type: Component,
            args: [{ selector: 'rds-divider', template: '', encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, styles: [".rds-divider{display:block;width:100%;border-bottom:1px solid var(--rds-color-grey-2);margin:0}.rds-divider.rds-divider-vertical{border-bottom:0;border-right:1px solid var(--rds-color-grey-2);height:100%}\n"] }]
        }], propDecorators: { hostRole: [{
                type: HostBinding,
                args: ['attr.role']
            }], class: [{
                type: HostBinding,
                args: ['class']
            }], orientation: [{
                type: HostBinding,
                args: ['attr.aria-orientation']
            }], vertical: [{
                type: HostBinding,
                args: ['class.rds-divider-vertical']
            }, {
                type: Input,
                args: [{ transform: booleanAttribute }]
            }] } });
