// @rds/angular-components/esm2022/lib/switch/switch.module.mjs
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { RdsCheckboxModule } from '../checkbox';
import { RdsSwitchRequiredValidatorDirective } from './switch-required-validator';
import { RdsSwitchComponent } from './switch.component';
import { ObserversModule } from '@angular/cdk/observers';
import { BidiModule } from '@angular/cdk/bidi';
import { CommonModule } from '@angular/common';
import { RdsHelperTextModule } from '../form-field/helper-text.module';
import * as i0 from "@angular/core";
export class RdsSwitchModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSwitchModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsSwitchModule, declarations: [RdsSwitchComponent, RdsSwitchRequiredValidatorDirective], imports: [BidiModule,
            FormsModule,
            RdsCheckboxModule,
            ObserversModule,
            CommonModule,
            RdsHelperTextModule], exports: [RdsSwitchComponent, RdsSwitchRequiredValidatorDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSwitchModule, imports: [BidiModule,
            FormsModule,
            RdsCheckboxModule,
            ObserversModule,
            CommonModule,
            RdsHelperTextModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSwitchModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        BidiModule,
                        FormsModule,
                        RdsCheckboxModule,
                        ObserversModule,
                        CommonModule,
                        RdsHelperTextModule,
                    ],
                    declarations: [RdsSwitchComponent, RdsSwitchRequiredValidatorDirective],
                    exports: [RdsSwitchComponent, RdsSwitchRequiredValidatorDirective],
                }]
        }] });
