// @rds/angular-components/esm2022/lib/switch/switch-required-validator.mjs
import { Directive, forwardRef } from '@angular/core';
import { CheckboxRequiredValidator, NG_VALIDATORS } from '@angular/forms';
import * as i0 from "@angular/core";
export const RDS_SWITCH_REQUIRED_VALIDATOR = {
    provide: NG_VALIDATORS,
    useExisting: forwardRef(() => RdsSwitchRequiredValidatorDirective),
    multi: true,
};
export class RdsSwitchRequiredValidatorDirective extends CheckboxRequiredValidator {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSwitchRequiredValidatorDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsSwitchRequiredValidatorDirective, selector: "rds-switch[required][formControlName],\n              rds-switch[required][formControl], rds-switch[required][ngModel]", providers: [RDS_SWITCH_REQUIRED_VALIDATOR], usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSwitchRequiredValidatorDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: `rds-switch[required][formControlName],
              rds-switch[required][formControl], rds-switch[required][ngModel]`,
                    providers: [RDS_SWITCH_REQUIRED_VALIDATOR],
                }]
        }] });
