// @rds/angular-components/esm2022/lib/list-item/abstract-selectable-list-item.component.mjs
import { ContentObserver } from '@angular/cdk/observers';
import { booleanAttribute, ChangeDetectorRef, Component, ContentChild, DestroyRef, EventEmitter, HostBinding, HostListener, inject, Input, Output, } from '@angular/core';
import { RdsAvatarComponent } from '../avatar';
import { RdsListItemComponent } from './list-item.component';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import * as i0 from "@angular/core";
const RDS_LIST_ITEM_SIZE_TO_AVATAR_SIZE_MAP = {
    s: 'xs',
    m: 's',
    l: 'm',
};
export class RdsAbstractSelectableListItemComponent extends RdsListItemComponent {
    constructor() {
        super(...arguments);
        this.destroyRef = inject(DestroyRef);
        this.contentObserver = inject(ContentObserver);
        this.cdr = inject(ChangeDetectorRef);
        this.checkedChanged = new EventEmitter();
        this.role = 'option';
        this.checked = false;
    }
    get cssClasses() {
        return [...super.cssClasses, this.getSelectTypeClass()];
    }
    ngAfterViewInit() {
        this.contentObserver
            .observe(this.elementRef.nativeElement)
            .pipe(takeUntilDestroyed(this.destroyRef))
            .subscribe(() => {
            if (this.avatar) {
                this.avatar.disableInteractions = true;
                this.adjustAvatarSize();
            }
            this.cdr.markForCheck();
        });
    }
    ngOnChanges(changes) {
        if (changes['size']) {
            this.adjustAvatarSize();
        }
    }
    toggle() {
        if (!this.disabled) {
            this.checked = !this.checked;
            this.checkedChanged.emit(this.checked);
            this.cdr.markForCheck();
        }
    }
    adjustAvatarSize() {
        if (this.avatar &&
            this.avatar.size !== RDS_LIST_ITEM_SIZE_TO_AVATAR_SIZE_MAP[this.size]) {
            this.avatar.size = RDS_LIST_ITEM_SIZE_TO_AVATAR_SIZE_MAP[this.size];
        }
    }
    focus(origin) {
        this.renderer.addClass(this.elementRef.nativeElement, 'cdk-focused');
        if (origin === 'keyboard') {
            this.renderer.addClass(this.elementRef.nativeElement, 'cdk-keyboard-focused');
        }
        this.cdr.markForCheck();
        this._focused.next(this);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAbstractSelectableListItemComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "16.1.0", version: "17.3.1", type: RdsAbstractSelectableListItemComponent, selector: "ng-component", inputs: { role: "role", checked: ["checked", "checked", booleanAttribute] }, outputs: { checkedChanged: "checkedChanged" }, host: { listeners: { "click": "toggle()" }, properties: { "attr.role": "this.role", "class": "this.cssClasses", "class.rds-list-item--checked": "this.checked" } }, queries: [{ propertyName: "avatar", first: true, predicate: RdsAvatarComponent, descendants: true }], usesInheritance: true, usesOnChanges: true, ngImport: i0, template: '', isInline: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsAbstractSelectableListItemComponent, decorators: [{
            type: Component,
            args: [{ template: '' }]
        }], propDecorators: { avatar: [{
                type: ContentChild,
                args: [RdsAvatarComponent]
            }], checkedChanged: [{
                type: Output
            }], role: [{
                type: Input
            }, {
                type: HostBinding,
                args: ['attr.role']
            }], cssClasses: [{
                type: HostBinding,
                args: ['class']
            }], checked: [{
                type: HostBinding,
                args: ['class.rds-list-item--checked']
            }, {
                type: Input,
                args: [{ transform: booleanAttribute }]
            }], toggle: [{
                type: HostListener,
                args: ['click']
            }] } });
