// @rds/angular-components/esm2022/lib/list-item/header-list-item.directive.mjs
import { Directive, HostBinding, Input } from '@angular/core';
import * as i0 from "@angular/core";
export class RdsHeaderListItemDirective {
    constructor() {
        this.size = 'l';
    }
    get class() {
        return `rds-header-list-item rds-header-list-item--${this.size}`;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderListItemDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.1", type: RdsHeaderListItemDirective, selector: "rds-header-list-item, [rds-header-list-item], [rdsHeaderListItem]", inputs: { size: "size" }, host: { properties: { "class": "this.class" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsHeaderListItemDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'rds-header-list-item, [rds-header-list-item], [rdsHeaderListItem]',
                }]
        }], propDecorators: { size: [{
                type: Input
            }], class: [{
                type: HostBinding,
                args: ['class']
            }] } });
