// @rds/angular-components/esm2022/lib/list-item/list-item.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RdsCheckboxModule } from '../checkbox';
import { RdsIconComponent } from '../icons';
import { RdsHeaderListItemDirective } from './header-list-item.directive';
import { RdsListItemComponent } from './list-item.component';
import { RdsMultiSelectListItemComponent } from './multi-select-list-item.component';
import { RdsNestableListItemComponent } from './nestable-list-item.component';
import { RdsListItemContainerDirective } from './list-item-container.directive';
import { RdsSingleSelectListItemComponent } from './single-select-list-item.component';
import * as i0 from "@angular/core";
const COMPONENTS = [
    RdsHeaderListItemDirective,
    RdsNestableListItemComponent,
    RdsListItemComponent,
    RdsListItemContainerDirective,
    RdsMultiSelectListItemComponent,
    RdsSingleSelectListItemComponent,
];
export class RdsListItemModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsListItemModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.1", ngImport: i0, type: RdsListItemModule, declarations: [RdsHeaderListItemDirective,
            RdsNestableListItemComponent,
            RdsListItemComponent,
            RdsListItemContainerDirective,
            RdsMultiSelectListItemComponent,
            RdsSingleSelectListItemComponent], imports: [CommonModule, RdsIconComponent, RdsCheckboxModule], exports: [RdsHeaderListItemDirective,
            RdsNestableListItemComponent,
            RdsListItemComponent,
            RdsListItemContainerDirective,
            RdsMultiSelectListItemComponent,
            RdsSingleSelectListItemComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsListItemModule, imports: [CommonModule, RdsCheckboxModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsListItemModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, RdsIconComponent, RdsCheckboxModule],
                    declarations: [...COMPONENTS],
                    exports: [...COMPONENTS],
                }]
        }] });
