// @rds/angular-components/esm2022/lib/list-item/multi-select-list-item.component.mjs
import { ChangeDetectionStrategy, Component, HostBinding, ViewEncapsulation, } from '@angular/core';
import { RdsListItemComponent } from './list-item.component';
import { RDS_MULTI_SELECT_LIST_ITEM_TEMPLATE } from './list-item.templates';
import { RdsAbstractSelectableListItemComponent } from './abstract-selectable-list-item.component';
import * as i0 from "@angular/core";
import * as i1 from "../checkbox/pseudo-checkbox.component";
export class RdsMultiSelectListItemComponent extends RdsAbstractSelectableListItemComponent {
    get checkedBindings() {
        return this.checked;
    }
    getSelectTypeClass() {
        return 'rds-multi-select-list-item';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsMultiSelectListItemComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "17.3.1", type: RdsMultiSelectListItemComponent, selector: "button[rds-multi-select-list-item], button[rdsMultiSelectListItem]", host: { properties: { "class.rds-list-item--checked": "this.checkedBindings", "attr.aria-checked": "this.checkedBindings" } }, providers: [
            {
                provide: RdsListItemComponent,
                useExisting: RdsMultiSelectListItemComponent,
            },
        ], usesInheritance: true, ngImport: i0, template: "<ng-content select=\"rds-avatar\"/>\n<ng-content select=\"rds-icon\"/>\n@if (!avatar && !icon) {\n    <rds-pseudo-checkbox\n      noSpacing\n      [checked]=\"checked\"\n      [disabled]=\"disabled\"\n    />\n}\n<span class=\"rds-list-item-text\">\n  <ng-content/>\n</span>\n@if (avatar || icon) {\n    <rds-pseudo-checkbox\n      noSpacing\n      [checked]=\"checked\"\n      [disabled]=\"disabled\"\n      class=\"rds-list-item__checkbox-right\"\n    />\n}\n", isInline: true, dependencies: [{ kind: "component", type: i1.RdsPseudoCheckboxComponent, selector: "rds-pseudo-checkbox", inputs: ["checked", "disabled", "indeterminate", "noSpacing"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsMultiSelectListItemComponent, decorators: [{
            type: Component,
            args: [{
                    template: RDS_MULTI_SELECT_LIST_ITEM_TEMPLATE,
                    selector: 'button[rds-multi-select-list-item], button[rdsMultiSelectListItem]',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    providers: [
                        {
                            provide: RdsListItemComponent,
                            useExisting: RdsMultiSelectListItemComponent,
                        },
                    ],
                }]
        }], propDecorators: { checkedBindings: [{
                type: HostBinding,
                args: ['class.rds-list-item--checked']
            }, {
                type: HostBinding,
                args: ['attr.aria-checked']
            }] } });
