// @rds/angular-components/esm2022/lib/list-item/nestable-list-item.component.mjs
import { booleanAttribute, ChangeDetectionStrategy, Component, HostBinding, Input, ViewEncapsulation, } from '@angular/core';
import { RdsListItemComponent } from './list-item.component';
import { RDS_NESTABLE_LIST_ITEM_TEMPLATE } from './list-item.templates';
import * as i0 from "@angular/core";
import * as i1 from "../icons/icon.component";
export class RdsNestableListItemComponent extends RdsListItemComponent {
    constructor() {
        super(...arguments);
        this.hasChildItems = false;
        this.activated = false;
    }
    get cssClasses() {
        const classes = [...super.cssClasses, 'rds-nestable-list-item'];
        if (!this.disabled && this.activated) {
            classes.push('rds-nestable-list-item-activated');
        }
        return classes;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsNestableListItemComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "17.3.1", type: RdsNestableListItemComponent, selector: "button[rds-nestable-list-item], button[rdsNestableListItem]", inputs: { hasChildItems: ["hasChildItems", "hasChildItems", booleanAttribute], activated: ["activated", "activated", booleanAttribute] }, host: { properties: { "class": "this.cssClasses" } }, providers: [
            {
                provide: RdsListItemComponent,
                useExisting: RdsNestableListItemComponent,
            },
        ], usesInheritance: true, ngImport: i0, template: "<ng-content select=\"rds-icon\"></ng-content>\n<span class=\"rds-list-item-text\">\n  <ng-content />\n</span>\n\n@if (hasChildItems) {\n  <rds-icon\n    namespace=\"outlined\"\n    icon=\"chevron_right\"\n    class=\"rds-list-item__has-child-items\"\n  />\n}", isInline: true, dependencies: [{ kind: "component", type: i1.RdsIconComponent, selector: "rds-icon", inputs: ["namespace", "icon"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsNestableListItemComponent, decorators: [{
            type: Component,
            args: [{
                    template: RDS_NESTABLE_LIST_ITEM_TEMPLATE,
                    selector: 'button[rds-nestable-list-item], button[rdsNestableListItem]',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    providers: [
                        {
                            provide: RdsListItemComponent,
                            useExisting: RdsNestableListItemComponent,
                        },
                    ],
                }]
        }], propDecorators: { cssClasses: [{
                type: HostBinding,
                args: ['class']
            }], hasChildItems: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }], activated: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }] } });
