// @rds/angular-components/esm2022/lib/list-item/single-select-list-item.component.mjs
import { ChangeDetectionStrategy, Component, HostBinding, ViewEncapsulation, } from '@angular/core';
import { RdsListItemComponent } from './list-item.component';
import { RDS_SINGLE_SELECT_LIST_ITEM_TEMPLATE } from './list-item.templates';
import { RdsAbstractSelectableListItemComponent } from './abstract-selectable-list-item.component';
import * as i0 from "@angular/core";
import * as i1 from "../icons/icon.component";
export class RdsSingleSelectListItemComponent extends RdsAbstractSelectableListItemComponent {
    get checkedBindings() {
        return this.checked;
    }
    getSelectTypeClass() {
        return 'rds-single-select-list-item';
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSingleSelectListItemComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.1", type: RdsSingleSelectListItemComponent, selector: "button[rds-single-select-list-item], button[rdsSingleSelectListItem]", host: { properties: { "class.rds-list-item--checked": "this.checkedBindings", "attr.aria-selected": "this.checkedBindings" } }, providers: [
            {
                provide: RdsListItemComponent,
                useExisting: RdsSingleSelectListItemComponent,
            },
        ], usesInheritance: true, ngImport: i0, template: "<ng-content select=\"rds-avatar\"/>\n<ng-content select=\"rds-icon\"></ng-content>\n<span class=\"rds-list-item-text\">\n  <ng-content />\n</span>\n\n<rds-icon icon=\"checkmark\" class=\"rds-list-item-checkmark\"/>", isInline: true, dependencies: [{ kind: "component", type: i1.RdsIconComponent, selector: "rds-icon", inputs: ["namespace", "icon"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.1", ngImport: i0, type: RdsSingleSelectListItemComponent, decorators: [{
            type: Component,
            args: [{
                    template: RDS_SINGLE_SELECT_LIST_ITEM_TEMPLATE,
                    selector: 'button[rds-single-select-list-item], button[rdsSingleSelectListItem]',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    encapsulation: ViewEncapsulation.None,
                    providers: [
                        {
                            provide: RdsListItemComponent,
                            useExisting: RdsSingleSelectListItemComponent,
                        },
                    ],
                }]
        }], propDecorators: { checkedBindings: [{
                type: HostBinding,
                args: ['class.rds-list-item--checked']
            }, {
                type: HostBinding,
                args: ['attr.aria-selected']
            }] } });
